--- 
title: "R statistics"
author: "Mark Goldberg"
date: "`r Sys.Date()`"
site: bookdown::bookdown_site
output: bookdown::gitbook
documentclass: book
bibliography: [book.bib, packages.bib]
biblio-style: apalike
link-citations: yes
github-repo: rstudio/bookdown-demo
description: "This is a minimal example of using the bookdown package to write a book. The output format for this example is bookdown::gitbook."
---

# Introduction

This is the collection of statistical methods in R.  

<!--chapter:end:index.Rmd-->

# Statistics R functions reference

## Get data

## Data inspection
```{r, eval = FALSE}
head(df)
typeof(df)  # data type
dim(df)     # dimention
nrow(df)    # number of rows
ncol(df)    # number of columns
str(df)     # data structure
summary(df) # data summary
names(df)   # names of columns
colnames(df) # also column names

table(df)	# frequency of categorical data
```

## Plots
```{r, eval=FALSE}
plot(x ~ y)
barplot(df)
boxplot(v)
hist(x)
pie(groupsize, labels, col, ...)
```

## Analysis of the distribution
```{r, eval = FALSE}
# mode
getmode <- function(v) {
   uniqv <- unique(v)
   uniqv[which.max(tabulate(match(v, uniqv)))]
}

getmode(v)        # mode
mean(v)           # mean
mean(v, trim=0.1) # trimmed mean
median(v)         # median

min(v)
max(v)
range(v)          # c(mun(v), max(v))
max(v)-min(v)   # range

sort(v)
rank(v)
sample(v)        # subsample
sample(v, size)

var(v)            # variance
sd(v)             # standard deviation
cor(v)            # correlation
cov(v)            # covariation
scale(v)          # z-scores

quantile(v)
IQR(v)            # interquantile range: IQR = Q3 ‚Äì Q1

qqnorm(v)         # normal probability plot
qqline(v)         # adds a line to a normal probability plot passing through 1Q and 3Q
```

## Distributions

**d** - probability density  
**p** - cumulative distribution  
**q** - quantile function (inverse cumulative distribution)  
**r** - random variables from distribution  
```{r, eval=FALSE}
# Normal distribution
dnorm(x, mean, sd)     # Probability Density Function (PDF)
pnorm(q, mean, sd)     # Cumulative Distribution Function (CDF)
qnorm(p, mean, sd)     # quantile function - inverse of pnorm
rnorm(n, mean, sd)     # random numbers from normal distribution

# Binomial
dbinom(x)     # Probability Density Function (PDF)
pbinom(q)     # Cumulative Distribution Function (CDF)
qbinom(p)     # quantile function - inverse of pnorm
rbinom(n)     # random numbers from normal distribution

# Poisson
dpois()
ppois()
qpois()
rpois()

# Exponential
dexp()
pexp()
qexp()
rexp()

# Chi-squared distribution

dchisq(v, df)     # density of the interval
pchisq()          # distribution of the interval (AUC)
qchisq()          # quantiles
rchisq()          # random deviates
```

## t-Test
```{r, eval = FALSE}
t.test(x, mu = 0, alternative = c("two.sided", "less", "greater"), 
       paired = FALSE, var.equal = FALSE, conf.level = 0.95)
t.test(v, mu)     # one-sample t-test, mu - null hypothesized value
t.test(v1, v2)    # two-sample t-test
t.test(v1, v2, var.equal=T)
t.test(var1, var2, paired=T)
wilcox.test(v1, v2, paired=T)
```

## ANOVA
```{r, eval = FALSE}
# One way ANOVA
oneway.test(x ~ f)
aov(x ~ f)

anova(m1, m2)      # compair two models
```


<!--chapter:end:02_stat_fun_ref.Rmd-->

# Basic Statistics

## Definitions
**population** - all existing samples  
**sample** - subset of statistical population  
**simple random sample** - random subset  
**stratified sample** - fist clustering, than random sample from  
**cluster sample** - random choosing from several existing clusters  
**variables** - discret, continuous, ordinal (—Ä–∞–Ω–≥–æ–≤–∞—è)  

## Probability
A standard French-suited deck of playing cards contains 52 cards; 13 each of hearts (‚ô•), spades (‚ô†), clubs (‚ô¶), and diamonds (‚ô£). Assuming that you have a well-shuffled deck in front of you, the probability of drawing any given card is 1/52 ‚âà 1.92%.  
Calculate the probability of drawing any of the four aces! That is, calculate the probability of drawing üÇ° or üÇ± or üÉÅ or üÉë using the sum rule and assign it to prob_to_draw_ace.  

```{r}
# Calculate the probability of drawing any of the four aces
prob_to_draw_ace <- 1/52 + 1/52 + 1/52 + 1/52
```
Cards and the product rule
Again, assuming that you have a well-shuffled deck in front of you, the probability of drawing any given card is 1/52 ‚âà 1.92% . The probability of drawing any of the four aces is 1/52 + 1/52 + 1/52 + 1/52 = 4/52. Once an ace has been drawn, the probability of picking any of the remaining three is 3/51. If another ace is drawn the probability of picking any of the remaining two is 2/50, and so on.  
Use the product rule to calculate the probability of picking the four aces in a row from the top of a well-shuffled deck and assign it to prob_to_draw_four_aces.  
```{r}
# Calculate the probability of picking four aces in a row
prob_to_draw_four_aces <- 4/52 * 3/51 * 2/50 * 1/49
```

## Analysis of sample distribution
### Histogram
```{r, eval=TRUE}
# sample of random integers
v <- round(rnorm(n=50, sd=5, mean=100))

par(mfrow=c(1,2))
stripchart(v, method = "stack", pch=19, cex=2, offset=.5, at=.15,
           main = "Dotplot of random value", xlab = "Random value")

hist(v)

# add density
x <- density(v)$x
y <- (10/max(density(v)$y))*density(v)$y  # scale y to plot with histogram
lines(x, y, col="red", lwd=2)
```

### Outliers
Outliers are rare values that appear far away from the majority of the data. 
Outliers can bias the results and potentially lead to incorrect conclusions if not handled properly. 
One method for dealing with outliers is to simply remove them. 
However, removing data points can introduce other types of bias into the results, and potentially result in losing critical information. 
If outliers seem to have a lot of influence on the results, a nonparametric test such as the **Wilcoxon Signed Rank Test** may be appropriate to use instead. 
Outliers can be identified visually using a boxplot.  

### Normality
It is possible to use histogram to estimate normality of the distribution.  

```{r }
# QQ-plot - fit normal distibution
qqnorm(v); qqline(v)

var(v)     # variance: sd = sqrt(var)
sd(v)      # standard deviation

sd(v)/sqrt(length(v))  # standard error sd/sqrt(n)

# Z-score (standartization)
# transform distribution to mean=0, variance=1
# z = (x - mean(n))/sd
vs <- scale(v)[,1]
vs

par(mfrow=c(1,2))
hist(v, breaks=10)
hist(vs, breaks=10)
```

## Confidence interval
```{r}
# sample of random integers
x <- round(rnorm(n=50, sd=5, mean=100))

# Confidence interval for normal distribution with p=0.95
m <- mean(x)
s <- sd(x)
n <- length(x)
error <- qnorm(0.95)*s/sqrt(n)
confidence <- c(m-error, m+error)
confidence

# Confidence interval for t-distribution with p=0.95
a <- 5
s <- 2
n <- 20
error <- qt(0.975,df=n-1)*s/sqrt(n)
# confidence interval
c(left=a-error, right=a+error)
```

<!--chapter:end:03_basic_statistics.Rmd-->

# Primary data analysis - Case studies


Abbakumov, 2016, lectures  

```{r, eval=FALSE}

# GET DATA
df <- read.table("./DATA/Swiss_Bank_Notes.csv", header=T, sep=" ", dec=",")
head(df)

# Data explanation: parameters of Swiss Banknotes
# Size of data: 200 (100 are real, 100 are false)
# Length - length
# H_l - height left
# H_r - height right
# dist_l - border left
# dist_up - border up
# Diag - diagonal
# ? Find false banknotes

# 1. Let's add a column with 100 filled 0 and 100 filled with 1.
origin <- 0
df <- data.frame(df, origin)
df$origin[1:100] <-1
# Set origin as factor - binary data (0,1)
df$origin <- as.factor(df$origin)
is.factor(df$origin)
```

```{r, eval=FALSE}
# 2. HISTORGRAM
par(mfrow=c(length(colnames(df))/2,2))
for (i in 1:length(colnames(df))) { hist(df[,i], main = paste(colnames(df)[i])) }

# Histogram for Diagonals
par(mfrow=c(1,1))
hist(df$Diag, breaks=18, probability=TRUE)

# Barplot
barplot(VADeaths, beside=TRUE, legend=TRUE, ylim=c(0, 100),
        ylab="Deaths per 1000", main="Death rates in Virginia")

# Pieplot
groupsize <- c(18, 30, 32, 10, 10)
labels <- c("A", "B", "C", "D", "F")
pie(groupsize, labels, col=c("red", "white", "grey", "black", "blue"))

# All pairs of data scatter plot
plot(df)

# Length ~ Dial
plot(df$Length, df$Diag)
# true notes
points(df$Length[df$origin==1],
       df$Length[df$origin==0], pch=3, col="green")
# false notes
points(df$Length[df$origin==1],
       df$Length[df$origin==0], pch=1, col="red")

# If factors are given, plot makes boxplot
plot(df$origin, df$Diag)
title("Swiss Bank Notes")

# GET DATA - TOWNS
town <- read.table("DATA/town_1959_2.csv", header=T, sep="\t", dec=".")
town
summary(town)
# Median is more stable to outliers
summary(town[,3])
# lets remove 2 first towns from the data
summary(town[3:1004,3])
hist(town[,3])

# log scale allows us to see outliers better
hist(log(town[,3]), breaks=50)

# Truncated mean is better than mean
mean(town[,3], trim=0.05)

# GET DATA - BASKETBALL
bb <- read.table("DATA/basketball.csv", header=F, sep=";", dec=".")
bb
# NBA Player characteristics:
# percent of positives vs:
# SF - light forvard
# PF - heavy forvard
# C - center
# G - defender

summary(bb[,1])

par(mfrow=c(1,1))
plot(bb[,1]~bb[,2])

par(mfrow=c(2,2))
for (i in 1:4) { hist(bb[,1], breaks=5*i, main=paste("Breaks", 5*i), ylab="") }

for (i in unique(bb[,2])) {
     hist(bb[bb[,2]==i ,1],  breaks=6, 
     xlim=c(min(bb[,1])-5, max(bb[,1]+5)), col="white", main=i, ylab="")
}
# Conclusion: for several groups of data boxplots may be more informative than histograms
```

<!--chapter:end:04_primary_analysis_cases.Rmd-->

# Primary data analysis

## Handling missing data

* **Ignore**: Discard samples with missing values.  
* **Impute**: 'Fill in' the missing values with other values.  
* **Accept**: Apply methods that are unaffected by the missing data.  

```{r, eval=FALSE}
library(naniar)
mammographic <- read.csv('./DATA/mammographic.data')
any_na(mammographic)
# Replace ? with NAs: bands
mammographic <- replace_with_na_all(mammographic, ~.x == '?')
any_na(mammographic)
miss_var_summary(mammographic)
```

**Vizualysing missing data**  
```{r, eval = FALSE}
library(ggpubr)

a <- vis_miss(mammographic)
# comulative
b <- vis_miss(mammographic, cluster=TRUE)
c <- gg_miss_case(mammographic)

ggarrange(a, b, c + rremove("x.text"), 
          labels = c("frame view", "cumulative", "missing"),
          ncol = 3, nrow = 1)
```

**Missing data types**  
- **MCAR**: Missing Completely At Random  
- **MAR**: Missing At Random  
- **MNAR**: Missing Not At Random  

| **Type** | **Imputation**    | **Deletion**          | **Visual cues**                                                              |
|----------|-------------------|-----------------------|------------------------------------------------------------------------------|
| **MCAR** | Recommended       | Will not lead to bias | Random or noisy patterns in missingness clusters                             |
| **MAR**  | Recommended       | May lead to bias      | Well-defined missingness clusters when arrangin for a particular variable(s) |
| **MNAR** | Will lead to bias | Will lead to bias     | Neither visual pattern above holds                                           |

It can be difficult to ascertain the missingness type using visual inspection!  

**Internal evaluation**  
Compair distributions with/without imputed values:  

* Mean  
* Variance  
* Scale  

**Exterlan evaluation**
Build ML models with/without imputated values and evaluate impact of imputation method on ML model performance:  

* Classification  
* Regression  
* Clustering  
* etc.  

Ideally imputation should not bring big differences.  

**Mean and linear imputations**  
```{r, eval=FALSE}
library(naniar)
library(simputation)

# Impute with the mean
imp_mean <- bands %>%
    bind_shadow(only_miss = TRUE) %>% 
    add_label_shadow() %>% 
    impute_mean_all()

# Impute with lm
imp_lm <- bands %>%
    bind_shadow(only_miss = TRUE) %>%
    add_label_shadow() %>%
    impute_lm(Blade_pressure ~ Ink_temperature) %>%
    impute_lm(Roughness ~ Ink_temperature) %>%
    impute_lm(Ink_pct ~ Ink_temperature)
```

**Combining multiple imputation models**  
```{r, eval=FALSE}
# Aggregate the imputation models
imp_models <- bind_rows(mean = imp_mean,
                        lm = lmp_lm,
                        .id = "imp_model")
head(imp_models)
```

## Dealing with outliers

* The 3-sigma rule (for normally distributed data)  
* The 1.5*IQR rule (more general)  

Outliers:  

+ any value lower that $Q1 - 1.5 x IQR$  
+ or any higher than $Q3 + 1.5 x IQR$

**Multivariate methods**  
* **Distance-based**: K-nearest neighbors (kNN) distance  
* **Density-based**: Local outlier factor (LOF)  

**1.5*IQR rule**  
Outliers:  

+ any value lower that $Q1 - 1.5 x IQR$  
+ or any higher than $Q3 + 1.5 x IQR$


**Distance-based methods**  

* Average distance to the K-nearest neighbors  

**Density-based methods**  

* Number of the neighboring points within a certain distance  

Assumption: outliers often lie far from their neighbors  

Local Outlier Factor (LOF)  
* Measures the local deviation of a data point with respect to its neighbors.  
* Outliers are observations with substantially lower density than their neighbors.  
`get.knn()` from FNN package  

* Each observation $x$ has an associated score LOF($x$)  
LOF($x$) $\approx$ 1 similar density to its neighbors  
LOF($x$) < 1 higher density than neighbors (inlier)  
LOF($x$) > 1 lower density than neighbors (outlier)   

`lof()` function from `dbscan` package

**What to do with outlier observations?**  
1. **Retention**: Keep them in your dataset and, if possible, use algorithms that are robust to outliers.
- e.g. K nearest-neighbors (kNN), tree-based methods (decision tree, random forest)  
2. **Imputation**: Use an imputation method to replace their value with a less extreme observation.  
- e.g. mode imputation, linear imputation, kNN imputation.  
3. **Capping**: Replace them with the value of the 5-th percentile (lower limit) or 95-th percentile (upper limit).  
4. **Exclusion**: Not recommended, especially in small datasets or those where a normal distribution cannot be assumed.  

```{r}
cars <- read.csv('./DATA/cars.csv')
cars <- cars[,1:3]
head(cars)
boxplot(cars$consume)
consume_quartiles <- quantile(cars$consume)
consume_quartiles

# Scale data and create scatterplot: cars_scaled
require(dplyr)
glimpse(cars)
cars_scaled <- as.data.frame(scale(cars))
plot(distance ~ consume, data = cars_scaled, 
     main = 'Fuel consumption vs. distance')

# Calculate upper threshold: upper_th
upper_th <- consume_quartiles[4] + 
    1.5 * (consume_quartiles[4] - consume_quartiles[2])
upper_th
# Print the sorted vector of distinct potential outliers
sort(unique(cars$consume[cars$consume > upper_th]))

library(FNN)
# Compute KNN score
cars_knn <- get.knn(data = cars_scaled, k = 7)
cars1 <- cars
cars1$knn_score <- rowMeans(cars_knn$nn.dist)

# Print top 5 KNN scores and data point indices: top5_knn
(top5_knn <- order(cars1$knn_score, decreasing = TRUE)[1:5])
print(cars1$knn_score[top5_knn])

# Plot variables using KNN score as size of points
plot(distance ~ consume, data = cars1, cex = knn_score, pch = 20)

library(dbscan)
# Add lof_score column to cars1
cars1 <- cars
cars1$lof_score <- lof(cars_scaled, minPts = 7)

# Print top 5 LOF scores and data point indices: top5_lof
(top5_lof <- order(cars1$lof_score, decreasing = TRUE)[1:5])
print(cars1$lof_score[top5_lof])

# Plot variables using LOF score as size of points
plot(distance ~ consume, data = cars1, 
     cex = lof_score, pch = 20)
```

# Data normalization

Data normalization (feature scaling) is not always needed for e.g. decision-tree-based models.  

Data normalization is beneficial for:  

* Support Vector Machines, K-narest neighbors, Logistic Regression
* Neural networks  
* Clustering algorithms (K-means, K-medoids, DBSCAN, etc.)  
* Feature extraction (Principal Component Analysis, Linear Discriminant Analysis, etc)  

**Min-max scaling**  

* Maps a numerical value $x$ to the [0,1] interval  

$$x' = \frac{x - min}{max - min}$$

* Ensures that all features will share the exact same scale.  
* Does not cope well with outliers.  

**Standardization (Z-score normalization)**  

* Maps a numerical value to $x$ to a new distribution with mean $\mu = 0$ and standard deviation $\sigma = 1$  

$$x' = \frac{x - \mu}{\sigma}$$

* More robust to outliers then min-max normalization.  
* Normalized data may be on different scales.  

**Example**  
```{r}
library(dplyr)

# generate data with different ranges
df <- data.frame(
    a = sample(seq(0, 2, length.out=20)),
    b = sample(seq(100, 500, length.out=20)))
# view data
glimpse(df)

# data ranges to see if normalization is needed
sapply(df, range)


# ranges are different => normalize
# Apply min-max and standardization
nrm <- df %>% 
    mutate(a_MinMax = (a - min(a)) / (max(a) - min(a)), 
           b_MinMax = (b - min(b)) / (max(b) - min(b)),
           a_ZScore = (a - mean(a)) / sd(a),
           b_ZScore = (b - mean(b)) / sd(b)
    )

# ranges for normalized data
sapply(nrm, range)

# plots
par(mfrow=c(1,3))
boxplot(nrm[, 1:2], main = 'Original')
boxplot(nrm[, 3:4], main = 'Min-Max')
boxplot(nrm[, 5:6], main = 'Z-Score')
```

**Sources**  
[Practicing Machine Learning Interview Questions in R on DataCamp](https://learn.datacamp.com/courses/practicing-machine-learning-interview-questions-in-r)  

<!--chapter:end:04_primary_data_analysis.Rmd-->

# Statistical distributions
## Normal Distribution

```{r, eval=TRUE}
x <- seq(-6, 6, length=100)
y1 <- dnorm(x, sd=1)
y2 <- dnorm(x, sd=0.5)
y3 <- dnorm(x, sd=2)
plot(x, y1, 
     xlim=c(-6,6), ylim=c(0,0.8),
     type="l", lwd=2,
     xlab="x value", ylab="Density",
     main="Normal distribution")
lines(x, y2, col="red")
lines(x, y3, col="green")
legend("topright", legend=c("sd = 1", "sd = 0.5", "sd = 2"),
       col=c("black", "red", "green"), 
       lty=c(1,1,1), lwd=2)

# Area under the curve
# For normal distribution auc = 1 (probability for all)
library(DescTools)
AUC(x, y1)
AUC(x,y2)
AUC(x,y3) # to broad distribution, some samples out of the area
```

## Bernoulli Distribution
**Bernoulli distribution** is the discrete probability distribution of a random variable which takes the value 1 with probability p and the value 0 with probability q=1-p [wiki](https://en.wikipedia.org/wiki/Bernoulli_distribution).  
Less formally, it can be thought of as a model for the set of possible outcomes of any single experiment that asks a yes‚Äìno question.  
The Bernoulli distribution is a special case of the binomial distribution with  n=1.  
Bernoulli process, a random process consisting of a sequence of independent Bernoulli trials.  

## Binomial Distribution
It describes the outcome of n independent trials in an experiment. Each trial is assumed to have only two outcomes, either success or failure. If the probability of a successful trial is p, then the probability of having x successful outcomes in an experiment of n independent trials is as follows.
$$f(x) = p^x(1-p)^{(n-x)}$$
where x = 0,1,2,3,4,.... 
The binomial distribution is the basis for the popular binomial test of statistical significance.  

Rules:
1. Must be a fixed number of trials.  
2. Trials must be independent (the outcome of one rial does not affect the others).  
3. Each trial has only two outcomes: success or failure.  
4. The probability of success remains the same in all trials.  

$n$ - number of trials  
$p$ - probability of successful outcome for each trial  
$q$ - probability of failure outcome for each trial  
$x$ - number of successes  
$P(x)$ - probability of the number of successes  

$q = 1 - p$  
$p = 1 - q$  

**Example**  

Suppose there are twelve multiple choice questions in an English class quiz. Each question has five possible answers, and only one of them is correct. Find the probability of having four or less correct answers if a student attempts to answer every question at random.

Solution  
Since only one out of five possible answers is correct, the probability of answering a question correctly by random is 1/5=0.2. We can find the probability of having exactly 4 correct answers by random attempts as follows.  

```{r}
dbinom(4, size=12, prob=0.2) 
```
To find the probability of having four or less correct answers by random attempts, we apply the function dbinom with x = 0,‚Ä¶,4.  
```{r}
dbinom(0, size=12, prob=0.2) + 
dbinom(1, size=12, prob=0.2) + 
dbinom(2, size=12, prob=0.2) + 
dbinom(3, size=12, prob=0.2) + 
dbinom(4, size=12, prob=0.2)
```
Alternatively, we can use the cumulative probability function for binomial distribution pbinom.  
```{r}
pbinom(4, size=12, prob=0.2)
```

**Tasks**  
60% of people who purchase sports cars are men. If 10 sports car owners are randomly selected, find the probability that exactly 7 are men (A: 0.215).  

**Sources**  

[Binomial Distribution](http://www.r-tutor.com/elementary-statistics/probability-distributions/binomial-distribution)

[Statistics Lecture 5.3: A study of Binomial Probability Distributions](https://youtu.be/iGKSxMGX0Do)

Youtube 'Binomial distributions - Probabilities of probabilities' at 3Blue1Brown channel [part 1](https://youtu.be/8idr1WZ1A7Q)[part 2](https://youtu.be/ZA4JkHKZM50)  

## Beta distribution

The **beta distribution** is a family of continuous probability distributions defined on 
the interval [0, 1] parameterized by two positive **shape parameters**, denoted by $\alpha$ and $\beta$. 
[wiki](https://en.wikipedia.org/wiki/Beta_distribution).  

## Geometric Distribution
In probability theory and statistics, the geometric distribution is either one of two discrete probability distributions:

- The probability distribution of the number X of Bernoulli trials needed to get one success, supported on the set { 1, 2, 3, ... }  
- The probability distribution of the number Y = X ‚àí 1 of failures before the first success, supported on the set { 0, 1, 2, 3, ... }  
[wiki](https://en.wikipedia.org/wiki/Geometric_distribution)

$$(1 - p)^{k-1} for k trials where k \in {1,2,3,...}$$
$$(1 - p)^kp for k failures where k \in {0,1,2,...}$$

## Uniform Distributions
**Continuous uniform distribution**  is a family of symmetric probability distributions. The distribution describes an experiment where there is an arbitrary outcome that lies between certain bounds.  
The bounds are defined by the parameters, $a$ and $b$, which are the minimum and maximum values. The interval can either be closed $[a, b]$ or open $(a, b)$.  
Therefore, the distribution is often abbreviated $U(a, b)$, where $U$ stands for uniform distribution [wiki](https://en.wikipedia.org/wiki/Continuous_uniform_distribution).  

Probability density function:  
$$ \begin{cases} \frac{1}{b - a} & \text{for} \, x \in [a,b] \\ 1 & \text{for} \, x < a \, \text{or} \, x\, > b \end{cases}$$

## Poisson Distribution
**Poisson distribution** is a discrete probability distribution that expresses the probability of **a given number of events occurring in a 
fixed interval of time or space** if these events occur with a known constant mean rate and independently of the time since the last event. [wiki](https://en.wikipedia.org/wiki/Poisson_distribution)
Also knows as **Poisson Process**.  

$$P(k) = e^{\lambda}\frac{\lambda^k}{k!}$$

The Poisson distribution is also the limit of a binomial distribution, for which the probability of success for each trial equals Œª divided by the number of trials, as the number of trials approaches infinity.  

According to the Central Limit Theorem for big $\lambda$ the Normal distribution will be the approximation for the Poisson distribution.  

Criteria of the Poisson process:  
1. Events are independent of each other. The occurrence of one event does not affect the probability another event will occur.
2. The average rate (events per time period) is constant.
3. Two events cannot occur at the same time.

Models:  
- Number of calls per minute in a call center.  
- Number of decay events that occur from a radioactive source in a given observation period.  
- The number of meteorites greater than 1 meter diameter that strike Earth in a year.  
- The number of patients arriving in an emergency room between 10 and 11 pm.  
- The number of laser photons hitting a detector in a particular time interval.

Example:  
12 cars cross the bridge in 1 min in average. What is the probability that 17 or more cars will cross the bridge?  
```{r}
ppois(16, lambda=12, lower=FALSE)

# simulate Poisson
x <- rpois(n = 10000, lambda = 3)
hist(x)
# Calculate the probability of break-even
mean(x >= 15)
```

## Exponential Distribution
**Exponential distribution** is the probability distribution of the time between events in a Poisson point process, i.e., 
a process in which events occur continuously and independently at a constant average rate. [wiki](https://en.wikipedia.org/wiki/Exponential_distribution)

$$P(x) = \lambda e ^{ - \lambda x }$$

It is a particular case of the gamma distribution.

Suppose the mean checkout time of a supermarket cashier is 3 minutes. Find the probability of a customer checkout being completed by the cashier in less than 2 minutes.  

```{r}
pexp(2, rate=1/3)
```

## Chi-squared Distribution
In probability theory and statistics, the $\{chi}^2$ distribution with k degrees of freedom is the distribution of a sum of the squares of 
$k$ independent standard normal random variables. [wiki](https://en.wikipedia.org/wiki/Chi-square_distribution)
If $X_1,X_2,‚Ä¶,X_m$ are $m$ independent random variables having the standard normal distribution, then the following quantity follows a **Chi-Squared distribution** with $m$ degrees of freedom. Its mean is $m$, and its variance is $2m$.

$$V = X_1^2 + X_2^2 + ... + X_m^{2} \tilde{} X_{(m)}^2$$

Summ of squares of standard normal, independen random variables are distributed according to the chi-square distribution with k degrees of freedom, where k is the number of random variables being summed.  

Chi-squared distribution applications:  
* Chi-square test of independence in contingency tables  
* Chi-square test of goodness of fit of observed data to hypothetical distributions  
* Likelihood-ratio test for nested models  
* Log-rank test in survival analysis  
* Cochran‚ÄìMantel‚ÄìHaenszel test for stratified contingency tables  
```{r}
x = seq(0,25, length=100)
v = dchisq(x, df=7)
plot(x,v, type='l')
# quantiles of chi-squared
q = c(.25, .5, .75, .999)
qchi = qchisq(q, df = 7, lower.tail = TRUE)
abline(v=qchi, lty=2, col='red')
qchi
```

**Sources**  
[Seven Must-Know Statistical Distributions and Their Simulations for Data Science](https://towardsdatascience.com/seven-must-know-statistical-distributions-and-their-simulations-for-data-science-681c5ac41e32#:~:text=A%20statistical%20distribution%20is%20a,the%20random%20value%20it%20models.)
[Probability distributions](http://www.r-tutor.com/elementary-statistics/probability-distributions)
Hartmann, K., Krois, J., Waske, B. (2018): E-Learning Project SOGA: Statistics and Geospatial Data Analysis. Department of Earth Sciences, Freie Universitaet Berlin.  

<!--chapter:end:05_statistical_distributions.Rmd-->

# Hypothesis testing
## Hypothesis testing theory

Null hypothesis (H<sub>0</sub>):  
1. H<sub>0</sub>: m = Œº  
2. H<sub>0</sub>: m \(\leq\) Œº  
3. H<sub>0</sub>: m \(\geq\) Œº  

Alternative hypotheses (H<sub>a</sub>):  
1. H<sub>a</sub>:m ‚â† Œº (different)  
2. H<sub>a</sub>:m > Œº (greater)  
3. H<sub>a</sub>:m < Œº (less)  

*Note*: Hypothesis 1. are called **two-tailed tests** and hypotheses 2. & 3. are called **one-tailed tests**.  

The p-value is the probability that the observed data could happen, under the condition that the null hypothesis is true.  

*Note*: p-value is not the probability that the null hypothesis is true.  
*Note*: Absence of evidence ‚ßß evidence of absence.  

Cutoffs for hypothesis testing \*p < 0.05, \**p < 0.01, \***p < 0.001.  
If p value is less than significance level alpha (0.05), the hull hypothesies is rejected.  

|                     | not rejected ('negative')      | rejected ('positive')         |
|---------------------|--------------------------------|-------------------------------|
| H<sub>0</sub> true  | True negative (specificity)    | False Positive (Type I error) |
| H<sub>0</sub> false | False Negative (Type II error) | True positive (sensitivity)   |

Type II errors are usually more dangerous.  

Significance level $/alpha$ is the probability of mistakenly rejecting the null hypothesis, if the null hypothesis is true. This is also called false positive and type I error.

## Hypothesis test (Practice)

1. Check distribution (is normal?): shapiro.test()  
2. Distribution uniformity (are both normal?)  
3. Dependence of variable (are x and y dependent?)  
4. Difference of distribution parameters (are means different?)  

```{r, eval=FALSE}
###. 1 Is distribution normal?
town <- read.table("~/DataAnalysis/DATA/town_1959_2.csv", header=T, sep="\t", dec=".")
town
# histogram
hist(town[,3])

# log scale
hist(log(town[,3]), breaks=50)

# ? Is log scaled number of sitizens per town is normally distributed?
# test for normal distribution of dataset
data <- log(town[,3])
shapiro.test(data)

# H0: data is normally distributed
# W = 0.97467 - 
# p-value = 3.15e-12 -> p < alpha (0.01) - H0 is incorrect
# Our distripution is not normal

# Different tests for normality in package "nortest"
install.packages("nortest")
library(nortest)
ad.test(data)   # Anderson-Darling test
lillie.test(data) # Lilliefors (Kolmogorov-Smirnov) test

# Emperical rule: if n < 2000 -> shapiro.test, else -> lillie.test

# In theory it is possible to use method for normally distributed data, 
# even if they are not normal, but almost normal:
# 1. No outlayers
# 2. Symmetry of histogram
# 3. Bell-shaped form of histogram

# Method to make normal from not-normal distribution
# 1. Remove outlayers
# 2. Transform data to get symmetry: log, Boxcox
# 3. Bimodality: split samples into groups

# Boxcox transformation
install.packages("AID")
library(AID)
bctr = boxcoxnc(data)
bctr$results

### 2. Compair centers of two distributions
# Emperical rule: if notmal -> mean, else -> median
# if median -> Mann‚ÄìWhitney-Wilcoxon (MWW) test
# if mean -> one of Student's tests

# Student t-test
# t.test(x,y, alternative="two.sided", paired=FALSE, var.equal=FALSE)

# How to compair if dispersions are the same?
# Fligner-Killeen test
# Brown-Forsythe test
#fligner.test(x~g, data=data.table)

# Example
x <- read.table ("~/DataAnalysis/R_data_analysis/DATA/Albuquerque Home Prices_data.txt", header=T)

names(x)
summary(x)

# Change -9999 to NA
x$AGE[x$AGE==-9999] <- NA
x$TAX[x$TAX==-9999] <- NA

fligner.test(PRICE~COR, data=x)
# Dispersion is the same for both distribution
# p > alpha (0.01)

t.test(x$PRICE[x$COR==1], x$PRICE[x$COR==0],
       alternative="less",
       paired=FALSE,
       var.equal=TRUE)
# p-value = 0.1977 (H0 is true)

# Mood's median test
names(x)
x1 <- x[x$NE==1,1]
x2 <- x[x$NE==0,1]
m <- median(c(x1,x2))
f11 <- sum(x1>m)
f12 <- sum(x2>m)
f21 <- sum(x1<=m)
f22 <- sum(x2<=m)
table <- matrix(c(f11,f12, f21, f22), nrow=2,ncol=2)
chisq.test(table)
# p=0.47 > alpha=0.05: H0 is true
# medians are the same

## Mann‚ÄìWhitney-Wilcoxon (MWW) test
# Check if medians are the same
# U1 = n1*n2+{n1*(n1+1)/2 - T1}
# U1 = n1*n2+{n2*(n2+1)/2 - T2}
# U=min(U1,U2)
# Ti - —Å—É–º–º–∞ —Ä–∞–Ω–≥–æ–≤ –≤ –æ–±—ä–µ–¥–∏–Ω–µ–Ω–Ω–æ–π –≤—ã–±–æ—Ä–∫–µ –Ω–∞–±–ª—é–¥–µ–Ω–∏–π –∏–∑ –≤—ã–±–æ—Ä–∫–∏ i
# n1, n2 - —Ä–∞–∑–º–µ—Ä—ã –≤—ã–±–æ—Ä–æ–∫

wilcox.test(x,y,
            alternative="two.sided",
            paired=FALSE,
            exact=TRUE,
            correct=FALSE)

###. 3. Dependency of variables
# H0: x, y independent
# Correlation analysis
# Correlation (pearson) mesure linear dependency!
# Correlation k depends on outlayers (must be removed)

plot(x$SQFT, x$TAX)

cor(x$SQFT, x$TAX, use = "complete.obs")

cor(x$SQFT, x$TAX, use = "complete.obs", method = "spearman")

cor(x, use = "complete.obs", method = "spearman")

cor(x, use = "complete.obs", method = "pearson")

cor.test(x$SQFT, x$TAX, use = "complete.obs")
```

<!--chapter:end:06_hypothesis_testing.Rmd-->

# t-Procedures
## t-test and normal distribution
t-distribution assumes that the observations are **independent** and that they follow a **normal distribution**.
If the data are **dependent**, then p-values will likely be totally wrong (e.g., for positive correlation, too optimistic). Type II errors?  
It is good to test if observations are normally distributed. Otherwise we assume that data is normally distributed.  
Independence of observations is usually not testable, but can be reasonably assumed if the data collection process was random without replacement.  

FIXME: I do not understand this. Deviation data from normalyty will lead to type-I errors. I data is deviated from normal distribution, use **Wilcoxon test** or **permutation tests**.  

## One-sample t-test
One-sample t-test is used to compare the mean of one sample to a known standard (or theoretical/hypothetical) mean (Œº).  
t-statistics:
\(t = \frac{m - \mu}{s/\sqrt{n}}\), where  
**m** is the sample **mean**  
**n** is the sample **size**  
**s** is the sample **standard deviation** with n‚àí1 **degrees of freedom**  
**Œº** is the **theoretical value**  


Q: And what should I do with this t-statistics?  
Q: What is the difference between t-test and ANOVA?  
Q: What is the smallest sample size which can be tested by t-test?  
Q: Show diagrams explaining why p-value of one-sided is smaller than two-sided tests.

## Practical example: t-test in R  

We want to test if **N** is different from given mean **Œº**=0:  
```{R}
N = c(-0.01, 0.65, -0.17, 1.77, 0.76, -0.16, 0.88, 1.09, 0.96, 0.25)
t.test(N, mu = 0, alternative = "less")
t.test(N, mu = 0, alternative = "two.sided")
t.test(N, mu = 0, alternative = "greater")
```
FIXME: why it accepts all alternatives at the same time (less and greater?)  

## Two samples t-test
Do two different samples have the same mean?  
H<sub>0</sub>:  
1. H<sub>0</sub>: m<sub>1</sub> - m<sub>2</sub> = 0  
2. H<sub>0</sub>: m<sub>1</sub> - m<sub>2</sub> \(\leq\) 0  
3. H<sub>0</sub>: m<sub>1</sub> - m<sub>2</sub> \(\geq\) 0  

H<sub>a</sub>:  
1. H<sub>a</sub>: m<sub>1</sub> - m<sub>2</sub> ‚â† 0 (different)  
2. H<sub>a</sub>: m<sub>1</sub> - m<sub>2</sub> > 0 (greater)  
3. H<sub>a</sub>: m<sub>1</sub> - m<sub>2</sub> < 0 (less)  

The paired sample t-test has four main assumptions:

1. The dependent variable must be **continuous** (interval/ratio).  
2. The observations are **independent** of one another.  
3. The dependent variable should be approximately **normally distributed**.  
4. The dependent variable should not contain any **outliers**.  

Continuous data can take on any value within a range (income, height, weight, etc.). The opposite of continuous data is discrete data, which can only take on a few values (Low, Medium, High, etc.). Occasionally, discrete data can be used to approximate a continuous scale, such as with **Likert-type scales**.  

t-statistics:
\(t=\frac{y - x}{SE}\), where
y and x are the samples means.
SE is the standard error for the difference.
If H<sub>0</sub> is correct, test statistic follows a t-distribution with n+m-2 degrees of freedom (n, m the number of observations in each sample).  

To apply t-test samples must be tested if they have equal variance:  
equal variance (homoscedastic). Type 3 means two samples, unequal variance (heteroscedastic).  

## Compare Student's t and normal distributions

```{r}
x <- seq(-4, 4, length=100)
hx <- dnorm(x)

degf <- c(1, 3, 8, 30)
colors <- c("red", "blue", "darkgreen", "gold", "black")
labels <- c("df=1", "df=3", "df=8", "df=30", "normal")

plot(x, hx, type="l", lty=2, xlab="x value",
     ylab="Density", main="Comparison of t Distributions")

for (i in 1:4){
    lines(x, dt(x,degf[i]), lwd=2, col=colors[i])
}

legend("topright", inset=.05, title="Distributions",
       labels, lwd=2, lty=c(1, 1, 1, 1, 2), col=colors)
```


To generate data with known mean and sd:  
```{r}
rnorm2 <- function(n,mean,sd) { mean+sd*scale(rnorm(n)) }
r <- rnorm2(100,4,1)
```

```{R}
### t-test
a = c(175, 168, 168, 190, 156, 181, 182, 175, 174, 179)
b = c(185, 169, 173, 173, 188, 186, 175, 174, 179, 180)

# test homogeneity of variances using Fisher‚Äôs F-test
var.test(a,b)
# variance is homogene (can use var.equal=T in t.test)

# t-test
t.test(a,b, 
       var.equal=TRUE,   # variance is homogene (tested by var.test(a,b)) 
       paired=FALSE)     # samples are independent
```

## Non-parametric tests
## Mann-Whitney U Rank Sum Test
1. The dependent variable is ordinal or continuous.  
2. The data consist of a randomly selected sample of independent observations from two independent groups.  
3. The dependent variables for the two independent groups share a similar shape.  

## Wilcoxon test
The **Wilcoxon** is a **non-parametric test** which works on normal and non-normal data. However, we usually prefer not to use it if we can assume that the data is normally distributed. The non-parametric test comes with less statistical power, this is a price that one has to pay for more flexible assumptions.  

# Tests for categorical variables
**Categorical variable** can take fixed number of possible values, assigning each individual or other unit of observation to a particular group or nominal category on the basis of some qualitative property.

## Chi-squared tests
The chi-squared test is most suited to large datasets. As a general rule, the chi-squared test is appropriate if at least 80% of the cells have an expected frequency of 5 or greater. In addition, none of the cells should have an expected frequency less than 1. If the expected values are very small, categories may be combined (if it makes sense to do so) to create fewer larger categories. Alternatively, Fisher‚Äôs exact test can be used.  

```{R}
data = rbind(c(83,35), c(92,43))
data
chisq.test(data, correct=F)
```
 chisq.test(testor,correct=F)
## Fisher‚Äôs Exact test
R Example:  

| Group       | TumourShrinkage-No | TumourShrinkage-Yes | Total |
|-------------|--------------------|---------------------|-------|
| 1 Treatment | 8                  | 3                   | 11    |
| 2 Placebo   | 9                  | 4                   | 13    |
| 3 Total     | 17                 | 7                   | 24    |

The **null hypothesis** is that there is **no association** between treatment and tumour shrinkage.  
The **alternative hypothesis** is that there is **some association** between treatment group and tumour shrinkage.  

```{R}
data = rbind(c(8,3), c(9,4))
data
fisher.test(data)
```
The output Fisher‚Äôs exact test tells us that the probability of observing such an extreme combination of frequencies is high, our p-value is 1.000 which is clearly greater than 0.05. In this case, there is **no evidence of an association** between treatment group and tumour shrinkage.  

# Multiple testing
When performing a large number of tests, the type I error is inflated: for Œ±=0.05 and performing n tests, the probability of no false positive result is: 
0.095 x 0.95 x ... (n-times) <<< 0.095  
The larger the number of tests performed, the higher the probability of a false rejection!  
Many data analysis approaches in genomics rely on itemby-item (i.e. multiple) testing:  
Microarray or RNA-Seq expression profiles of ‚Äúnormal‚Äù vs ‚Äúperturbed‚Äù samples: gene-by-gene  
ChIP-chip: locus-by-locus  
RNAi and chemical compound screens  
Genome-wide association studies: marker-by-marker  
QTL analysis: marker-by-marker and trait-by-trait  

**False positive rate** (FPR) - the proportion of false positives among all resulst.  

**False discovery rate** (FDR) - the proportion of false positives among all significant results.  

Example: 20,000 genes, 100 hits, 10 of them wrong.  
FPR: 0.05%  
FDR: 10%  

## The Bonferroni correction
The Bonferroni correction sets the significance cut-off at Œ±/n.  

# Sources
[One-Sample T-test in R](http://www.sthda.com/english/wiki/one-sample-t-test-in-r)

## t-test

The data set shows energy expend in two groups of women: stature
```{r include=TRUE}
library(ISwR)
data(energy)
attach(energy)
head(energy)
tapply(expend, stature, mean)
```

H~0~: there is no difference in averages between lean and obese.  
```{r include=TRUE}
t.test(expend ~ stature)
```
Alternative hypothesis is true - means are different.  
Mean difference is in between -3.5 and 1.0 with a probability 95%.  
The risk of error is 0.15%

### Two-tailed test
Compair two sets of variables.
```{r include=TRUE}
data(intake) # from package ISwR
attach(intake)
head(intake)
mean(post - pre)
```
Is difference of means significant?  
```{r include=TRUE}
t.test(pre, post, paired=TRUE)
```
The difference is significant with a probability 95%.  
The difference is in between 1074.1 and 1566.8 kJ/day  

<!--chapter:end:06_t_tests.Rmd-->

# Analysis of Variance (ANOVA)
## One-way ANOVA
**variance** = SS/df, where SS - sum of squares and df - degree of freedom  
\(SS = \displaystyle\sum_{i=1}^{n}{(x_i - \mu)^2}\), where  
\(\mu\) is the sample **mean**  
**n** is the sample **size** 

\(var(x) = \frac{1}{n}{\displaystyle\sum_{i=1}^{n}{(x_i - \mu)^2}}\)  



SST = SSE + SSC = W + B, where  
SST - Total Sum of Squares  
SSE - Error Sum of Squares - within (W)  
SSC - Sum of Squares Columns (treatmens) - between (B)

C - columns (treatments)  
N - total number of observations  

Mean squared of columns - MSC = SSC/df_columns, where df_columns = C-1  
Mean squared of error - MSE = SSE/df_error, where df_error = N-C  
Sum of squares (total) - SST, where df_total = N-1
F-statistics - F = MSC/MSE

Let's calculate degree of freedom for our example:  
df_columns = 3-1 = 2, MSC = SSC/2  
df_error = 21-3 = 18, MSE = SSE/18  
df_total = 21-1 = 20

```{r}
# 3 groups of students with scores (1-100):
a = c(82,93,61,74,69,70,53)
b = c(71,62,85,94,78,66,71)
c = c(64,73,87,91,56,78,87)

sq = function(x) { sum((x - mean(x))^2) }

sq(a)
sq(b)
sq(c)

```

Using R packages:
```{r}
# data
# Number of calories consumed by month:
may <- c(2166, 1568, 2233, 1882, 2019)
sep <- c(2279, 2075, 2131, 2009, 1793)
dec <- c(2226, 2154, 2583, 2010, 2190)

d <- stack(list(may=may, sep=sep, dec=dec))
d
names(d)

oneway.test(values ~ ind, data=d, var.equal=TRUE)
```

```{r}
# alternative using aov
res <- aov(values ~ ind, data = d)
res
summary(res)
```

## Sources
Example for one-way ANOVA: [youtube](https://www.youtube.com/playlist?list=PLIeGtxpvyG-KA-BLkL391X__r0kU4_hm5) by Brandon Foltz  

<!--chapter:end:07_anova.Rmd-->

# t-test ANOVA difference
The t-test and ANOVA examine whether group means differ from one another. The t-test compares two groups, while ANOVA can do more than two groups.
The t-test ANOVA have three assumptions: independence assumption (the elements of one sample are not related to those of the other sample), normality assumption (samples are randomly drawn from the normally distributed populstions with unknown population means; otherwise the means are no longer best measures of central tendency, thus test will not be valid), and equal variance assumption (the population variances of the two groups are equal)
ANCOVA (analysis of covariance) includes covariates, interval independent variables, in the right-hand side to control their impacts. MANOVA (multivariate analysis of variance) has more than one left-hand side variable.  

![t-test and ANOVA usage.](./_bookdown_files/images/anova.gif)


<!--chapter:end:07_ttest_anova_dif.Rmd-->

# Chi-squared test


## Multinomial Goodness of Fit
A population is called multinomial if its data is categorical and belongs to a collection of discrete non-overlapping classes.

The null hypothesis for goodness of fit test for multinomial distribution is that the observed frequency fi is equal to an expected count $$e_i$$ in each category. 
It is to be rejected if the p-value of the following **Chi-squared test** statistics is less than a given significance level Œ±.

Example
Survey response about the student‚Äôs smoking habit: "Heavy", "Regul" (regularly), "Occas" (occasionally) and "Never". 
The Smoke data is multinomial.

```{r}
library(MASS)
levels(survey$Smoke) 
smoke_freq = table(survey$Smoke) 
smoke_freq

# estimated probabilities
smoke_prob = c(heavy = .045, 
               never = .795, 
               occas = .085, 
               regul = .075)
```

Determine whether the sample data in *smoke_freq* supports estimated probabilities in *smoke_prob* at .05 significance level.

```{r}
chisq.test(smoke_freq, p=smoke_prob)
```

As the p-value 0.991 is greater than the .05 significance level, we do not reject the null hypothesis that the sample data in survey supports the smoking statistics.  

**Sources**

[Multinomial Goodness of Fit](https://www.r-tutor.com/elementary-statistics/goodness-fit/multinomial-goodness-fit)

<!--chapter:end:08_chi_squared_test.Rmd-->

# Non-parametric Methods
A statistical method is called non-parametric if it makes no assumption on the population distribution or sample size.

This is in contrast with most parametric methods in elementary statistics that assume the data is quantitative, the population has a normal distribution and the sample size is sufficiently large.

In general, conclusions drawn from non-parametric methods are not as powerful as the parametric ones. However, as non-parametric methods make fewer assumptions, they are more flexible, more robust, and applicable to non-quantitative data.

## Sign Test
## Wilcoxon Signed-Rank Test
## Mann-Whitney-Wilcoxon Test
## Kruskal-Wallis Test

<!--chapter:end:08_non-parametric_tests.Rmd-->

# Wilcoxon signed-rank test

The Wilcoxon signed-rank test is a non-parametric statistical hypothesis test used to compare two related samples, 
matched samples, or repeated measurements on a single sample to assess whether their population mean ranks differ 
(i.e. it is a paired difference test). It can be used as an alternative to the paired Student's t-test 
(also known as "t-test for matched pairs" or "t-test for dependent samples") when the distribution of the 
difference between two samples' means cannot be assumed to be normally distributed. A Wilcoxon signed-rank test can be 
used to determine whether two dependent samples were selected from populations having the same distribution. 
[wiki](https://en.wikipedia.org/wiki/Wilcoxon_signed-rank_test)  

Two data samples are matched if they come from repeated observations of the same subject. 
Using the Wilcoxon Signed-Rank Test, we can decide whether the corresponding data population distributions are 
identical without assuming them to follow the normal distribution.

Example
In the built-in data set named immer, the barley yield in years 1931 and 1932 of the same field are recorded. 
The yield data are presented in the data frame columns Y1 and Y2.  

```{r}
library(MASS)
head(immer)
```

Problem
Without assuming the data to have normal distribution, test at .05 significance level if the barley yields of 1931 and 1932 in data set immer have identical data distributions.  

Solution
The null hypothesis is that the barley yields of the two sample years are identical populations. 
To test the hypothesis, we apply the wilcox.test function to compare the matched samples. 
For the paired test, we set the "paired" argument as TRUE. As the p-value turns out to be 0.005318, 
and is less than the .05 significance level, we reject the null hypothesis.  

```{r}
wilcox.test(immer$Y1, immer$Y2, paired=TRUE)
```

Answer
At .05 significance level, we conclude that the barley yields of 1931 and 1932 from the data set immer are nonidentical populations.  

**Source**
[Wilcoxon Signed-Rank Test](http://www.r-tutor.com/elementary-statistics/non-parametric-methods/wilcoxon-signed-rank-test)  

<!--chapter:end:09_wilcoxon_signed-rank_test.Rmd-->

# Support Vector Machine

**Definition**: Support vector machine is a representation of the training data as points in space 
separated into categories by a clear gap that is as wide as possible. New examples are then mapped 
into that same space and predicted to belong to a category based on which side of the gap they fall.  
**Advantages**: Effective in high dimensional spaces and uses a subset of training points in the decision function so it is also memory efficient.  
**Disadvantages**: The algorithm does not directly provide probability estimates, these are calculated using an expensive five-fold cross-validation.  

Parameters of SVM:  
* Type of kernel  
* Gamma value  
* C value  

<!--chapter:end:11_svm.Rmd-->

# Correlation

```{r, eval=FALSE}
df <- mtcars
# correlation miles per galon of petrol ~ horse power
cor.test(x = df$mpg, y = df$hp)

# get parameters of analysed correlation
fit <- cor.test(x = df$mpg, y = df$hp)
str(fit)
fit$p.value

# the same correlation in 'formula' form (see. ?cor.test)
?cor.test
cor.test(~mpg+hp, df)

# Plot using generic plot function of ggplot from ggplot2 package
plot(x = df$mpg, y = df$hp)

library(ggplot2)
# add color by number of engine's cylinders
ggplot(df, aes(x = mpg, y = hp, col = factor(cyl)))+geom_point(size=5)

# Subset of necessary data from mtcars
df.sub <- df[,c(1,3:7)]
# Scatterplots for all pairs of variables
pairs(df.sub)

# Correlations of all pairs of variables
cor(df.sub)

# Correlation using corr.test function from 'psych' package
# Correlation for all pairs of variables
library(psych)
fit.sub <- corr.test(df.sub)
fit.sub$r    # correlations
fit.sub$p    # dependecies of variables
```

<!--chapter:end:19_correlation.Rmd-->

# Methods and algorithms of machine learning  
**Regression Analysis**  
* Ordinary Least Squares Regression (OLSR)  
* Linear Regression  
* Logistic Regression  
* Stepwise Regression  
* Polynomial Regression  
* Locally Estimated Scatterplot Smoothing (LOESS)  
  
**Distance-based algorithms**  
* k-Nearest Neighbor (kNN)  
* Learning Vector Quantization (LVQ)  
* Self-organizing Map (SOM)  
  
**Regularization Algorithms**  
* Ridge Regression  
* Least Absolute Shrinkage and Selection Operator (LASSO)  
* Elastic Net  
* Least-Angle Regression (LARS)  
  
**Decision Tree Algorithms**  
* Classification and Regression Tree (CART)  
* Iterative Dichotomiser 3 (ID3)  
* C4.5 and C5.0  
* Chi-squared Automatic Interation Detection (CHAID)  
* Random Forest  
* Conditional Decision Trees  
  
**Bayesian Algorithms**  
* Naive Bayes  
* Gaussian Naive Bayes  
* Multinomial Naive Bayes  
* Bayesian Belief Network (BBN)  
* Bayesian Network (BN)  
  
**Clustering Algorithms**  
* k-Means  
* k-Medians  
* Partitioning Around MEdoids (PAM)  
* Hierarchical Clustering  
  
**Association Rule Mining Algorithms**  
* Apriori algorithm  
* Eclat algorithm  
* FP-growth algorithm  
* Context Based Rule Mining  
  
**Artifical Neural Network Algorithms**  
* Perceptron  
* Back-Propagation  
* Hopfield Network  
* Radial Basis Function Network (RBFN)  
  
**Deep Learining Algorithms**  
* Deep Boltzmann Machine (DBM)  
* Deep Belief Networks (DBN)  
* Convolutional NEural Network (CNN)  
* Stacked Auto-Encoders  
  
**Dimensionality Reduction Algorithms**  
* Principal Component Analysis (PCA)  
* Principal Compnent Regression (PCR)  
* Partial LEast Squares Regression (PLSR)  
* Multidimensional Scaling (MDS)  
* Linear Discriminant Analysis (LDA)  
* Mixtrue Discriminant Analysis (MDA)  
* Quadratic Discriminant Analysis (QDA)
1. PCA (linear)  
2. t-SNE (non-parametric/ nonlinear)  
3. Sammon mapping (nonlinear)  
4. Isomap (nonlinear)  
5. LLE (nonlinear)  
6. CCA (nonlinear)  
7. SNE (nonlinear)  
8. MVU (nonlinear)  
9. Laplacian Eigenmaps (nonlinear)  

  
**Ensemble Algorithms**  
* Boosting  
* Bagging  
* AdaBoost  
* Stacked Generalization (blending)  
* Gradient Boosting Machines (GBM)  
  
**Text Mining**  
* Automatic summarization  
* Named entity recognition (NER)  
* Optical character recognition (OCR)  
* Part-of-speech tagging
* Sentiment analysis
* Speech recognition
* Topic Modeling

<!--chapter:end:20_machine_learning.Rmd-->

# Machine Learning Functions Reference
## Linear Regression
```{r, eval=FALSE}
lm_model <- lm(y ‚àº x1 + x2, data=as.data.frame(cbind(y,x1,x2)))
summary(lm_model)

lm(y ~ x1 + x2 + x3)     # multiple linear regression

lm(log(y) ~ x)           # log transformed
lm(sqrt(y) ~ x)          # sqrt transformed
lm( y ~ log(x))          # fields transformed
llm(log(y) ~ log(x))     # everything is transformed

lm(y ~ .)                # use all fields for regression model

lm(y ~ x + 0)            # forced zero intercept

lm(y ~ x*k)              # interaction of two variables

lm(y ~ x + k + x:k)      # product of xkl but without interaction

lm(y ~ (x + k + ... + l)^2) # all first order interactions

lm(y ~ I(x1 + x2))       # sum of variables
lm(y ~ I(x1^2))          # product of variables (not interation)

lm(y ~ x + I(x^2) + I(x^3)) # polynomial regression
lm(y ~ poly(x,3))           # same as previous

# Forward/backward stepwise regression
# improve model
fit <- lm(y ~ x1 + x2)
bwd.fit <- step(fit, direction = 'backward')
fwd.fit <- step(fit, direction = 'forward', scope( ~ x1 + x2))

# Test linear model
plot(m)            # plot residuals
car::outlier.test(m)
dwtest(m)          # Durbin-Watson Test of the model residuals

# Prediction
predicted_values <- predict(lm_model, newdata=as.data.frame(cbind(x1_test, x2_test)))

# Apriori
dataset <- read.csv("C:\\Datasets\\mushroom.csv", header = TRUE)
mushroom_rules <- apriori(as.matrix(dataset), parameter = list(supp = 0.8, conf = 0.9))
summary(mushroom_rules)
inspect(mushroom_rules)

# Logistic Regression
glm()
glm_mod <-glm(y ‚àº x1+x2, family=binomial(link="logit"), data=as.data.frame(cbind(y,x1,x2))) 

# K-Means Clustering
kmeans_model <- kmeans(x=X, centers=m)

# k-Nearest Neighbor Classification
knn_model <- knn(train=X_train, test=X_test, cl=as.factor(labels), k=K)

# Naƒ±ve Bayes
library(e1071)
nB_model <- naiveBayes(y ‚àº x1 + x2, data=as.data.frame(cbind(y,x1,x2)))

# Decision Trees (CART)
library(rpart)
# Tree-based models

rpart(formula, data, method,control)

formula	outcome~predictor1+predictor2+predictor3+etc.
data    data frame
method 'class' for a classification tree
       'anova' for a regression tree
control optional parameters for controlling tree growth.
        minsplit = 50 - minimal number of observation in a node
        cp=0.001 - cost coplexity factor

cart_model <- rpart(y ‚àº x1 + x2, data=as.data.frame(cbind(y,x1,x2)), method="class")
        
printcp(fit)     # display cp table
plotcp(fit)		 # plot cross-validation results
rsq.rpart(fit)   # plot approximate R-squared and relative error for different splits (2 plots). labels are only appropriate for the "anova" method.
print(fit)       # print results
summary(fit)     # detailed results including surrogate splits
plot(fit)        # plot decision tree
text(fit)        # label the decision tree plot
post(fit, file=) # create postscript plot of decision tree

plot.rpart(cart_model)
text.rpart(cart_model)

# AdaBoost
# boosting functions - uses decision trees as base classifiers
library(rpart)
library(ada)
# Let X be the matrix of features, and labels be a vector of 0-1 class labels.
boost_model <- ada(x=X, y=labels)

# Support Vector Machines (SVM)
library(e1071)
# Let X be the matrix of features, and labels be a vector of 0-1 class labels. Let the regularization parameter be C. Use the following commands to generate the SVM model and view details:
svm_model <- svm(x=X, y=as.factor(labels), kernel ="radial", cost=C)
summary(svm_model)
```

<!--chapter:end:21_ml_fun_ref.Rmd-->

# Split data into train and test subsets
  
Here you can find several simple approaches to split data into train and test subset to fit and to test parameters of your model. 
We want to take 0.8 of our initial data to train our model.  
Data: `datasets::iris`.

1. First approach is to create a **vector containing randomly selected row ids** and to apply this vector to split data.  

```{r}
inTrain = sample(nrow(iris), nrow(iris)*0.8)

# split data
train = iris[inTrain, ]
test = iris[-inTrain, ]
```

2. The same idea to split data as before using `caret` package.  
The advantage is that `createDataPartition` function allows to split data many `times` and  use these subsets to estimate parameters of our model.  
```{r, message=F}
library(caret)
trainIndex <- createDataPartition(iris$Species, p=.8,
                                  list = FALSE,        # if FALSE - create a vector/matrix, if TRUE - create a list
                                  times = 1)           # how many subsets
# split data
train <- iris[trainIndex, ]
test <- iris[-trainIndex, ]
```

3. Another approch is to create a **logical vecotor** containing randomly distributed true/false and apply this vector to subset data.  
```{r}
inTrain = sample(c(TRUE, FALSE), nrow(iris), replace = T, prob = c(0.8,0.2))

# select data
train = iris[inTrain, ]
test = iris[!inTrain, ]
```

4. Using [`caTools`](https://cran.r-project.org/web/packages/caTools/index.html).
```{r}
library(caTools)
inTrain = sample.split(iris, SplitRatio = .8)
train = subset(iris, inTrain == TRUE)
test  = subset(iris, inTrain == FALSE)
```

5. Using [`dplyr`](https://cran.r-project.org/web/packages/dplyr/)
```{r, message=F}
library(dplyr)
iris$id <- 1:nrow(iris)
train <- iris %>% dplyr::sample_frac(.8)
test  <- dplyr::anti_join(iris, train, by = 'id')
```

<!--chapter:end:22_split_data_ways.Rmd-->

# Linear Regression

## Linear regression - theory
Linear regression model is a line y=ax+b, where sum of distances between all y=axi+b and given yi (sum of squares) is minimal.  

Assume that there is approximately a linear relationship between X and Y:  
$$ Y \approx \beta_0 + \beta_1X$$
where \(\beta\)~0~ is an __intercept__ and \(\beta\)~1~ is a __slope__

Parameters of the line could be calculated using __least squares methods__:

$$\beta_1 = \frac{\sum_{i=1}^{n} (x_i - \bar{x})(y_i - \bar{y})}{\sum_{i=1}^{n}{(x_i - \bar{x})^2}} $$
$$\beta_0 = \bar{y} - \beta_1\bar{x} $$

## Generate random data set for the linear model
Suppose we want to simulate from the following linear model:
y = \(\beta\)~0~ + \(\beta\)~1~x + \(\epsilon\),  
where \(\epsilon\) ~ N(0,2^2^). Assume x ~ N(0,1^2^), \(\beta\)~0~ = 0.5, \(\beta\)~1~ = 2.  

```{r}
set.seed(20)
x <-rnorm(100)
e <- rnorm(100, 0, 2)
y <- 0.5 + 2*x + e
summary(y)
plot(x,y)
```

## Practical example
Practical example from [Wikipedia](https://en.wikipedia.org/wiki/Linear_least_squares_(mathematics))  
Set of data: (1,6), (2,5), (3, 7), (4,10)

```{r}
x <- c(1,2,3,4)
y <- c(6,5,7,10)
plot(y~x, xlim=c(0,5), ylim=c(4,10))
abline(3.5, 1.4)
r <- lm(y~x)
segments(x, y, x, r$fitted.values, col="green")
```

We have to find the line corresponding to the minimal sum of errors (distances from the each point to this line):
1. For all points:  
$$\beta_1 + 1\beta_2 = 6$$

$$\beta_1 + 2\beta_2 = 5$$
$$\beta_1 + 3\beta_2 = 7$$
$$\beta_1 + 4\beta_2 = 10$$
the least squares S:  
$$S(\beta_1, \beta_2) = [6 - (\beta_1 + 1\beta_2)]^2 + [5 - (\beta_1 + 2\beta_2)]^2 + [7 - (\beta_1 + 3\beta_2)]^2 + [10 - (\beta_1 + 4\beta_2)]^2 = 4\beta_1^2 + 30\beta_2^2 + 20\beta_1\beta_2 - 56\beta_1 - 154\beta_2 + 210$$
The minimum is:
$$\frac{\partial{S}}{\partial{\beta_1}} = 0 = 8 \beta_1 + 60\beta_2 - 154$$
$$\frac{\partial{S}}{\partial{\beta_2}} = 0 = 20 \beta_1 + 20\beta_2 - 56$$
Result in a system of two equations in two unkowns gives:
$$\beta_1 = 3.5$$
$$\beta_2 = 1.4$$
The line of best fit:  
$y = 3.5 + 1.4x$  
All possible regression lines goes through the intersection point $(\bar{x}, \bar{y})$

## Mean squared error (MSE)

**Standard error of train data**  
$$MSE = \frac{1}{n}\sum_{i=1}^{n}(y_i - \hat{f}(x_i))^2$$

**Standard error of learn data**  
$$MSE = \frac{1}{n_o}\sum_{i=1}^{n_o}(y_i^o - \hat{f}(x_i^o))^2$$

## Linear model in R
```{r}
# Height and weight vectors for 19 children
height <- c(69.1,56.4,65.3,62.8,63,57.3,59.8,62.5,62.5,59.0,51.3,64,56.4,66.5,72.2,65.0,67.0,57.6,66.6)
weight <- c(113,84,99,103,102,83,85,113,84,99,51,90,77,112,150,128,133,85,112)

plot(height,weight)
# Fit linear model
model <- lm(weight ~ height) # weight = slope*weight + intercept
abline(model)   # Regression line

# correlation between variables
cor(height,weight)

# Get data from the model
#get the intercept(b0) and the slope(b1) values
model

# detailed information about the model
summary(model)

# check all attributes calculated by lm
attributes(model)

# getting only the intercept
model$coefficients[1] #or model$coefficients[[1]]

# getting only the slope
model$coefficients[2] #or model$coefficients[[2]]

# checking the residuals
residuals(model)

# predict the weight for a given height, say 60 inches
model$coefficients[[2]]*60 + model$coefficients[[1]]

# Mean squared error (MSE)
predicted.weights <- predict(model, newdata = as.data.frame(weight))
mse <- mean(( weight - predicted.weights)^2, na.rm = TRUE)
mse
```

## Linear regression model for multiple parameters

## Choosing explanatory variables for the model
In general, less MSE then better prediction. We can arrange variables by their importance 
to predict and choose the number of significant variables.  

1. run model with different number of variables.  
2. Arrange MSE for each variable.  
3. Compair MSE for different number of variables using t-test.  
If difference is significant => use more variables.  

In the same way we can compair different models.  

```{r}
library(mosaicData)
head(CPS85)

# relation wage ~ education
# wage - response variable
# educ, exper, age are explanatory variables

# linear model
model1 <- lm(wage ~ educ, data = CPS85)
model2 <- lm(wage ~ educ + age, data = CPS85)
model3 <- lm(wage ~ educ + age + exper, data = CPS85)

pred1 <- predict(model1, newdata = CPS85)
pred2 <- predict(model2, newdata = CPS85)
pred3 <- predict(model3, newdata = CPS85)

# Compair MSE
mse1 <- mean(( CPS85$wage - pred1)^2, na.rm = TRUE)
mse2 <- mean(( CPS85$wage - pred2)^2, na.rm = TRUE)
mse3 <- mean(( CPS85$wage - pred3)^2, na.rm = TRUE)

mse <- data.frame(model_1 = mse1,
                  model_2 = mse2,
                  model_3 = mse3)
mse

# Using both educ and age variables reduese MSE => improve model, where
# adding exper does not improve model
```

## Assessment of model performance for categorical data.
Errors for categorical data can be calculated as number of errors prediction model makes.  
Test whether predicted values match actual values.  
Likelihood: extract the probability that the model assigned to the observed outcome.  

## Confidence intervals for linear model
```{r}
# 0. Build linear model 
data("cars", package = "datasets")
model <- lm(dist ~ speed, data = cars)
# 1. Add predictions 
pred.int <- predict(model, interval = "prediction")
mydata <- cbind(cars, pred.int)
# 2. Regression line + confidence intervals
library("ggplot2")
p <- ggplot(mydata, aes(speed, dist)) +
  geom_point() +
  stat_smooth(method = lm)
# 3. Add prediction intervals
p + geom_line(aes(y = lwr), color = "red", linetype = "dashed")+
    geom_line(aes(y = upr), color = "red", linetype = "dashed")
```

<!--chapter:end:23_linear_regression.Rmd-->

## Practical examples for linear model regression
In this simple example we have 6 persons (3 males and 3 femails) and their score from 0 to 10.  
We want to build a model to see the dependence of score on gender: score ~ gender + $\epsilon$, where $\epsilon$ is an error  

```{r}
# create data frame for the dataset
df = data.frame(gender=c(rep(0,3), rep(1,3)), score=c(10,8,7, 1,3,2))
df

# build linear model
x = lm(score ~ gender, df)
summary(x)
```

<!--chapter:end:24_linear_model_categorical.Rmd-->

# Linear regression complex cases

## Cars 
```{r, eval = FALSE}
df <- mtcars

# Subset of necessary data from mtcars
df.sub <- df[,c(1,3:7)]

# Linear regression model
fit <- lm(mpg ~ hp, df)
fit
summary(fit)

# Plot regression models using ggplot
# parameter geom_smooth builds regression model
library(ggplot2)
# auto model
ggplot(df, aes(hp, mpg))+geom_point(size=2)+geom_smooth()
# for linear model: method = 'lm'
ggplot(df, aes(hp, mpg))+geom_point(size=2)+geom_smooth(method = "lm")

# Split data into two groups by am - Transmission (0 = automatic, 1 = manual)
ggplot(df, aes(hp, mpg, col=factor(am)))+
    geom_point(size=2)+
    geom_smooth(method = "lm")

# Prediction of values using linear regression model
fitted_values_mpg <- data.frame(mpg=df$mpg, fitted=fit$fitted.values)
fitted_values_mpg
View(fitted_values_mpg)

# Lets predict galons of petrol for given horse powers
new_hp <- data.frame(hp=c(100, 150, 129, 300))

predict(fit, new_hp)

new_hp$mpg <- predict(fit, new_hp)
new_hp

# Lets make regression model for cylinders as numeric (not factor)
fit <- lm(mpg ~ cyl, df)
fit

# Regression line for cyclinders
ggplot(df, aes(cyl, mpg))+
    geom_point()+
    geom_smooth(method="lm")+
    theme(axis.text=element_text(size=25),
        axis.title=element_text(size=25, face="bold"))
```

## Linear regression modeling, compair with kNN
```{r, eval=FALSE}
library('GGally')
library('lmtest')
library('FNN')

# –∫–æ–Ω—Å—Ç–∞–Ω—Ç—ã
my.seed <- 12345
train.percent <- 0.85

# –∑–∞–≥—Ä—É–∑–∫–∞ –¥–∞–Ω–Ω—ã—Ö
fileURL <- 'https://sites.google.com/a/kiber-guu.ru/msep/mag-econ/salary_data.csv?attredirects=0&d=1'

# –ø—Ä–µ–æ–±—Ä–∞–∑—É–µ–º –∫–∞—Ç–µ–≥–æ—Ä–∏–∞–ª—å–Ω—ã–µ –ø–µ—Ä–µ–º–µ–Ω–Ω—ã–µ –≤ —Ñ–∞–∫—Ç–æ—Ä—ã
wages.ru <- read.csv(fileURL, row.names = 1, sep = ';', as.is = T)
wages.ru$male <- as.factor(wages.ru$male)
wages.ru$educ <- as.factor(wages.ru$educ)
wages.ru$forlang <- as.factor(wages.ru$forlang)

# –æ–±—É—á–∞—é—â–∞—è –≤—ã–±–æ—Ä–∫–∞
set.seed(my.seed)
inTrain <- sample(seq_along(wages.ru$salary), 
                  nrow(wages.ru) * train.percent)
df.train <- wages.ru[inTrain, c(colnames(wages.ru)[-1], colnames(wages.ru)[1])]
df.test <- wages.ru[-inTrain, -1]

# Variable description
# salary ‚Äì —Å—Ä–µ–¥–Ω–µ–º–µ—Å—è—á–Ω–∞—è –∑–∞—Ä–ø–ª–∞—Ç–∞ –ø–æ—Å–ª–µ –≤—ã—á–µ—Ç–∞ –Ω–∞–ª–æ–≥–æ–≤ –∑–∞ –ø–æ—Å–ª–µ–¥–Ω–∏–µ 12 –º–µ—Å—è—Ü–µ–≤ (—Ä—É–±–ª–µ–π);
# male ‚Äì –ø–æ–ª: 1 ‚Äì –º—É–∂—á–∏–Ω–∞, 0 ‚Äì –∂–µ–Ω—â–∏–Ω–∞;
# educ ‚Äì —É—Ä–æ–≤–µ–Ω—å –æ–±—Ä–∞–∑–æ–≤–∞–Ω–∏—è:
# 1 ‚Äì 0-6 –∫–ª–∞—Å—Å–æ–≤,
# 2 ‚Äì –Ω–µ–∑–∞–∫–æ–Ω—á–µ–Ω–Ω–æ–µ —Å—Ä–µ–¥–Ω–µ–µ (7-8 –∫–ª–∞—Å—Å–æ–≤),
# 3 - –Ω–µ–∑–∞–∫–æ–Ω—á–µ–Ω–Ω–æ–µ —Å—Ä–µ–¥–Ω–µ–µ –ø–ª—é—Å —á—Ç–æ-—Ç–æ –µ—â—ë,
# 4 ‚Äì –∑–∞–∫–æ–Ω—á–µ–Ω–Ω–æ–µ —Å—Ä–µ–¥–Ω–µ–µ,
# 5 ‚Äì –∑–∞–∫–æ–Ω—á–µ–Ω–Ω–æ–µ —Å—Ä–µ–¥–Ω–µ–µ —Å–ø–µ—Ü–∏–∞–ª—å–Ω–æ–µ, 6 ‚Äì –∑–∞–∫–æ–Ω—á–µ–Ω–Ω–æ–µ –≤—ã—Å—à–µ–µ –æ–±—Ä–∞–∑–æ–≤–∞–Ω–∏–µ –∏ –≤—ã—à–µ;
# forlang - –∏–Ω–æ—Å—Ç. —è–∑—ã–∫: 1 ‚Äì –≤–ª–∞–¥–µ–µ—Ç, 0 ‚Äì –Ω–µ—Ç;
# exper ‚Äì –æ—Ñ–∏—Ü–∏–∞–ª—å–Ω—ã–π —Å—Ç–∞–∂ c 1.01.2002 (–ª–µ—Ç).

summary(df.train)
ggp <- ggpairs(df.train)
print(ggp, progress = F)
# —Ü–≤–µ—Ç–∞ –ø–æ —Ñ–∞–∫—Ç–æ—Ä—É male
ggp <- ggpairs(df.train[, c('exper', 'male', 'salary')], 
               mapping = ggplot2::aes(color = male))
print(ggp, progress = F)
# —Ü–≤–µ—Ç–∞ –ø–æ —Ñ–∞–∫—Ç–æ—Ä—É educ
ggp <- ggpairs(df.train[, c('exper', 'educ', 'salary')], 
               mapping = ggplot2::aes(color = educ))
print(ggp, progress = F)

# —Ü–≤–µ—Ç–∞ –ø–æ —Ñ–∞–∫—Ç–æ—Ä—É forlang
ggp <- ggpairs(df.train[, c('exper', 'forlang', 'salary')], 
               mapping = ggplot2::aes(color = forlang))
print(ggp, progress = F)

# Linear regression model
model.1 <- lm(salary ~ . + exper:educ + exper:forlang + exper:male, data = df.train)
summary(model.1)

## Exclude uninfluencial parameters
# Exclude eper:educ as paramaeters are not important
model.2 <- lm(salary ~ . + exper:forlang + exper:male, data = df.train)
summary(model.2)

# Exclude male1:exper
model.3 <- lm(salary ~ . + exper:forlang, data = df.train)
summary(model.3)

# forlang1 is less important, has no sence
model.4 <- lm(salary ~ male + educ + exper, data = df.train)
summary(model.4)

df.train$educ <- as.numeric(df.train$educ)
df.test$educ <- as.numeric(df.test$educ)

model.6 <- lm(salary ~ ., data = df.train)
summary(model.6)

# Model 6 is week, let's add exper:male interactions
df.train$educ <- as.numeric(df.train$educ)

model.7 <- lm(salary ~ . + exper:male, data = df.train)
summary(model.7)

# Obviously the best decision is not to use interactions for modeling
# Test remainers
# —Ç–µ—Å—Ç –ë—Ä–æ–π—à–∞-–ü–∞–≥–∞–Ω–∞
bptest(model.6)
# —Å—Ç–∞—Ç–∏—Å—Ç–∏–∫–∞ –î–∞—Ä–±–∏–Ω–∞-–£–æ—Ç—Å–æ–Ω–∞
dwtest(model.6)
# –≥—Ä–∞—Ñ–∏–∫–∏ –æ—Å—Ç–∞—Ç–∫–æ–≤
par(mar = c(4.5, 4.5, 2, 1))
par(mfrow = c(1, 3))
plot(model.7, 1)
plot(model.7, 4)
plot(model.7, 5)

### Comparison with kNN-method
par(mfrow = c(1, 1))
# —Ñ–∞–∫—Ç–∏—á–µ—Å–∫–∏–µ –∑–Ω–∞—á–µ–Ω–∏—è y –Ω–∞ —Ç–µ—Å—Ç–æ–≤–æ–π –≤—ã–±–æ—Ä–∫–µ
y.fact <- wages.ru[-inTrain, 1]
y.model.lm <- predict(model.6, df.test)
MSE.lm <- sum((y.model.lm - y.fact)^2) / length(y.model.lm)

# kNN —Ç—Ä–µ–±—É–µ—Ç –Ω–∞ –≤—Ö–æ–¥ —Ç–æ–ª—å–∫–æ —á–∏—Å–ª–æ–≤—ã–µ –ø–µ—Ä–µ–º–µ–Ω–Ω—ã–µ
df.train.num <- as.data.frame(apply(df.train, 2, as.numeric))
df.test.num <- as.data.frame(apply(df.test, 2, as.numeric))

for (i in 2:50){
    model.knn <- knn.reg(train = df.train.num[, !(colnames(df.train.num) %in% 'salary')], 
                         y = df.train.num[, 'salary'], 
                         test = df.test.num, k = i)
    y.model.knn <- model.knn$pred
    if (i == 2){
        MSE.knn <- sum((y.model.knn - y.fact)^2) / length(y.model.knn)
    } else {
        MSE.knn <- c(MSE.knn, 
                     sum((y.model.knn - y.fact)^2) / length(y.model.knn))
    }
}

# –≥—Ä–∞—Ñ–∏–∫
par(mar = c(4.5, 4.5, 1, 1))
plot(2:50, MSE.knn, type = 'b', col = 'darkgreen',
     xlab = '–∑–Ω–∞—á–µ–Ω–∏–µ k', ylab = 'MSE –Ω–∞ —Ç–µ—Å—Ç–æ–≤–æ–π –≤—ã–±–æ—Ä–∫–µ')
lines(2:50, rep(MSE.lm, 49), lwd = 2, col = grey(0.2), lty = 2)
legend('bottomright', lty = c(1, 2), pch = c(1, NA), 
       col = c('darkgreen', grey(0.2)), 
       legend = c('k –±–ª–∏–∂–∞–π—à–∏—Ö —Å–æ—Å–µ–¥–∞', '—Ä–µ–≥—Ä–µ—Å—Å–∏—è (–≤—Å–µ —Ñ–∞–∫—Ç–æ—Ä—ã)'), 
       lwd = rep(2, 2))
```
**Source**

[Course 'Math modeling' practical work, State University of Management, Moscow](https://sites.google.com/a/kiber-guu.ru/r-practice/home)

## More complex example
```{r, eval=FALSE}
# Linear regression modeling, compair with kNN

# Source: Course 'Math modeling' practical work, State University of Management, Moscow
# link: https://sites.google.com/a/kiber-guu.ru/r-practice/home

library('GGally')
library('lmtest')
library('FNN')

# variable
my.seed <- 12345
train.percent <- 0.85

wages.ru <- read.csv("~/Projects/data_analysis/DATA/salary.csv", row.names = 1, sep = '\t', 
                     as.is = T)  # transform data into factors when possible
wages.ru$male <- as.factor(wages.ru$male)
wages.ru$educ <- as.factor(wages.ru$educ)
wages.ru$forlang <- as.factor(wages.ru$forlang)

# test data
set.seed(my.seed)
inTrain <- sample(seq_along(wages.ru$salary), 
                  nrow(wages.ru) * train.percent)
df.train <- wages.ru[inTrain, c(colnames(wages.ru)[-1], colnames(wages.ru)[1])]
df.test <- wages.ru[-inTrain, -1]

summary(df.train)
ggp <- ggpairs(df.train)
print(ggp, progress = F)
# —Ü–≤–µ—Ç–∞ –ø–æ —Ñ–∞–∫—Ç–æ—Ä—É male
ggp <- ggpairs(df.train[, c('exper', 'male', 'salary')], 
               mapping = ggplot2::aes(color = male))
print(ggp, progress = F)
# —Ü–≤–µ—Ç–∞ –ø–æ —Ñ–∞–∫—Ç–æ—Ä—É educ
ggp <- ggpairs(df.train[, c('exper', 'educ', 'salary')], 
               mapping = ggplot2::aes(color = educ))
print(ggp, progress = F)

# —Ü–≤–µ—Ç–∞ –ø–æ —Ñ–∞–∫—Ç–æ—Ä—É forlang
ggp <- ggpairs(df.train[, c('exper', 'forlang', 'salary')], 
               mapping = ggplot2::aes(color = forlang))
print(ggp, progress = F)

# Linear regression model
model.1 <- lm(salary ~ . + exper:educ + exper:forlang + exper:male, data = df.train)
summary(model.1)

## Exclude uninfluencial parameters
# Exclude eper:educ as paramaeters are not important
model.2 <- lm(salary ~ . + exper:forlang + exper:male, data = df.train)
summary(model.2)

# Exclude male1:exper
model.3 <- lm(salary ~ . + exper:forlang, data = df.train)
summary(model.3)

# forlang1 is less important, has no sence
model.4 <- lm(salary ~ male + educ + exper, data = df.train)
summary(model.4)

df.train$educ <- as.numeric(df.train$educ)
df.test$educ <- as.numeric(df.test$educ)

model.6 <- lm(salary ~ ., data = df.train)
summary(model.6)

# Model 6 is week, let's add exper:male interactions
df.train$educ <- as.numeric(df.train$educ)

model.7 <- lm(salary ~ . + exper:male, data = df.train)
summary(model.7)

# Obviously the best decision is not to use interactions for modeling
# Test remainers
# —Ç–µ—Å—Ç –ë—Ä–æ–π—à–∞-–ü–∞–≥–∞–Ω–∞
bptest(model.6)
# —Å—Ç–∞—Ç–∏—Å—Ç–∏–∫–∞ –î–∞—Ä–±–∏–Ω–∞-–£–æ—Ç—Å–æ–Ω–∞
dwtest(model.6)
# –≥—Ä–∞—Ñ–∏–∫–∏ –æ—Å—Ç–∞—Ç–∫–æ–≤
par(mar = c(4.5, 4.5, 2, 1))
par(mfrow = c(1, 3))
plot(model.7, 1)
plot(model.7, 4)
plot(model.7, 5)

### Comparison with kNN-method
par(mfrow = c(1, 1))
# —Ñ–∞–∫—Ç–∏—á–µ—Å–∫–∏–µ –∑–Ω–∞—á–µ–Ω–∏—è y –Ω–∞ —Ç–µ—Å—Ç–æ–≤–æ–π –≤—ã–±–æ—Ä–∫–µ
y.fact <- wages.ru[-inTrain, 1]
y.model.lm <- predict(model.6, df.test)
MSE.lm <- sum((y.model.lm - y.fact)^2) / length(y.model.lm)

# kNN —Ç—Ä–µ–±—É–µ—Ç –Ω–∞ –≤—Ö–æ–¥ —Ç–æ–ª—å–∫–æ —á–∏—Å–ª–æ–≤—ã–µ –ø–µ—Ä–µ–º–µ–Ω–Ω—ã–µ
df.train.num <- as.data.frame(apply(df.train, 2, as.numeric))
df.test.num <- as.data.frame(apply(df.test, 2, as.numeric))

for (i in 2:50){
    model.knn <- knn.reg(train = df.train.num[, !(colnames(df.train.num) %in% 'salary')], 
                         y = df.train.num[, 'salary'], 
                         test = df.test.num, k = i)
    y.model.knn <- model.knn$pred
    if (i == 2){
        MSE.knn <- sum((y.model.knn - y.fact)^2) / length(y.model.knn)
    } else {
        MSE.knn <- c(MSE.knn, 
                     sum((y.model.knn - y.fact)^2) / length(y.model.knn))
    }
}

# –≥—Ä–∞—Ñ–∏–∫
par(mar = c(4.5, 4.5, 1, 1))
plot(2:50, MSE.knn, type = 'b', col = 'darkgreen',
     xlab = '–∑–Ω–∞—á–µ–Ω–∏–µ k', ylab = 'MSE –Ω–∞ —Ç–µ—Å—Ç–æ–≤–æ–π –≤—ã–±–æ—Ä–∫–µ')
lines(2:50, rep(MSE.lm, 49), lwd = 2, col = grey(0.2), lty = 2)
legend('bottomright', lty = c(1, 2), pch = c(1, NA), 
       col = c('darkgreen', grey(0.2)), 
       legend = c('k –±–ª–∏–∂–∞–π—à–∏—Ö —Å–æ—Å–µ–¥–∞', '—Ä–µ–≥—Ä–µ—Å—Å–∏—è (–≤—Å–µ —Ñ–∞–∫—Ç–æ—Ä—ã)'), 
       lwd = rep(2, 2))
```

## NEXT part
```{r, eval=FALSE}
# Linear regression (part 1)


# Linear Regression (part 2)
x <- read.table("~/DataAnalysis/R_data_analysis/DATA/Diamond.dat", header=T)

# Check 3 models
# 1. ax+b
# 2. ax^2+b
# 3. ax^2+bx+c
res.1 <- lm(x[,2]~x[,1])
summary(res.1)

ves2 <- x[,1]*x[,1]
res.2 <- lm(x[,2]~ves2)
summary(res.2)

res.3 <- lm(x[,2]~x[,1]+ves2)
summary(res.3)

# weight of diamands ~ weight^2
plot(x[,1]~ves2) 

# Conclusion: from 3 models, the 2d is optimal

### –í—Ä–µ–º–µ–Ω–Ω—ã–µ —Ä—è–¥—ã

ser.g.01 <- read.table("~/DataAnalysis/R_data_analysis/DATA/series_g.csv", header=T, sep=";")

# visual inspection
ser.g.01
dim(ser.g.01)
names(ser.g.01)

# plot vaiable
plot(ser.g.01$series_g, type="l")

log.ser.g <- log(ser.g.01$series_g)

plot(log.ser.g, type="l")
time. <- 1:(144+12)

month.01 <- rep(c(1,0,0,0,0,0,0,0,0,0,0,0), 12+1)
month.02 <- rep(c(0,1,0,0,0,0,0,0,0,0,0,0), 12+1)
month.03 <- rep(c(0,0,1,0,0,0,0,0,0,0,0,0), 12+1)
month.04 <- rep(c(0,0,0,1,0,0,0,0,0,0,0,0), 12+1)
month.05 <- rep(c(0,0,0,0,1,0,0,0,0,0,0,0), 12+1)
month.06 <- rep(c(0,0,0,0,0,1,0,0,0,0,0,0), 12+1)
month.07 <- rep(c(0,0,0,0,0,0,1,0,0,0,0,0), 12+1)
month.08 <- rep(c(0,0,0,0,0,0,0,1,0,0,0,0), 12+1)
month.09 <- rep(c(0,0,0,0,0,0,0,0,1,0,0,0), 12+1)
month.10 <- rep(c(0,0,0,0,0,0,0,0,0,1,0,0), 12+1)
month.11 <- rep(c(0,0,0,0,0,0,0,0,0,0,1,0), 12+1)
month.12 <- rep(c(0,0,0,0,0,0,0,0,0,0,0,1), 12+1)

log.ser.g[145:(144+12)] <-NA

ser.g.02 <- data.frame(log.ser.g, time., month.01, month.02, month.03, 
                       month.04, month.05, month.06, month.07, month.08, month.09, month.10, 
                       month.11, month.12)

ser.g.02

res.01 <- lm(log.ser.g ~ time. +              month.02 + month.03 + month.04 + 
                 month.05 + month.06 + month.07 + month.08 + month.09 + month.10 + 
                 month.11 + month.12, ser.g.02)

summary(res.01)
res.01$fitted.values

plot(ser.g.02$log.ser.g, type="l")
lines(res.01$fitted.values)

plot(ser.g.02$log.ser.g, type="l", col="green")
lines(res.01$fitted.values, col="red")

x.lg = predict.lm(res.01, ser.g.02)

x.lg

plot(x.lg, type="l", col="red")
lines(ser.g.02$log.ser.g, col="green")

y <- exp(x.lg)

plot(y, type="l", col="red")
lines(ser.g.01$series_g, col="green")
```

## NEXT Part
```{r, eval=FALSE}
# –ø—Ä–æ–≤–µ—Å—Ç–∏ –æ—Ç–±–æ—Ä –æ–ø—Ç–∏–º–∞–ª—å–Ω–æ–≥–æ –ø–æ–¥–º–Ω–æ–∂–µ—Å—Ç–≤–∞ –ø–µ—Ä–µ–º–µ–Ω–Ω—ã—Ö;
# –æ—Ç–æ–±—Ä–∞—Ç—å –ø—Ä–µ–¥–∏–∫—Ç–æ—Ä—ã –º–µ—Ç–æ–¥–∞–º–∏ –ø–æ—à–∞–≥–æ–≤–æ–≥–æ –≤–∫–ª—é—á–µ–Ω–∏—è –∏ –∏—Å–∫–ª—é—á–µ–Ω–∏—è;
# –∫–∞–∫ –ø–æ—Å—Ç—Ä–æ–∏—Ç—å —Ä–∏–¥–∂- –∏ –ª–∞—Å—Å–æ-—Ä–µ–≥—Ä–µ—Å—Å–∏—é;
# –∫–∞–∫ –∏—Å–ø–æ–ª—å–∑–æ–≤–∞—Ç—å —Å–Ω–∏–∂–µ–Ω–∏–µ —Ä–∞–∑–º–µ—Ä–Ω–æ—Å—Ç–∏: PCR –∏ PLS;
# –∫–∞–∫ –ø—Ä–∏–º–µ–Ω—è—Ç—å —ç—Ç–∏ –º–µ—Ç–æ–¥—ã –≤ —Å–æ—á–µ—Ç–∞–Ω–∏–∏ —Å –∫—Ä–æ—Å—Å-–≤–∞–ª–∏–¥–∞—Ü–∏–µ–π.

library('ISLR')              # –Ω–∞–±–æ—Ä –¥–∞–Ω–Ω—ã—Ö Hitters
library('leaps')             # —Ñ—É–Ω–∫—Ü–∏—è regsubset() -- –æ—Ç–±–æ—Ä –æ–ø—Ç–∏–º–∞–ª—å–Ω–æ–≥–æ 
#  –ø–æ–¥–º–Ω–æ–∂–µ—Å—Ç–≤–∞ –ø–µ—Ä–µ–º–µ–Ω–Ω—ã—Ö
library('glmnet')            # —Ñ—É–Ω–∫—Ü–∏—è glmnet() -- –ª–∞—Å—Å–æ
library('pls')               # —Ä–µ–≥—Ä–µ—Å—Å–∏—è –Ω–∞ –≥–ª–∞–≤–Ω—ã–µ –∫–æ–º–ø–æ–Ω–µ–Ω—Ç—ã -- pcr()
#  –∏ —á–∞—Å—Ç–Ω—ã–π –ú–ù–ö -- plsr()

my.seed <- 1
?Hitters
fix(Hitters)
names(Hitters)
dim(Hitters)
sum(is.na(Hitters$Salary))
Hitters <- na.omit(Hitters)
dim(Hitters)
sum(is.na(Hitters$Salary))

##  –û—Ç–±–æ—Ä –æ–ø—Ç–∏–º–∞–ª—å–Ω–æ–≥–æ –ø–æ–¥–º–Ω–æ–∂–µ—Å—Ç–≤–∞
# –ø–æ–¥–≥–æ–Ω—è–µ–º –º–æ–¥–µ–ª–∏ —Å —Å–æ—á–µ—Ç–∞–Ω–∏—è–º–∏ –ø—Ä–µ–¥–∏–∫—Ç–æ—Ä–æ–≤ –¥–æ 8 –≤–∫–ª—é—á–∏—Ç–µ–ª—å–Ω–æ
regfit.full <- regsubsets(Salary ~ ., Hitters)
summary(regfit.full)

# –ø–æ–¥–≥–æ–Ω—è–µ–º –º–æ–¥–µ–ª–∏ —Å —Å–æ—á–µ—Ç–∞–Ω–∏—è–º–∏ –ø—Ä–µ–¥–∏–∫—Ç–æ—Ä–æ–≤ –¥–æ 19 (–º–∞–∫—Å–∏–º—É–º –≤ –¥–∞–Ω–Ω—ã—Ö)
regfit.full <- regsubsets(Salary ~ ., Hitters, nvmax = 19)
reg.summary <- summary(regfit.full)
reg.summary

# —Å—Ç—Ä—É–∫—Ç—É—Ä–∞ –æ—Ç—á—ë—Ç–∞ –ø–æ –º–æ–¥–µ–ª–∏ (–∏—â–µ–º —Ö–∞—Ä–∞–∫—Ç–µ—Ä–∏—Å—Ç–∏–∫–∏ –∫–∞—á–µ—Å—Ç–≤–∞)
names(reg.summary)

# R^2 –∏ —Å–∫–æ—Ä—Ä–µ–∫—Ç–∏—Ä–æ–≤–∞–Ω–Ω—ã–π R^2
round(reg.summary$rsq, 3)

# –Ω–∞ –≥—Ä–∞—Ñ–∏–∫–µ
plot(1:19, reg.summary$rsq, type = 'b',
     xlab = '–ö–æ–ª–∏—á–µ—Å—Ç–≤–æ –ø—Ä–µ–¥–∏–∫—Ç–æ—Ä–æ–≤', ylab = 'R-–∫–≤–∞–¥—Ä–∞—Ç')
# —Å–æ–¥–∞ –∂–µ –¥–æ–±–∞–≤–∏–º —Å–∫–æ—Ä—Ä–µ–∫—Ç–∏—Ä–æ–≤–∞–Ω–Ω—ã–π R-–∫–≤–∞–¥—Ä–∞—Ç
points(1:19, reg.summary$adjr2, col = 'red')
# –º–æ–¥–µ–ª—å —Å –º–∞–∫—Å–∏–º–∞–ª—å–Ω—ã–º —Å–∫–æ—Ä—Ä–µ–∫—Ç–∏—Ä–æ–≤–∞–Ω–Ω—ã–º R-–∫–≤–∞–¥—Ä–∞—Ç–æ–º
which.max(reg.summary$adjr2)

### 11
points(which.max(reg.summary$adjr2), 
       reg.summary$adjr2[which.max(reg.summary$adjr2)],
       col = 'red', cex = 2, pch = 20)
legend('bottomright', legend = c('R^2', 'R^2_adg'),
       col = c('black', 'red'), lty = c(1, NA),
       pch = c(1, 1))

# C_p
reg.summary$cp

# —á–∏—Å–ª–æ –ø—Ä–µ–¥–∏–∫—Ç–æ—Ä–æ–≤ —É –æ–ø—Ç–∏–º–∞–ª—å–Ω–æ–≥–æ –∑–Ω–∞—á–µ–Ω–∏—è –∫—Ä–∏—Ç–µ—Ä–∏—è
which.min(reg.summary$cp)

### 10
# –≥—Ä–∞—Ñ–∏–∫
plot(reg.summary$cp, xlab = '–ß–∏—Å–ª–æ –ø—Ä–µ–¥–∏–∫—Ç–æ—Ä–æ–≤',
     ylab = 'C_p', type = 'b')
points(which.min(reg.summary$cp),
       reg.summary$cp[which.min(reg.summary$cp)], 
       col = 'red', cex = 2, pch = 20)

# BIC
reg.summary$bic

# —á–∏—Å–ª–æ –ø—Ä–µ–¥–∏–∫—Ç–æ—Ä–æ–≤ —É –æ–ø—Ç–∏–º–∞–ª—å–Ω–æ–≥–æ –∑–Ω–∞—á–µ–Ω–∏—è –∫—Ä–∏—Ç–µ—Ä–∏—è
which.min(reg.summary$bic)

### 6
# –≥—Ä–∞—Ñ–∏–∫
plot(reg.summary$bic, xlab = '–ß–∏—Å–ª–æ –ø—Ä–µ–¥–∏–∫—Ç–æ—Ä–æ–≤',
     ylab = 'BIC', type = 'b')
points(which.min(reg.summary$bic),
       reg.summary$bic[which.min(reg.summary$bic)], 
       col = 'red', cex = 2, pch = 20)

# –º–µ—Ç–æ–¥ plot –¥–ª—è –≤–∏–∑—É–∞–ª–∏–∑–∞—Ü–∏–∏ —Ä–µ–∑—É–ª—å—Ç–∞—Ç–æ–≤
?plot.regsubsets
plot(regfit.full, scale = 'r2')

plot(regfit.full, scale = 'adjr2')

plot(regfit.full, scale = 'Cp')

plot(regfit.full, scale = 'bic')

# –∫–æ—ç—Ñ—Ñ–∏—Ü–∏–µ–Ω—Ç—ã –º–æ–¥–µ–ª–∏ —Å –Ω–∞–∏–º–µ–Ω—å—à–∏–º BIC
round(coef(regfit.full, 6), 3)

##  –û—Ç–±–æ—Ä –ø—É—Ç—ë–º –ø–æ—à–∞–≥–æ–≤–æ–≥–æ –≤–∫–ª—é—á–µ–Ω–∏—è –∏ –∏—Å–∫–ª—é—á–µ–Ω–∏—è –ø–µ—Ä–µ–º–µ–Ω–Ω—ã—Ö
# –ü–æ—à–∞–≥–æ–≤–æ–µ –≤–∫–ª—é—á–µ–Ω–∏–µ
regfit.fwd <- regsubsets(Salary ~ ., data = Hitters, nvmax = 19, method = 'forward')
summary(regfit.fwd)

regfit.bwd <- regsubsets(Salary ~ ., data = Hitters,
                         nvmax = 19, method = 'backward')
summary(regfit.bwd)

round(coef(regfit.full, 7), 3)

round(coef(regfit.fwd, 7), 3)

round(coef(regfit.bwd, 7), 3)

### –ù–∞—Ö–æ–∂–¥–µ–Ω–∏–µ –æ–ø—Ç–∏–º–∞–ª—å–Ω–æ–π –º–æ–¥–µ–ª–∏ –ø—Ä–∏ –ø–æ–º–æ—â–∏ –º–µ—Ç–æ–¥–æ–≤ –ø—Ä–æ–≤–µ—Ä–æ—á–Ω–æ–π –≤—ã–±–æ—Ä–∫–∏ –∏ –ø–µ—Ä–µ–∫—Ä—ë—Å—Ç–Ω–æ–π –ø—Ä–æ–≤–µ—Ä–∫–∏
set.seed(my.seed)
train <- sample(c(T, F), nrow(Hitters), rep = T)
test <- !train

# –æ–±—É—á–∞–µ–º –º–æ–¥–µ–ª–∏
regfit.best <- regsubsets(Salary ~ ., data = Hitters[train, ],
                          nvmax = 19)
# –º–∞—Ç—Ä–∏—Ü–∞ –æ–±—ä—è—Å–Ω—è—é—â–∏—Ö –ø–µ—Ä–µ–º–µ–Ω–Ω—ã—Ö –º–æ–¥–µ–ª–∏ –¥–ª—è —Ç–µ—Å—Ç–æ–≤–æ–π –≤—ã–±–æ—Ä–∫–∏
test.mat <- model.matrix(Salary ~ ., data = Hitters[test, ])

# –≤–µ–∫—Ç–æ—Ä –æ—à–∏–±–æ–∫
val.errors <- rep(NA, 19)
# —Ü–∏–∫–ª –ø–æ –∫–æ–ª–∏—á–µ—Å—Ç–≤—É –ø—Ä–µ–¥–∏–∫—Ç–æ—Ä–æ–≤
for (i in 1:19){
    coefi <- coef(regfit.best, id = i)
    pred <- test.mat[, names(coefi)] %*% coefi
    # –∑–∞–ø–∏—Å—ã–≤–∞–µ–º –∑–Ω–∞—á–µ–Ω–∏–µ MSE –Ω–∞ —Ç–µ—Å—Ç–æ–≤–æ–π –≤—ã–±–æ—Ä–∫–µ –≤ –≤–µ–∫—Ç–æ—Ä
    val.errors[i] <- mean((Hitters$Salary[test] - pred)^2)
}
round(val.errors, 0)

# –Ω–∞—Ö–æ–¥–∏–º —á–∏—Å–ª–æ –ø—Ä–µ–¥–∏–∫—Ç–æ—Ä–æ–≤ —É –æ–ø—Ç–∏–º–∞–ª—å–Ω–æ–π –º–æ–¥–µ–ª–∏
which.min(val.errors)

### 10
# –∫–æ—ç—Ñ—Ñ–∏—Ü–∏–µ–Ω—Ç—ã –æ–ø—Ç–∏–º–∞–ª—å–Ω–æ–π –º–æ–¥–µ–ª–∏
round(coef(regfit.best, 10), 3)

# —Ñ—É–Ω–∫—Ü–∏—è –¥–ª—è –ø—Ä–æ–≥–Ω–æ–∑–∞ –¥–ª—è —Ñ—É–Ω–∫—Ü–∏–∏ regsubset()
predict.regsubsets <- function(object, newdata, id, ...){
    form <- as.formula(object$call[[2]])
    mat <- model.matrix(form, newdata)
    coefi <- coef(object, id = id)
    xvars <- names(coefi)
    mat[, xvars] %*% coefi
}

# –Ω–∞–±–æ—Ä —Å –æ–ø—Ç–∏–º–∞–ª—å–Ω—ã–º –∫–æ–ª–∏—á–µ—Å—Ç–≤–æ–º –ø–µ—Ä–µ–º–µ–Ω–Ω—ã—Ö –Ω–∞ –ø–æ–ª–Ω–æ–º –Ω–∞–±–æ—Ä–µ –¥–∞–Ω–Ω—ã—Ö
regfit.best <- regsubsets(Salary ~ ., data = Hitters,
                          nvmax = 19)
round(coef(regfit.best, 10), 3)

# k-–∫—Ä–∞—Ç–Ω–∞—è –∫—Ä–æ—Å—Å-–≤–∞–ª–∏–¥–∞—Ü–∏—è
# –æ—Ç–±–∏—Ä–∞–µ–º 10 –±–ª–æ–∫–æ–≤ –Ω–∞–±–ª—é–¥–µ–Ω–∏–π
k <- 10
set.seed(my.seed)
folds <- sample(1:k, nrow(Hitters), replace = T)

# –∑–∞–≥–æ—Ç–æ–≤–∫–∞ –ø–æ–¥ –º–∞—Ç—Ä–∏—Ü—É —Å –æ—à–∏–±–∫–∞–º–∏
cv.errors <- matrix(NA, k, 19, dimnames = list(NULL, paste(1:19)))

# –∑–∞–ø–æ–ª–Ω—è–µ–º –º–∞—Ç—Ä–∏—Ü—É –≤ —Ü–∏–∫–ª–µ –ø–æ –±–ª–æ–∫–∞–º –¥–∞–Ω–Ω—ã—Ö
for (j in 1:k){
    best.fit <- regsubsets(Salary ~ ., data = Hitters[folds != j, ],
                           nvmax = 19)
    # —Ç–µ–ø–µ—Ä—å —Ü–∏–∫–ª –ø–æ –∫–æ–ª–∏—á–µ—Å—Ç–≤—É –æ–±—ä—è—Å–Ω—è—é—â–∏—Ö –ø–µ—Ä–µ–º–µ–Ω–Ω—ã—Ö
    for (i in 1:19){
        # –º–æ–¥–µ–ª—å–Ω—ã–µ –∑–Ω–∞—á–µ–Ω–∏—è Salary
        pred <- predict(best.fit, Hitters[folds == j, ], id = i)
        # –≤–ø–∏—Å—ã–≤–∞–µ–º –æ—à–∏–±–∫—É –≤ –º–∞—Ç—Ä–∏—Ü—É
        cv.errors[j, i] <- mean((Hitters$Salary[folds == j] - pred)^2)
    }
}

# —É—Å—Ä–µ–¥–Ω—è–µ–º –º–∞—Ç—Ä–∏—Ü—É –ø–æ –∫–∞–∂–¥–æ–º—É —Å—Ç–æ–ª–±—Ü—É (—Ç.–µ. –ø–æ –±–ª–æ–∫–∞–º –Ω–∞–±–ª—é–¥–µ–Ω–∏–π), 
#  —á—Ç–æ–±—ã –ø–æ–ª—É—á–∏—Ç—å –æ—Ü–µ–Ω–∫—É MSE –¥–ª—è –∫–∞–∂–¥–æ–π –º–æ–¥–µ–ª–∏ —Å —Ñ–∏–∫—Å–∏—Ä–æ–≤–∞–Ω–Ω—ã–º 
#  –∫–æ–ª–∏—á–µ—Å—Ç–≤–æ–º –æ–±—ä—è—Å–Ω—è—é—â–∏—Ö –ø–µ—Ä–µ–º–µ–Ω–Ω—ã—Ö
mean.cv.errors <- apply(cv.errors, 2, mean)
round(mean.cv.errors, 0)

# –Ω–∞ –≥—Ä–∞—Ñ–∏–∫–µ
plot(mean.cv.errors, type = 'b')
points(which.min(mean.cv.errors), mean.cv.errors[which.min(mean.cv.errors)],
       col = 'red', pch = 20, cex = 2)

# –ø–µ—Ä–µ—Å—Ç—Ä–∞–∏–≤–∞–µ–º –º–æ–¥–µ–ª—å —Å 11 –æ–±—ä—è—Å–Ω—è—é—â–∏–º–∏ –ø–µ—Ä–µ–º–µ–Ω–Ω—ã–º–∏ –Ω–∞ –≤—Å—ë–º –Ω–∞–±–æ—Ä–µ –¥–∞–Ω–Ω—ã—Ö
reg.best <- regsubsets(Salary ~ ., data = Hitters, nvmax = 19)
round(coef(reg.best, 11), 3)

# –∏–∑-–∑–∞ —Å–∏–Ω—Ç–∞–∫—Å–∏—Å–∞ glmnet() —Ñ–æ—Ä–º–∏—Ä—É–µ–º —è–≤–Ω–æ –º–∞—Ç—Ä–∏—Ü—É –æ–±—ä—è—Å–Ω—è—é—â–∏—Ö...
x <- model.matrix(Salary ~ ., Hitters)[, -1]

# –∏ –≤–µ–∫—Ç–æ—Ä –∑–Ω–∞—á–µ–Ω–∏–π –∑–∞–≤–∏—Å–∏–º–æ–π –ø–µ—Ä–µ–º–µ–Ω–Ω–æ–π
y <- Hitters$Salary

### –ì—Ä–µ–±–Ω–µ–≤–∞—è —Ä–µ–≥—Ä–µ—Å—Å–∏—è
# –≤–µ–∫—Ç–æ—Ä –∑–Ω–∞—á–µ–Ω–∏–π –≥–∏–ø–µ—Ä–ø–∞—Ä–∞–º–µ—Ç—Ä–∞ –ª—è–º–±–¥–∞
grid <- 10^seq(10, -2, length = 100)

# –ø–æ–¥–≥–æ–Ω—è–µ–º —Å–µ—Ä–∏—é –º–æ–¥–µ–ª–µ–π —Ä–∏–¥–∂-—Ä–µ–≥—Ä–µ—Å—Å–∏–∏
ridge.mod <- glmnet(x, y, alpha = 0, lambda = grid)

# —Ä–∞–∑–º–µ—Ä–Ω–æ—Å—Ç—å –º–∞—Ç—Ä–∏—Ü—ã –∫–æ—ç—Ñ—Ñ–∏—Ü–∏–µ–Ω—Ç–æ–≤ –º–æ–¥–µ–ª–µ–π
dim(coef(ridge.mod))
## [1]  20 100
# –∑–Ω–∞—á–µ–Ω–∏–µ –ª—è–º–±–¥–∞ –ø–æ–¥ –Ω–æ–º–µ—Ä–æ–º 50
round(ridge.mod$lambda[50], 0)
## [1] 11498
# –∫–æ—ç—Ñ—Ñ–∏—Ü–∏–µ–Ω—Ç—ã —Å–æ–æ—Ç–≤–µ—Ç—Å—Ç–≤—É—é—â–µ–π –º–æ–¥–µ–ª–∏
round(coef(ridge.mod)[, 50], 3)

# –Ω–æ—Ä–º–∞ —ç–ª—å-–¥–≤–∞
round(sqrt(sum(coef(ridge.mod)[-1, 50]^2)), 2)

# –≤—Å—ë —Ç–æ –∂–µ –¥–ª—è –ª—è–º–±–¥—ã –ø–æ–¥ –Ω–æ–º–µ—Ä–æ–º 60
# –∑–Ω–∞—á–µ–Ω–∏–µ –ª—è–º–±–¥–∞ –ø–æ–¥ –Ω–æ–º–µ—Ä–æ–º 50
round(ridge.mod$lambda[60], 0)

# –∫–æ—ç—Ñ—Ñ–∏—Ü–∏–µ–Ω—Ç—ã —Å–æ–æ—Ç–≤–µ—Ç—Å—Ç–≤—É—é—â–µ–π –º–æ–¥–µ–ª–∏
round(coef(ridge.mod)[, 60], 3)

# –Ω–æ—Ä–º–∞ —ç–ª—å-–¥–≤–∞
round(sqrt(sum(coef(ridge.mod)[-1, 60]^2)), 1)

# –º—ã –º–æ–∂–µ–º –ø–æ–ª—É—á–∏—Ç—å –∑–Ω–∞—á–µ–Ω–∏—è –∫–æ—ç—Ñ—Ñ–∏—Ü–∏–µ–Ω—Ç–æ–≤ –¥–ª—è –Ω–æ–≤–æ–π –ª—è–º–±–¥—ã
round(predict(ridge.mod, s = 50, type = 'coefficients')[1:20, ], 3)

## –ú–µ—Ç–æ–¥ –ø—Ä–æ–≤–µ—Ä–æ—á–Ω–æ–π –≤—ã–±–æ—Ä–∫–∏
set.seed(my.seed)
train <- sample(1:nrow(x), nrow(x)/2)
test <- -train
y.test <- y[test]

# –ø–æ–¥–≥–æ–Ω—è–µ–º —Ä–∏–¥–∂-–º–æ–¥–µ–ª–∏ —Å –±–æ–ª—å—à–µ–π —Ç–æ—á–Ω–æ—Å—Ç—å—é (thresh –Ω–∏–∂–µ –∑–Ω–∞—á–µ–Ω–∏—è –ø–æ —É–º–æ–ª—á–∞–Ω–∏—é)
ridge.mod <- glmnet(x[train, ], y[train], alpha = 0, lambda = grid,
                    thresh = 1e-12)
plot(ridge.mod)

# –ø—Ä–æ–≥–Ω–æ–∑—ã –¥–ª—è –º–æ–¥–µ–ª–∏ —Å –ª—è–º–±–¥–∞ = 4
ridge.pred <- predict(ridge.mod, s = 4, newx = x[test, ])
round(mean((ridge.pred - y.test)^2), 0)
# —Å—Ä–∞–≤–Ω–∏–º —Å MSE –¥–ª—è –Ω—É–ª–µ–≤–æ–π –º–æ–¥–µ–ª–∏ (–ø—Ä–æ–≥–Ω–æ–∑ = —Å—Ä–µ–¥–Ω–µ–µ)
round(mean((mean(y[train]) - y.test)^2), 0)
# –Ω–∞—Å–∫–æ–ª—å–∫–æ –º–æ–¥–µ–ª—å —Å –ª—è–º–±–¥–∞ = 4 –æ—Ç–ª–∏—á–∞–µ—Ç—Å—è –æ—Ç –æ–±—ã—á–Ω–æ–π –ü–õ–†
ridge.pred <- predict(ridge.mod, s = 0, newx = x[test, ], exact = T,
                      x = x[train, ], y = y[train])
round(mean((ridge.pred - y.test)^2), 0)
# predict —Å –ª—è–º–±–¥–æ–π (s) = 0 –¥–∞—ë—Ç –º–æ–¥–µ–ª—å –ü–õ–†
lm(y ~ x, subset = train)

round(predict(ridge.mod, s = 0, exact = T, type = 'coefficients',
              x = x[train, ], y = y[train])[1:20, ], 3)

## –ü–æ–¥–±–æ—Ä –æ–ø—Ç–∏–º–∞–ª—å–Ω–æ–≥–æ –∑–Ω–∞—á–µ–Ω–∏—è –ª—è–º–±–¥–∞ —Å –ø–æ–º–æ—â—å—é –ø–µ—Ä–µ–∫—Ä—ë—Å—Ç–Ω–æ–π –ø—Ä–æ–≤–µ—Ä–∫–∏
# k-–∫—Ä–∞—Ç–Ω–∞—è –∫—Ä–æ—Å—Å-–≤–∞–ª–∏–¥–∞—Ü–∏—è
set.seed(my.seed)
# –æ—Ü–µ–Ω–∫–∞ –æ—à–∏–±–∫–∏
cv.out <- cv.glmnet(x[train, ], y[train], alpha = 0)
plot(cv.out)

# –∑–Ω–∞—á–µ–Ω–∏–µ –ª—è–º–±–¥–∞, –æ–±–µ—Å–ø–µ—á–∏–≤–∞—é—â–µ–µ –º–∏–Ω–∏–º–∞–ª—å–Ω—É—é –æ—à–∏–±–∫—É –ø–µ—Ä–µ–∫—Ä—ë—Å—Ç–Ω–æ–π –ø—Ä–æ–≤–µ—Ä–∫–∏
bestlam <- cv.out$lambda.min
round(bestlam, 0)
## [1] 212
# MSE –Ω–∞ —Ç–µ—Å—Ç–æ–≤–æ–π –¥–ª—è —ç—Ç–æ–≥–æ –∑–Ω–∞—á–µ–Ω–∏—è –ª—è–º–±–¥—ã
ridge.pred <- predict(ridge.mod, s = bestlam, newx = x[test, ])
round(mean((ridge.pred - y.test)^2), 0)
## [1] 96016
# –Ω–∞–∫–æ–Ω–µ—Ü, –ø–æ–¥–≥–æ–Ω—è–µ–º –º–æ–¥–µ–ª—å –¥–ª—è –æ–ø—Ç–∏–º–∞–ª—å–Ω–æ–π –ª—è–º–±–¥—ã, 
#  –Ω–∞–π–¥–µ–Ω–Ω–æ–π –ø–æ –ø–µ—Ä–µ–∫—Ä—ë—Å—Ç–Ω–æ–π –ø—Ä–æ–≤–µ—Ä–∫–µ
out <- glmnet(x, y, alpha = 0)
round(predict(out, type = 'coefficients', s = bestlam)[1:20, ], 3)

## –õ–∞—Å—Å–æ
lasso.mod <- glmnet(x[train, ], y[train], alpha = 1, lambda = grid)
plot(lasso.mod)

# –ü–æ–¥–±–æ—Ä –æ–ø—Ç–∏–º–∞–ª—å–Ω–æ–≥–æ –∑–Ω–∞—á–µ–Ω–∏—è –ª—è–º–±–¥–∞ —Å –ø–æ–º–æ—â—å—é –ø–µ—Ä–µ–∫—Ä—ë—Å—Ç–Ω–æ–π –ø—Ä–æ–≤–µ—Ä–∫–∏
set.seed(my.seed)
cv.out <- cv.glmnet(x[train, ], y[train], alpha = 1)
plot(cv.out)

bestlam <- cv.out$lambda.min
lasso.pred <- predict(lasso.mod, s = bestlam, newx = x[test, ])
round(mean((lasso.pred - y.test)^2), 0)

# –∫–æ—ç—Ñ—Ñ–∏—Ü–∏–µ–Ω—Ç—ã –ª—É—á—à–µ–π –º–æ–¥–µ–ª–∏
out <- glmnet(x, y, alpha = 1, lambda = grid)
lasso.coef <- predict(out, type = 'coefficients', s = bestlam)[1:20, ]
round(lasso.coef, 3)
round(lasso.coef[lasso.coef != 0], 3)

### –õ–∞–±–æ—Ä–∞—Ç–æ—Ä–Ω–∞—è —Ä–∞–±–æ—Ç–∞ 3: —Ä–µ–≥—Ä–µ—Å—Å–∏—è –ø—Ä–∏ –ø–æ–º–æ—â–∏ –º–µ—Ç–æ–¥–æ–≤ PCR –∏ PLS
### 6.7.1 –†–µ–≥—Ä–µ—Å—Å–∏—è –Ω–∞ –≥–ª–∞–≤–Ω—ã–µ –∫–æ–º–ø–æ–Ω–µ–Ω—Ç—ã
# –∫—Ä–æ—Å—Å-–≤–∞–ª–∏–¥–∞—Ü–∏—è 
set.seed(2)   # –Ω–µ–ø–æ–Ω—è—Ç–Ω–æ –ø–æ—á–µ–º—É –æ–Ω–∏ —Å–º–µ–Ω–∏–ª–∏ –∑–µ—Ä–Ω–æ; –ø–æ—Ö–æ–∂–µ, –æ–ø–µ—á–∞—Ç–∫–∞
pcr.fit <- pcr(Salary ~ ., data = Hitters, scale = T, validation = 'CV')
summary(pcr.fit)

# –≥—Ä–∞—Ñ–∏–∫ –æ—à–∏–±–æ–∫
validationplot(pcr.fit, val.type = 'MSEP')
# –ü–æ–¥–±–æ—Ä –æ–ø—Ç–∏–∞–ª—å–Ω–æ–≥–æ M: –∫—Ä–æ—Å—Å-–≤–∞–ª–∏–¥–∞—Ü–∏—è –Ω–∞ –æ–±—É—á–∞—é—â–µ–π –≤—ã–±–æ—Ä–∫–µ
set.seed(my.seed)
pcr.fit <- pcr(Salary ~ ., data = Hitters, subset = train, scale = T,
               validation = 'CV')
validationplot(pcr.fit, val.type = 'MSEP')

# MSE –Ω–∞ —Ç–µ—Å—Ç–æ–≤–æ–π –≤—ã–±–æ—Ä–∫–µ
pcr.pred <- predict(pcr.fit, x[test, ], ncomp = 7)
round(mean((pcr.pred - y.test)^2), 0)

# –ø–æ–¥–≥–æ–Ω—è–µ–º –º–æ–¥–µ–ª—å –Ω–∞ –≤—Å–µ–π –≤—ã–±–æ—Ä–∫–µ –¥–ª—è M = 7 
#  (–æ–ø—Ç–∏–º–∞–ª—å–Ω–æ –ø–æ –º–µ—Ç–æ–¥—É –ø–µ—Ä–µ–∫—Ä—ë—Å—Ç–Ω–æ–π –ø—Ä–æ–≤–µ—Ä–∫–∏)
pcr.fit <- pcr(y ~ x, scale = T, ncomp = 7)
summary(pcr.fit)

# –†–µ–≥—Ä–µ—Å—Å–∏—è –ø–æ –º–µ—Ç–æ–¥—É —á–∞—Å—Ç–Ω—ã—Ö –Ω–∞–∏–º–µ–Ω—å—à–∏—Ö –∫–≤–∞–¥—Ä–∞—Ç–æ–≤
set.seed(my.seed)
pls.fit <- plsr(Salary ~ ., data = Hitters, subset = train, scale = T,
                validation = 'CV')
summary(pls.fit)

# —Ç–µ–ø–µ—Ä—å –ø–æ–¥–≥–æ–Ω—è–µ–º –º–æ–¥–µ–ª—å –¥–ª—è –Ω–∞–π–¥–µ–Ω–Ω–æ–≥–æ –æ–ø—Ç–∏–º–∞–ª—å–Ω–æ–≥–æ M = 2 
#  –∏ –æ—Ü–µ–Ω–∏–≤–∞–µ–º MSE –Ω–∞ —Ç–µ—Å—Ç–æ–≤–æ–π
pls.pred <- predict(pls.fit, x[test, ], ncomp = 2)
round(mean((pls.pred - y.test)^2), 0)
## [1] 101417
# –ø–æ–¥–≥–æ–Ω—è–µ–º –º–æ–¥–µ–ª—å –Ω–∞ –≤—Å–µ–π –≤—ã–±–æ—Ä–∫–µ
pls.fit <- plsr(Salary ~ ., data = Hitters, scale = T, ncomp = 2)
summary(pls.fit)

```


<!--chapter:end:24_linear_regression_complex.Rmd-->

# Nonlinear regression

Nonlinear regression is a form of regression analysis in which observational data are modeled by a function which is a nonlinear combination of the model parameters and **depends on one or more independent variables**.  
Some nonlinear data sets can be transformed to a linear model.  
Sone can not be transformed. For such modeling methods of Numerical analysis should be applied such as Newton's method, Gauss-Newton method and Levenberg‚ÄìMarquardt method.  

```{r, eval=FALSE}
–ú–∞—Ç–µ–º–∞—Ç–∏—á–µ—Å–∫–æ–µ –º–æ–¥–µ–ª–∏—Ä–æ–≤–∞–Ω–∏–µ
–ü—Ä–∞–∫—Ç–∏–∫–∞ 7
–ù–µ–ª–∏–Ω–µ–π–Ω—ã–µ –º–æ–¥–µ–ª–∏
–í –ø—Ä–∞–∫—Ç–∏—á–µ—Å–∫–∏—Ö –ø—Ä–∏–º–µ—Ä–∞—Ö –Ω–∏–∂–µ –ø–æ–∫–∞–∑–∞–Ω–æ –∫–∞–∫:
    
    –æ—Ü–µ–Ω–∏–≤–∞—Ç—å –ø–æ–ª–∏–Ω–æ–º–∏–∞–ª—å–Ω—É—é —Ä–µ–≥—Ä–µ—Å—Å–∏—é;
–∞–ø–ø—Ä–æ–∫—Å–∏–º–∏—Ä–æ–≤–∞—Ç—å –Ω–µ–ª–∏–Ω–µ–π–Ω—ã–µ –º–æ–¥–µ–ª–∏ —Å—Ç—É–ø–µ–Ω—á–∞—Ç—ã–º–∏ —Ñ—É–Ω–∫—Ü–∏—è–º–∏;
—Å—Ç—Ä–æ–∏—Ç—å —Å–ø–ª–∞–π–Ω—ã;
—Ä–∞–±–æ—Ç–∞—Ç—å —Å –ª–æ–∫–∞–ª—å–Ω–æ–π —Ä–µ–≥—Ä–µ—Å—Å–∏–µ–π;
—Å—Ç—Ä–æ–∏—Ç—å –æ–±–æ–±—â—ë–Ω–Ω—ã–µ –ª–∏–Ω–µ–π–Ω—ã–µ –º–æ–¥–µ–ª–∏ (GAM).
–ú–æ–¥–µ–ª–∏: –ø–æ–ª–∏–Ω–æ–º–∏–∞–ª—å–Ω–∞—è —Ä–µ–≥—Ä–µ—Å—Å–∏—è, –ø–æ–ª–∏–Ω–æ–º–∏–∞–ª—å–Ω–∞—è –ª–æ–≥–∏—Å—Ç–∏—á–µ—Å–∫–∞—è —Ä–µ–≥—Ä–µ—Å—Å–∏—è, —Å—Ç—É–ø–µ–Ω—á–∞—Ç–∞—è –º–æ–¥–µ–ª—å, –æ–±–æ–±—â—ë–Ω–Ω–∞—è –ª–∏–Ω–µ–π–Ω–∞—è –º–æ–¥–µ–ª—å.
–î–∞–Ω–Ω—ã–µ: Wage {ISLR}

–ü–æ–¥—Ä–æ–±–Ω—ã–µ –∫–æ–º–º–µ–Ω—Ç–∞—Ä–∏–∏ –∫ –∫–æ–¥—É –ª–∞–±–æ—Ä–∞—Ç–æ—Ä–Ω—ã—Ö —Å–º. –≤ [1], –≥–ª–∞–≤–∞ 7.

library('ISLR')              # –Ω–∞–±–æ—Ä –¥–∞–Ω–Ω—ã—Ö Auto
library('splines')           # —Å–ø–ª–∞–π–Ω—ã
library('gam')               # –æ–±–æ–±—â—ë–Ω–Ω—ã–µ –∞–¥–¥–∏—Ç–∏–≤–Ω—ã–µ –º–æ–¥–µ–ª–∏
## Warning: package 'gam' was built under R version 3.3.3
## Loading required package: foreach
## Warning: package 'foreach' was built under R version 3.3.3
## Loaded gam 1.14
library('akima')             # –≥—Ä–∞—Ñ–∏–∫ –¥–≤—É–º–µ—Ä–Ω–æ–π –ø–ª–æ—Å–∫–æ—Å—Ç–∏
## Warning: package 'akima' was built under R version 3.3.3
library('ggplot2')           # –∫—Ä–∞—Å–∏–≤—ã–µ –≥—Ä–∞—Ñ–∏–∫–∏
## Warning: package 'ggplot2' was built under R version 3.3.3
my.seed <- 1
–†–∞–±–æ—Ç–∞–µ–º —Å –Ω–∞–±–æ—Ä–æ–º –¥–∞–Ω–Ω—ã—Ö –ø–æ –∑–∞—Ä–ø–ª–∞—Ç–∞–º 3000 —Ä–∞–±–æ—Ç–Ω–∏–∫–æ–≤-–º—É–∂—á–∏–Ω —Å—Ä–µ–¥–Ω–µ–∞—Ç–ª–∞–Ω—Ç–∏—á–µ—Å–∫–æ–≥–æ —Ä–µ–≥–∏–æ–Ω–∞ Wage. –ü—Ä–∏—Å–æ–µ–¥–∏–Ω—è–µ–º –µ–≥–æ –∫ –ø—Ä–æ—Å—Ç—Ä–∞–Ω—Å—Ç–≤—É –∏–º—ë–Ω —Ñ—É–Ω–∫—Ü–∏–µ–π attach(), –∏ –¥–∞–ª—å—à–µ –æ–±—Ä–∞—â–∞–µ–º—Å—è –Ω–∞–ø—Ä—è–º—É—é –∫ —Å—Ç–æ–ª–±—Ü–∞–º —Ç–∞–±–ª–∏—Ü—ã.

attach(Wage)
–†–∞–±–æ—Ç–∞–µ–º —Å–æ —Å—Ç–æ–ª–±—Ü–∞–º–∏:
    * wage ‚Äì –∑–∞—Ä–∞–±–æ—Ç–Ω–∞—è –ø–ª–∞—Ç–∞ —Ä–∞–±–æ—Ç–Ω–∏–∫–∞ –¥–æ —É–ø–ª–∞—Ç—ã –Ω–∞–ª–æ–≥–æ–≤;
* age ‚Äì –≤–æ–∑—Ä–∞—Å—Ç —Ä–∞–±–æ—Ç–Ω–∏–∫–∞ –≤ –≥–æ–¥–∞—Ö.

–ü–æ–ª–∏–Ω–æ–º–∏–∞–ª—å–Ω–∞—è —Ä–µ–≥—Ä–µ—Å—Å–∏—è
–ó–∞–≤–∏—Å–∏–º–æ—Å—Ç—å –∑–∞—Ä–ø–ª–∞—Ç—ã –æ—Ç –≤–æ–∑—Ä–∞—Å—Ç–∞
–°—É–¥—è –ø–æ –≥—Ä–∞—Ñ–∏–∫—É –Ω–∏–∂–µ, –∑–∑–∞–∏–º–æ—Å–≤—è–∑—å –∑–∞—Ä–∞–±–æ—Ç–Ω–æ–π –ø–ª–∞—Ç—ã –∏ –≤–æ–∑—Ä–∞—Å—Ç–∞ –Ω–µ–ª–∏–Ω–µ–π–Ω–∞. –ù–∞–±–ª—é–¥–∞–µ—Ç—Å—è —Ç–∞–∫–∂–µ –≥—Ä—É–ø–ø–∞ –Ω–∞–±–ª—é–¥–µ–Ω–∏–π —Å –≤—ã—Å–æ–∫–∏–º –∑–Ω–∞—á–µ–Ω–∏–µ–º wage, –≥—Ä–∞–Ω–∏—Ü–∞ –ø—Ä–æ—Ö–æ–¥–∏—Ç –ø—Ä–∏–º–µ—Ä–Ω–æ –Ω–∞ —É—Ä–æ–≤–Ω–µ 250.

gp <- ggplot(data = Wage, aes(x = age, y = wage))
gp <- gp + geom_point() + geom_abline(slope = 0, intercept = 250, col = 'red')
gp


–ü–æ–¥–≥–æ–Ω—è–µ–º –ø–æ–ª–∏–Ω–æ–º —á–µ—Ç–≤—ë—Ä—Ç–æ–π —Å—Ç–µ–ø–µ–Ω–∏ –¥–ª—è –∑–∞–≤–∏—Å–∏–º–æ—Å—Ç–∏ –∑–∞—Ä–∞–±–æ—Ç–Ω–æ–π –ø–ª–∞—Ç—ã –æ—Ç –≤–æ–∑—Ä–∞—Å—Ç–∞.

fit <- lm(wage ~ poly(age, 4), data = Wage)
round(coef(summary(fit)), 2)
##               Estimate Std. Error t value Pr(>|t|)
## (Intercept)     111.70       0.73  153.28     0.00
## poly(age, 4)1   447.07      39.91   11.20     0.00
## poly(age, 4)2  -478.32      39.91  -11.98     0.00
## poly(age, 4)3   125.52      39.91    3.14     0.00
## poly(age, 4)4   -77.91      39.91   -1.95     0.05
–§—É–Ω–∫—Ü–∏—è poly(age, 4) —Å–æ–∑–¥–∞—ë—Ç —Ç–∞–±–ª–∏—Ü—É —Å –±–∞–∑–∏—Å–æ–º –æ—Ä—Ç–æ–≥–æ–Ω–∞–ª—å–Ω—ã—Ö –ø–æ–ª–∏–Ω–æ–º–æ–≤: –ª–∏–Ω–µ–π–Ω—ã–µ –∫–æ–º–±–∏–Ω–∞—Ü–∏–∏ –∑–Ω–∞—á–µ–Ω–∏–π –ø–µ—Ä–µ–º–µ–Ω–Ω–æ–π age –≤ —Å—Ç–µ–ø–µ–Ω—è—Ö –æ—Ç 1 –¥–æ 4.

round(head(poly(age, 4)), 3)
##           1      2      3      4
## [1,] -0.039  0.056 -0.072  0.087
## [2,] -0.029  0.026 -0.015 -0.003
## [3,]  0.004 -0.015  0.000  0.014
## [4,]  0.001 -0.015  0.005  0.013
## [5,]  0.012 -0.010 -0.011  0.010
## [6,]  0.018 -0.002 -0.017 -0.001
# –º–æ–∂–Ω–æ –ø–æ–ª—É—á–∏—Ç—å —Å–∞–º–∏ –∑–Ω–∞—á–µ–Ω–∏—è age –≤ –∑–∞–¥–∞–Ω–Ω—ã—Ö —Å—Ç–µ–ø–µ–Ω—è—Ö
round(head(poly(age, 4, raw = T)), 3)
##       1    2      3       4
## [1,] 18  324   5832  104976
## [2,] 24  576  13824  331776
## [3,] 45 2025  91125 4100625
## [4,] 43 1849  79507 3418801
## [5,] 50 2500 125000 6250000
## [6,] 54 2916 157464 8503056
# –Ω–∞ –ø—Ä–æ–≥–Ω–æ–∑ –Ω–µ –ø–æ–≤–ª–∏—è–µ—Ç, –Ω–æ –æ—Ü–µ–Ω–∫–∏ –ø–∞—Ä–∞–º–µ—Ç—Ä–æ–≤ –∏–∑–º–µ–Ω—è—é—Ç—Å—è
fit.2 <- lm(wage ~ poly(age, 4, raw = T), data = Wage)
round(coef(summary(fit.2)), 2)
##                        Estimate Std. Error t value Pr(>|t|)
## (Intercept)             -184.15      60.04   -3.07     0.00
## poly(age, 4, raw = T)1    21.25       5.89    3.61     0.00
## poly(age, 4, raw = T)2    -0.56       0.21   -2.74     0.01
## poly(age, 4, raw = T)3     0.01       0.00    2.22     0.03
## poly(age, 4, raw = T)4     0.00       0.00   -1.95     0.05
# –≥—Ä–∞–Ω–∏—Ü—ã –∏–∑–º–µ–Ω–µ–Ω–∏—è –ø–µ—Ä–µ–º–µ–Ω–Ω–æ–π age
agelims <- range(age)

# –∑–Ω–∞—á–µ–Ω–∏—è age, –¥–ª—è –∫–æ—Ç–æ—Ä—ã—Ö –¥–µ–ª–∞–µ–º –ø—Ä–æ–≥–Ω–æ–∑ (–æ—Ç min –¥–æ max —Å —à–∞–≥–æ–º 1)
age.grid <- seq(from = agelims[1], to = agelims[2])

# —Ä–∞—Å—Å—á–∏—Ç–∞—Ç—å –ø—Ä–æ–≥–Ω–æ–∑—ã –∏ –∏—Ö —Å—Ç–∞–Ω–¥–∞—Ä—Ç–Ω—ã–µ –æ—à–∏–±–∫–∏
preds <- predict(fit, newdata = list(age = age.grid), se = T)

# –≥—Ä–∞–Ω–∏—Ü—ã –¥–æ–≤–µ—Ä–∏—Ç–µ–ª—å–Ω–æ–≥–æ –∏–Ω—Ç–µ—Ä–≤–∞–ª–∞ –¥–ª—è –∑–∞—Ä–∞–±–æ—Ç–Ω–æ–π –ø–ª–∞—Ç—ã
se.bands <- cbind(lower.bound = preds$fit - 2*preds$se.fit,
                  upper.bound = preds$fit + 2*preds$se.fit)

# —Å–º–æ—Ç—Ä–∏–º —Ä–µ–∑—É–ª—å—Ç–∞—Ç
round(head(se.bands), 2)
##   lower.bound upper.bound
## 1       41.33       62.53
## 2       49.76       67.24
## 3       57.39       71.76
## 4       64.27       76.09
## 5       70.44       80.27
## 6       75.94       84.28
–†–∏—Å—É–µ–º –ª–µ–≤—É—é –ø–∞–Ω–µ–ª—å –≥—Ä–∞—Ñ–∏–∫–∞ —Å–æ —Å–ª–∞–π–¥–∞ 4 –ø—Ä–µ–∑–µ–Ω—Ç–∞—Ü–∏–∏ (—Ä–∏—Å. 7.1 –∫–Ω–∏–≥–∏). –§—É–Ω–∫—Ü–∏—è matlines() —Ä–∏—Å—É–µ—Ç –≥—Ä—Ñ–∏–∫ —Å—Ç–æ–ª–±—Ü–æ–≤ –æ–¥–Ω–æ–π –º–∞—Ç—Ä–∏—Ü—ã –ø—Ä–æ—Ç–∏–≤ —Å—Ç–æ–ª–±—Ü–æ–≤ –¥—Ä—É–≥–æ–π.

# –Ω–∞–±–ª—é–¥–µ–Ω–∏—è
plot(age, wage, xlim = agelims, cex = 0.5, col = 'darkgrey')

# –∑–∞–≥–æ–ª–æ–≤–æ–∫
title('–ü–æ–ª–∏–Ω–æ–º —á–µ—Ç–≤—ë—Ä—Ç–æ–π —Å—Ç–µ–ø–µ–Ω–∏')

# –º–æ–¥–µ–ª—å
lines(age.grid, preds$fit, lwd = 2, col = 'blue')

# –¥–æ–≤–µ—Ä–∏—Ç–µ–ª—å–Ω—ã–µ –∏–Ω—Ç–µ—Ä–≤–∞–ª—ã –ø—Ä–æ–≥–Ω–æ–∑–∞
matlines(x = age.grid, y = se.bands, lwd = 1, col = 'blue', lty = 3)


–£–±–µ–¥–∏–º—Å—è, —á—Ç–æ –ø—Ä–æ–≥–Ω–æ–∑—ã –ø–æ –º–æ–¥–µ–ª—è–º —Å —Ä–∞–∑–ª–∏—á–Ω—ã–º–∏ –≤—ã–∑–æ–≤–∞–º–∏ poly() —Å–æ–≤–ø–∞–¥–∞—é—Ç.

# –ø—Ä–æ–≥–Ω–æ–∑—ã –ø–æ –≤—Ç–æ—Ä–æ–º—É –≤—ã–∑–æ–≤—É –º–æ–¥–µ–ª–∏
preds2 <- predict(fit.2, newdata = list(age = age.grid), se = T)

# –º–∞–∫—Å–∏–º–∞–ª—å–Ω–æ–µ —Ä–∞—Å—Ö–æ–∂–¥–µ–Ω–∏–µ –º–µ–∂–¥—É –ø—Ä–æ–≥–Ω–æ–∑–∞–º–∏ –ø–æ –¥–≤—É–º –≤–∞—Ä–∏–∞–Ω—Ç–∞–º –≤—ã–∑–æ–≤–∞ –º–æ–¥–µ–ª–∏
max(abs(preds$fit - preds2$fit))
## [1] 7.389644e-13
–¢–µ–ø–µ—Ä—å –ø–æ–¥–±–∏—Ä–∞–µ–º —Å—Ç–µ–ø–µ–Ω—å –ø–æ–ª–∏–Ω–æ–º–∞, —Å—Ä–∞–≤–Ω–∏–≤–∞—è –º–æ–¥–µ–ª–∏ —Å–æ —Å—Ç–µ–ø–µ–Ω—è–º–∏ –æ—Ç 1 –¥–æ 5 —Å –ø–æ–º–æ—â—å—é –¥–∏—Å–ø–µ—Ä—Å–∏–æ–Ω–Ω–æ–≥–æ –∞–Ω–∞–ª–∏–∑–∞ (ANOVA).

fit.1 <- lm(wage ~ age, data = Wage)
fit.2 <- lm(wage ~ poly(age, 2), data = Wage)
fit.3 <- lm(wage ~ poly(age, 3), data = Wage)
fit.4 <- lm(wage ~ poly(age, 4), data = Wage)
fit.5 <- lm(wage ~ poly(age, 5), data = Wage)

round(anova(fit.1, fit.2, fit.3, fit.4, fit.5), 2)
Res.Df
<dbl>
    RSS
<dbl>
    Df
<dbl>
    Sum of Sq
<dbl>
    F
<dbl>
    Pr(>F)
<dbl>
    2998	5022216	NA	NA	NA	NA
2997	4793430	1	228786.01	143.59	0.00
2996	4777674	1	15755.69	9.89	0.00
2995	4771604	1	6070.15	3.81	0.05
2994	4770322	1	1282.56	0.80	0.37
5 rows
–†–∞—Å—Å–º–∞—Ç—Ä–∏–≤–∞—é—Ç—Å—è –ø—è—Ç—å –º–æ–¥–µ–ª–µ–π, –≤ –∫–æ—Ç–æ—Ä—ã—Ö —Å—Ç–µ–ø–µ–Ω–∏ –ø–æ–ª–∏–Ω–æ–º–∞ –æ—Ç age –∏–¥—É—Ç –ø–æ –≤–æ–∑—Ä–∞—Å—Ç–∞–Ω–∏—é. –í –∫—Ä–∞–π–Ω–µ–º –ø—Ä–∞–≤–æ–º —Å—Ç–æ–ª–±—Ü–µ —Ç–∞–±–ª–∏—Ü–µ –ø—Ä–∏–≤–æ–¥—è—Ç—Å—è p-–∑–Ω–∞—á–µ–Ω–∏—è –¥–ª—è –ø—Ä–æ–≤–µ—Ä–∫–∏ –Ω—É–ª–µ–≤–æ–π –≥–∏–ø–æ—Ç–µ–∑—ã: —Ç–µ–∫—É—â–∞—è –º–æ–¥–µ–ª—å –Ω–µ –¥–∞—ë—Ç —Å—Ç–∞—Ç–∏—Å—Ç–∏—á–µ—Å–∫–∏ –∑–Ω–∞—á–∏–º–æ–≥–æ —Å–æ–∫—Ä–∞—â–µ–Ω–∏—è RSS –ø–æ —Å—Ä–∞–≤–Ω–µ–Ω–∏—é —Å –ø—Ä–µ–¥—ã–¥—É—â–µ–π –º–æ–¥–µ–ª—å—é. –ú–æ–∂–Ω–æ —Å–¥–µ–ª–∞—Ç—å –≤—ã–≤–æ–¥, —á—Ç–æ —Å—Ç–µ–ø–µ–Ω–∏ 3 –¥–æ—Å—Ç–∞—Ç–æ—á–Ω–æ, –¥–∞–ª—å–Ω–µ–π—à–µ–µ —É–≤–µ–ª–∏—á–µ–Ω–∏–µ —Å—Ç–µ–ø–µ–Ω–∏ –Ω–µ –¥–∞—ë—Ç –∑–Ω–∞—á–∏–º–æ–≥–æ —É–ª—É—á—à–µ–Ω–∏—è –∫–∞—á–µ—Å—Ç–≤–∞ –º–æ–¥–µ–ª–∏.

–ó–∞–≤–∏—Å–∏–º–æ—Å—Ç—å –≤–µ—Ä–æ—è—Ç–Ω–æ—Å—Ç–∏ –ø–æ–ª—É—á–∞—Ç—å –∑–∞—Ä–ø–ª–∞—Ç—É > 250 –æ—Ç –≤–æ–∑—Ä–∞—Å—Ç–∞
–¢–µ–ø–µ—Ä—å –≤–µ—Ä–Ω—ë–º—Å—è –∫ –≥—Ä—É–ø–ø–µ –Ω–∞–±–ª—é–¥–µ–Ω–∏–π —Å –≤—ã—Å–æ–∫–∏–º wage. –†–∞—Å—Å–º–æ—Ç—Ä–∏–º –∑–∞–≤–∏—Å–∏–º–æ—Å—Ç—å –≤–µ—Ä–æ—è—Ç–Ω–æ—Å—Ç–∏ —Ç–æ–≥–æ, —á—Ç–æ –≤–µ–ª–∏—á–∏–Ω–∞ –∑–∞—Ä–ø–ª–∞—Ç—ã –±–æ–ª—å—à–µ 250, –æ—Ç –≤–æ–∑—Ä–∞—Å—Ç–∞.
–ü–æ–¥–≥–æ–Ω—è–µ–º –ª–æ–≥–∏—Å—Ç–∏—á–µ—Å–∫—É—é —Ä–µ–≥—Ä–µ—Å—Å–∏—é –∏ –¥–µ–ª–∞–µ–º –ø—Ä–æ–≥–Ω–æ–∑—ã, –¥–ª—è —ç—Ç–æ–≥–æ –∏—Å–ø–æ–ª—å–∑—É–µ–º —Ñ—É–Ω–∫—Ü–∏—é –¥–ª—è –æ—Ü–µ–Ω–∫–∏ –æ–±–æ–±—â—ë–Ω–Ω–æ–π –ª–∏–Ω–µ–π–Ω–æ–π –º–æ–¥–µ–ª–∏  glm() –∏ —É–∫–∞–∑—ã–≤–∞–µ–º —Ç–∏–ø –º–æ–¥–µ–ª–∏ binomial:
    
    fit <- glm(I(wage > 250) ~ poly(age, 4), data = Wage, family = 'binomial')

# –ø—Ä–æ–≥–Ω–æ–∑—ã
preds <- predict(fit, newdata = list(age = age.grid), se = T)

# –ø–µ—Ä–µ—Å—á–∏—Ç—ã–≤–∞–µ–º –¥–æ–≤–µ—Ä–∏—Ç–µ–ª—å–Ω—ã–µ –∏–Ω—Ç–µ—Ä–≤–∞–ª—ã –∏ –ø—Ä–æ–≥–Ω–æ–∑—ã –≤ –∏—Å—Ö–æ–¥–Ω—ã–µ –ï–ò
pfit <- exp(preds$fit) / (1 + exp(preds$fit))
se.bands.logit <- cbind(lower.bound = preds$fit - 2*preds$se.fit,
                        upper.bound = preds$fit + 2*preds$se.fit)
se.bands <- exp(se.bands.logit)/(1 + exp(se.bands.logit))

# —Ä–µ–∑—É–ª—å—Ç–∞—Ç - –¥–æ–≤–µ—Ä–∏—Ç–µ–ª—å–Ω—ã–π –∏–Ω—Ç–µ—Ä–≤–∞–ª –¥–ª—è –≤–µ—Ä–æ—è—Ç–Ω–æ—Å—Ç–∏ —Å–æ–±—ã—Ç–∏—è 
#   "–ó–∞—Ä–∞–±–æ—Ç–Ω–∞—è –ø–ª–∞—Ç–∞ –≤—ã—à–µ 250".   
round(head(se.bands), 3)
##   lower.bound upper.bound
## 1           0       0.002
## 2           0       0.003
## 3           0       0.004
## 4           0       0.005
## 5           0       0.006
## 6           0       0.007
–î–æ—Å—Ç—Ä–∞–∏–≤–∞–µ–º –≥—Ä–∞—Ñ–∏–∫ —Å 4 —Å–ª–∞–π–¥–∞ –ø—Ä–µ–∑–µ–Ω—Ç–∞—Ü–∏–∏ (—Ä–∏—Å. 7.1 –∫–Ω–∏–≥–∏). –†–∏—Å—É–µ–º –ø—Ä–∞–≤—É—é –ø–∞–Ω–µ–ª—å.

# —Å–µ—Ç–∫–∞ –¥–ª—è –≥—Ä–∞—Ñ–∏–∫–∞ (–∏–∑–æ–±—Ä–∞–∂–∞–µ–º –≤–µ—Ä–æ—è—Ç–Ω–æ—Å—Ç–∏, –ø–æ—ç—Ç–æ–º—É –∏–Ω—Ç–µ—Ä–≤–∞–ª –∏–∑–º–µ–Ω–µ–Ω–∏—è y –º–∞–ª)
plot(age, I(wage > 250), xlim = agelims, type = 'n', ylim = c(0, 0.2),
     ylab = 'P(Wage > 250 | Age)')

# —Ñ–∞–∫—Ç–∏—á–µ—Å–∫–∏–µ –Ω–∞–±–ª—é–¥–µ–Ω–∏—è –ø–æ–∫–∞–∑—ã–≤–∞–µ–º –∑–∞—Å–µ—á–∫–∞–º–∏
points(jitter(age), I((wage > 250) / 5), cex = 0.5, pch = '|', col = 'darkgrey')

# –º–æ–¥–µ–ª—å
lines(age.grid, pfit, lwd = 2, col = 'blue')

# –¥–æ–≤–µ—Ä–∏—Ç–µ–ª—å–Ω—ã–µ –∏–Ω—Ç–µ—Ä–≤–∞–ª—ã
matlines(age.grid, se.bands, lwd = 1, col = 'blue', lty = 3)

# –∑–∞–≥–æ–ª–æ–≤–æ–∫
title('–ü–æ–ª–∏–Ω–æ–º —á–µ—Ç–≤—ë—Ä—Ç–æ–π —Å—Ç–µ–ø–µ–Ω–∏')


–°—Ç—É–ø–µ–Ω—á–∞—Ç—ã–µ —Ñ—É–Ω–∫—Ü–∏–∏
–î–ª—è –Ω–∞—á–∞–ª–∞ –æ–ø—Ä–µ–¥–µ–ª–∏–º –Ω–µ—Å–∫–æ–ª—å–∫–æ –∏–Ω—Ç–µ—Ä–≤–∞–ª–æ–≤, –Ω–∞ –∫–∞–∂–¥–æ–º –∏–∑ –∫–æ—Ç–æ—Ä—ã—Ö –±—É–¥–µ–º –º–æ–¥–µ–ª–∏—Ä–æ–≤–∞—Ç—å –∑–∞–≤–∏—Å–∏–º–æ—Å—Ç—å wage –æ—Ç age —Å–≤–æ–∏–º —Å—Ä–µ–¥–Ω–∏–º —É—Ä–æ–≤–Ω–µ–º.

# –Ω–∞—Ä–µ–∑–∞–µ–º –ø—Ä–µ–¥–∏–∫—Ç–æ—Ä age –Ω–∞ 4 —Ä–∞–≤–Ω—ã—Ö –∏–Ω—Ç–µ—Ä–≤–∞–ª–∞
table(cut(age, 4))
## 
## (17.9,33.5]   (33.5,49]   (49,64.5] (64.5,80.1] 
##         750        1399         779          72
# –ø–æ–¥–≥–æ–Ω—è–µ–º –ª–∏–Ω–µ–π–Ω—É—é –º–æ–¥–µ–ª—å –Ω–∞ –∏–Ω—Ç–µ—Ä–≤–∞–ª–∞—Ö
fit <- lm(wage ~ cut(age, 4), data = Wage)
round(coef(summary(fit)), 2)
##                        Estimate Std. Error t value Pr(>|t|)
## (Intercept)               94.16       1.48   63.79     0.00
## cut(age, 4)(33.5,49]      24.05       1.83   13.15     0.00
## cut(age, 4)(49,64.5]      23.66       2.07   11.44     0.00
## cut(age, 4)(64.5,80.1]     7.64       4.99    1.53     0.13
# –ø—Ä–æ–≥–Ω–æ–∑ -- —ç—Ç–æ —Å—Ä–µ–¥–Ω–∏–µ –ø–æ `wage` –Ω–∞ –∫–∞–∂–¥–æ–º –∏–Ω—Ç–µ—Ä–≤–∞–ª–µ
preds.cut <- predict(fit, newdata = list(age = age.grid), se = T)

# –∏–Ω—Ç–µ—Ä–≤–∞–ª—å–Ω—ã–π –ø—Ä–æ–≥–Ω–æ–∑
se.bands.cut <- cbind(lower.bound = preds.cut$fit - 2*preds.cut$se.fit,
                      upper.bound = preds.cut$fit + 2*preds.cut$se.fit)
–í–æ—Å–ø—Ä–æ–∏–∑–≤–µ–¥—ë–º –≥—Ä–∞—Ñ–∏–∫ —Å–æ —Å–ª–∞–π–¥–∞ 7 –ø—Ä–µ–∑–µ–Ω—Ç–∞—Ü–∏–∏ (—Ä–∏—Å. 7.2 –∫–Ω–∏–≥–∏).

# –Ω–∞–±–ª—é–¥–µ–Ω–∏—è
plot(age, wage, xlim = agelims, cex = 0.5, col = 'darkgrey')

# –º–æ–¥–µ–ª—å
lines(age.grid, preds.cut$fit, lwd = 2, col = 'darkgreen')

# –¥–æ–≤–µ—Ä–∏—Ç–µ–ª—å–Ω—ã–µ –∏–Ω—Ç–µ—Ä–≤–∞–ª—ã –ø—Ä–æ–≥–Ω–æ–∑–∞
matlines(x = age.grid, y = se.bands.cut, lwd = 1, col = 'darkgreen', lty = 3)

# –∑–∞–≥–æ–ª–æ–≤–æ–∫
title('–°—Ç—É–ø–µ–Ω—á–∞—Ç–∞—è —Ñ—É–Ω–∫—Ü–∏—è')


–ü—Ä–∞–≤–∞—è —á–∞—Å—Ç—å –≥—Ä–∞—Ñ–∏–∫–∞, –¥–ª—è –≤–µ—Ä–æ—è—Ç–Ω–æ—Å—Ç–∏ —Ç–æ–≥–æ, —á—Ç–æ –∑–∞—Ä–ø–ª–∞—Ç–∞ –≤—ã—à–µ 250.

fit <- glm(I(wage > 250) ~ cut(age, 4), data = Wage, family = 'binomial')

# –ø—Ä–æ–≥–Ω–æ–∑—ã
preds <- predict(fit, newdata = list(age = age.grid), se = T)

# –ø–µ—Ä–µ—Å—á–∏—Ç—ã–≤–∞–µ–º –¥–æ–≤–µ—Ä–∏—Ç–µ–ª—å–Ω—ã–µ –∏–Ω—Ç–µ—Ä–≤–∞–ª—ã –∏ –ø—Ä–æ–≥–Ω–æ–∑—ã –≤ –∏—Å—Ö–æ–¥–Ω—ã–µ –ï–ò
pfit <- exp(preds$fit) / (1 + exp(preds$fit))
se.bands.logit <- cbind(lower.bound = preds$fit - 2*preds$se.fit,
                        upper.bound = preds$fit + 2*preds$se.fit)
se.bands <- exp(se.bands.logit)/(1 + exp(se.bands.logit))

# —Ä–µ–∑—É–ª—å—Ç–∞—Ç - –¥–æ–≤–µ—Ä–∏—Ç–µ–ª—å–Ω—ã–π –∏–Ω—Ç–µ—Ä–≤–∞–ª –¥–ª—è –≤–µ—Ä–æ—è—Ç–Ω–æ—Å—Ç–∏ —Å–æ–±—ã—Ç–∏—è 
#   "–ó–∞—Ä–∞–±–æ—Ç–Ω–∞—è –ø–ª–∞—Ç–∞ –≤—ã—à–µ 250".   
round(head(se.bands), 3)
##   lower.bound upper.bound
## 1       0.003       0.016
## 2       0.003       0.016
## 3       0.003       0.016
## 4       0.003       0.016
## 5       0.003       0.016
## 6       0.003       0.016
# —Å–µ—Ç–∫–∞ –¥–ª—è –≥—Ä–∞—Ñ–∏–∫–∞ (–∏–∑–æ–±—Ä–∞–∂–∞–µ–º –≤–µ—Ä–æ—è—Ç–Ω–æ—Å—Ç–∏, –ø–æ—ç—Ç–æ–º—É –∏–Ω—Ç–µ—Ä–≤–∞–ª –∏–∑–º–µ–Ω–µ–Ω–∏—è y –º–∞–ª)
plot(age, I(wage > 250), xlim = agelims, type = 'n', ylim = c(0, 0.2),
     ylab = 'P(Wage > 250 | Age)')

# —Ñ–∞–∫—Ç–∏—á–µ—Å–∫–∏–µ –Ω–∞–±–ª—é–¥–µ–Ω–∏—è –ø–æ–∫–∞–∑—ã–≤–∞–µ–º –∑–∞—Å–µ—á–∫–∞–º–∏
points(jitter(age), I((wage > 250) / 5), cex = 0.5, pch = '|', col = 'darkgrey')

# –º–æ–¥–µ–ª—å
lines(age.grid, pfit, lwd = 2, col = 'darkgreen')

# –¥–æ–≤–µ—Ä–∏—Ç–µ–ª—å–Ω—ã–µ –∏–Ω—Ç–µ—Ä–≤–∞–ª—ã
matlines(age.grid, se.bands, lwd = 1, col = 'darkgreen', lty = 3)

# –∑–∞–≥–æ–ª–æ–≤–æ–∫
title('–°—Ç—É–ø–µ–Ω—á–∞—Ç–∞—è —Ñ—É–Ω–∫—Ü–∏—è')


–°–ø–ª–∞–π–Ω—ã
–ü–æ—Å—Ç—Ä–æ–∏–º –∫—É–±–∏—á–µ—Å–∫–∏–π —Å–ø–ª–∞–π–Ω —Å —Ç—Ä–µ–º—è —É–∑–ª–∞–º–∏.

# –∫—É–±–∏—á–µ—Å–∫–∏–π —Å–ø–ª–∞–π–Ω —Å —Ç—Ä–µ–º—è —É–∑–ª–∞–º–∏
fit <- lm(wage ~ bs(age, knots = c(25, 40, 60)), data = Wage)
# –ø—Ä–æ–≥–Ω–æ–∑
preds.spl <- predict(fit, newdata = list(age = age.grid), se = T)
–¢–µ–ø–µ—Ä—å –ø–æ—Å—Ç—Ä–æ–∏–º –Ω–∞—Ç—É—Ä–∞–ª—å–Ω—ã–π –ø–æ —Ç—Ä—ë–º —É–∑–ª–∞–º. –¢—Ä–∏ —É–∑–ª–∞ —ç—Ç–æ 6 —Å—Ç–µ–ø–µ–Ω–µ–π —Å–≤–æ–±–æ–¥—ã. –ï—Å–ª–∏ —Ñ—É–Ω–∫—Ü–∏–∏ bs(), –∫–æ—Ç–æ—Ä–∞—è —Å–æ–∑–¥–∞—ë—Ç –º–∞—Ç—Ä–∏—Ü—É —Å –±–∞–∑–∏—Å–æ–º –¥–ª—è –ø–æ–ª–∏–Ω–æ–º–∏–∞–ª—å–Ω–æ–≥–æ —Å–ø–ª–∞–π–Ω–∞, –ø–µ—Ä–µ–¥–∞—Ç—å —Ç–æ–ª—å–∫–æ —Å—Ç–µ–ø–µ–Ω–∏ —Å–≤–æ–±–æ–¥—ã, –æ–Ω–∞ —Ä–∞—Å–ø—Ä–µ–¥–µ–ª–∏—Ç —É–∑–ª—ã —Ä–∞–≤–Ω–æ–º–µ—Ä–Ω–æ. –í –¥–∞–Ω–Ω–æ–º —Å–ª—É—á–∞–µ —ç—Ç–æ –∫–≤–∞—Ä—Ç–∏–ª–∏ —Ä–∞—Å–ø—Ä–µ–¥–µ–ª–µ–Ω–∏—è age.

# 3 —É–∑–ª–∞ -- 6 —Å—Ç–µ–ø–µ–Ω–µ–π —Å–≤–æ–±–æ–¥—ã (—Å—Ç–æ–ª–±—Ü—ã –º–∞—Ç—Ä–∏—Ü—ã)
dim(bs(age, knots = c(25, 40, 60)))
## [1] 3000    6
# –µ—Å–ª–∏ –Ω–µ —É–∫–∞–∑—ã–≤–∞–µ–º —É–∑–ª—ã —è–≤–Ω–æ...
dim(bs(age, df = 6))
## [1] 3000    6
#  –æ–Ω–∏ –ø—Ä–∏–≤—è–∑—ã–≤–∞—é—Ç—Å—è –∫ –∫–≤–∞—Ä—Ç–∏–ª—è–º
attr(bs(age, df = 6), 'knots')
##   25%   50%   75% 
## 33.75 42.00 51.00
# –Ω–∞—Ç—É—Ä–∞–ª—å–Ω—ã–π —Å–ø–ª–∞–π–Ω
fit2 <- lm(wage ~ ns(age, df = 4), data = Wage)
preds.spl2 <- predict(fit2, newdata = list(age = age.grid), se = T)
–ì—Ä–∞—Ñ–∏–∫ —Å—Ä–∞–≤–Ω–µ–Ω–∏—è –∫—É–±–∏—á–µ—Å–∫–æ–≥–æ –∏ –Ω–∞—Ç—É—Ä–∞–ª—å–Ω–æ–≥–æ —Å–ø–ª–∞–π–Ω–æ–≤.

par(mfrow = c(1, 1), mar = c(4.5, 4.5, 1, 8.5), oma = c(0, 0, 0, 0), xpd = T)

# –Ω–∞–±–ª—é–¥–µ–Ω–∏—è
plot(age, wage, col = 'grey')

# –º–æ–¥–µ–ª—å –∫—É–±–∏—á–µ—Å–∫–æ–≥–æ —Å–ø–ª–∞–π–Ω–∞
lines(age.grid, preds.spl$fit, lwd = 2)

# –¥–æ–≤–µ—Ä–∏—Ç–µ–ª—å–Ω—ã–π –∏–Ω—Ç–µ—Ä–≤–∞–ª
lines(age.grid, preds.spl$fit + 2*preds.spl$se, lty = 'dashed')
lines(age.grid, preds.spl$fit - 2*preds.spl$se, lty = 'dashed')

# –Ω–∞—Ç—É—Ä–∞–ª—å–Ω—ã–π —Å–ø–ª–∞–π–Ω
lines(age.grid, preds.spl2$fit, col = 'red', lwd = 2)

# –ª–µ–≥–µ–Ω–¥–∞
legend("topright", inset = c(-0.7, 0),
       c('–ö—É–±–∏—á–µ—Å–∫–∏–π \n —Å 3 —É–∑–ª–∞–º–∏', '–ù–∞—Ç—É—Ä–∞–ª—å–Ω—ã–π'),
       lwd = rep(2, 2), col = c('black', 'red'))

# –∑–∞–≥–æ–ª–æ–≤–æ–∫
title("–°–ø–ª–∞–π–Ω—ã")


–ü–æ—Å—Ç—Ä–æ–∏–º –≥—Ä–∞—Ñ–∏–∫ —Å–æ —Å–ª–∞–π–¥–∞ 20 (—Ä–∏—Å—É–Ω–æ–∫ 7.8 –∫–Ω–∏–≥–∏).

par(mfrow = c(1, 1), mar = c(4.5, 4.5, 1, 1), oma = c(0, 0, 4, 0))

# –Ω–∞–±–ª—é–¥–µ–Ω–∏—è
plot(age, wage, xlim = agelims, cex = 0.5, col = 'darkgrey')

# –∑–∞–≥–æ–ª–æ–≤–æ–∫
title('–°–≥–ª–∞–∂–∏–≤–∞—é—â–∏–π —Å–ø–ª–∞–π–Ω')

# –ø–æ–¥–≥–æ–Ω—è–µ–º –º–æ–¥–µ–ª—å —Å 16 —Å—Ç–µ–ø–µ–Ω—è–º–∏ —Å–≤–æ–±–æ–¥—ã
fit <- smooth.spline(age, wage, df = 16)

# –ø–æ–¥–≥–æ–Ω—è–µ–º –º–æ–¥–µ–ª—å —Å –ø–æ–¥–±–æ—Ä–æ–º –ª—è–º–±–¥—ã —Å –ø–æ–º–æ—â—å—é –ø–µ—Ä–µ–∫—Ä—ë—Å—Ç–Ω–æ–π –ø—Ä–æ–≤–µ—Ä–∫–∏
fit2 <- smooth.spline(age, wage, cv = T)
## Warning in smooth.spline(age, wage, cv = T): cross-validation with non-
## unique 'x' values seems doubtful
fit2$df
## [1] 6.794596
# —Ä–∏—Å—É–µ–º –º–æ–¥–µ–ª—å
lines(fit, col = 'red', lwd = 2)
lines(fit2, col = 'blue', lwd = 2)
legend('topright', 
       c('16 df', '6.8 df'),
       col = c('red', 'blue'), lty = 1, lwd = 2, cex = 0.8)


–õ–æ–∫–∞–ª—å–Ω–∞—è —Ä–µ–≥—Ä–µ—Å—Å–∏—è
–°—Ç—Ä–æ–∏–º –≥—Ä–∞—Ñ–∏–∫ —Å–æ —Å–ª–∞–π–¥–∞ 24 (—Ä–∏—Å. 7.10).

plot(age, wage, xlim = agelims, cex = 0.5, col = 'darkgrey')

title('–õ–æ–∫–∞–ª—å–Ω–∞—è —Ä–µ–≥—Ä–µ—Å—Å–∏—è')

# –ø–æ–¥–≥–æ–Ω—è–µ–º –º–æ–¥–µ–ª—å c –æ–∫–Ω–æ–º 0.2
fit <- loess(wage ~ age, span = 0.2, data = Wage)

# –ø–æ–¥–≥–æ–Ω—è–µ–º –º–æ–¥–µ–ª—å c –æ–∫–Ω–æ–º 0.5
fit2 <- loess(wage ~ age, span = 0.5, data = Wage)

# —Ä–∏—Å—É–º –º–æ–¥–µ–ª–∏
lines(age.grid, predict(fit, data.frame(age = age.grid)),
      col = 'red', lwd = 2)
lines(age.grid, predict(fit2, data.frame(age = age.grid)),
      col = 'blue', lwd = 2)

# –ª–µ–≥–µ–Ω–¥–∞
legend('topright', 
       c('s = 0.2', 's = 0.5'),
       col = c('red', 'blue'), lty = 1, lwd = 2, cex = 0.8)


–û–±–æ–±—â—ë–Ω–Ω—ã–µ –∞–¥–¥–∏—Ç–∏–≤–Ω—ã–µ –º–æ–¥–µ–ª–∏ (GAM) —Å –Ω–µ–ø—Ä–µ—Ä—ã–≤–Ω—ã–º –æ—Ç–∫–ª–∏–∫–æ–º
–ü–æ—Å—Ç—Ä–æ–∏–º GAM –Ω–∞ –Ω–∞—Ç—É—Ä–∞–ª—å–Ω—ã—Ö —Å–ø–ª–∞–π–Ω–∞—Ö —Å—Ç–µ–ø–µ–Ω–µ–π 4 (year), 5 (age) —Å –∫–∞—Ç–µ–≥–æ—Ä–∏–∞–ª—å–Ω—ã–º –ø—Ä–µ–¥–∏–∫—Ç–æ—Ä–æ–º edication.

# GAM –Ω–∞ –Ω–∞—Ç—É—Ä–∞–ª—å–Ω—ã—Ö —Å–ø–ª–∞–π–Ω–∞—Ö
gam.ns <- gam(wage ~ ns(year, 4) + ns(age, 5) + education, data = Wage)
–¢–∞–∫–∂–µ –ø–æ—Å—Ç—Ä–æ–∏–º –º–æ–¥–µ–ª—å –Ω–∞ —Å–≥–ª–∞–∂–∏–≤–∞—é—â–∏—Ö —Å–ø–ª–∞–π–Ω–∞—Ö.

# GAM –Ω–∞ —Å–≥–ª–∞–∂–∏–≤–∞—é—â–∏—Ö —Å–ø–ª–∞–π–Ω–∞—Ö
gam.m3 <- gam(wage ~ s(year, 4) + s(age, 5) + education, data = Wage)
–ì—Ä–∞—Ñ–∏–∫ —Å–æ —Å–ª–∞–π–¥–∞ 28 (—Ä–∏—Å. 7.12).

par(mfrow = c(1, 3))
plot(gam.m3, se = T, col = 'blue')


–ì—Ä–∞—Ñ–∏–∫ —Å–æ —Å–ª–∞–π–¥–∞ 27 (—Ä–∏—Å. 7.11).

par(mfrow = c(1, 3))
plot(gam.ns, se = T, col = 'red')


–ì—Ä–∞—Ñ–∏–∫ —Ñ—É–Ω–∫—Ü–∏–∏ –æ—Ç year –ø–æ—Ö–æ–∂ –Ω–∞ –ø—Ä—è–º—É—é. –°–¥–µ–ª–∞–µ–º ANOVA, —á—Ç–æ–±—ã –ø–æ–Ω—è—Ç—å, –∫–∞–∫–∞—è —Å—Ç–µ–ø–µ–Ω—å –¥–ª—è year –ª—É—á—à–µ.

gam.m1 <- gam(wage ~ s(age, 5) + education, data = Wage)          # –±–µ–∑ year
gam.m2 <- gam(wage ~ year + s(age, 5) + education, data = Wage)   # year^1

anova(gam.m1, gam.m2, gam.m3, test = 'F')
Resid. Df
<dbl>
    Resid. Dev
<dbl>
    Df
<dbl>
    Deviance
<dbl>
    F
<dbl>
    Pr(>F)
<dbl>
    2990	3711731	NA	NA	NA	NA
2989	3693842	1.000000	17889.243	14.477130	0.0001447167
2986	3689770	2.999989	4071.134	1.098212	0.3485661430
3 rows
–¢—Ä–µ—Ç—å—è –º–æ–¥–µ–ª—å —Å—Ç–∞—Ç–∏—Å—Ç–∏—á–µ—Å–∫–∏ –Ω–µ –ª—É—á—à–µ –≤—Ç–æ—Ä–æ–π. –ö—Ä–æ–º–µ —Ç–æ–≥–æ, –æ–¥–∏–Ω –∏–∑ –ø–∞—Ä–∞–º–µ—Ç—Ä–æ–≤ —ç—Ç–æ–π –º–æ–¥–µ–ª–∏ –Ω–µ–∑–Ω–∞—á–∏–º.

# —Å–≤–æ–¥–∫–∞ –ø–æ –º–æ–¥–µ–ª–∏ gam.m3
summary(gam.m3)
## 
## Call: gam(formula = wage ~ s(year, 4) + s(age, 5) + education, data = Wage)
## Deviance Residuals:
##     Min      1Q  Median      3Q     Max 
## -119.43  -19.70   -3.33   14.17  213.48 
## 
## (Dispersion Parameter for gaussian family taken to be 1235.69)
## 
##     Null Deviance: 5222086 on 2999 degrees of freedom
## Residual Deviance: 3689770 on 2986 degrees of freedom
## AIC: 29887.75 
## 
## Number of Local Scoring Iterations: 2 
## 
## Anova for Parametric Effects
##              Df  Sum Sq Mean Sq F value    Pr(>F)    
## s(year, 4)    1   27162   27162  21.981 2.877e-06 ***
## s(age, 5)     1  195338  195338 158.081 < 2.2e-16 ***
## education     4 1069726  267432 216.423 < 2.2e-16 ***
## Residuals  2986 3689770    1236                      
## ---
## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
## 
## Anova for Nonparametric Effects
##             Npar Df Npar F  Pr(F)    
## (Intercept)                          
## s(year, 4)        3  1.086 0.3537    
## s(age, 5)         4 32.380 <2e-16 ***
## education                            
## ---
## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
–†–∞–±–æ—Ç–∞–µ–º —Å –º–æ–¥–µ–ª—å—é gam.m2.

# –ø—Ä–æ–≥–Ω–æ–∑ –ø–æ –æ–±—É—á–∞—é—â–µ–π –≤—ã–±–æ—Ä–∫–µ
preds <- predict(gam.m2, newdata = Wage)
–¢–∞–∫–∂–µ –º–æ–∂–Ω–æ –∏—Å–ø–æ–ª—å–∑–æ–≤–∞—Ç—å –≤ GAM –ª–æ–∫–∞–ª—å–Ω—ã–µ —Ä–µ–≥—Ä–µ—Å—Å–∏–∏.

# GAM –Ω–∞ –ª–æ–∫–∞–ª—å–Ω—ã—Ö —Ä–µ–≥—Ä–µ—Å—Å–∏—è—Ö
gam.lo <- gam(wage ~ s(year, df = 4) + lo(age, span = 0.7) + education, 
              data = Wage)

par(mfrow = c(1, 3))
plot.gam(gam.lo, se = T, col = 'green')


# –º–æ–¥–µ–ª—å —Å–æ –≤–∑–∞–∏–º–æ–¥–µ–π—Å—Ç–≤–∏–µ–º —Ä–µ–≥—Ä–µ—Å—Å–æ—Ä–æ–≤ year –∏ age
gam.lo.i <- gam(wage ~ lo(year, age, span = 0.5) + education, data = Wage)
## Warning in lo.wam(x, z, wz, fit$smooth, which, fit$smooth.frame,
## bf.maxit, : liv too small. (Discovered by lowesd)
## Warning in lo.wam(x, z, wz, fit$smooth, which, fit$smooth.frame,
## bf.maxit, : lv too small. (Discovered by lowesd)
## Warning in lo.wam(x, z, wz, fit$smooth, which, fit$smooth.frame,
## bf.maxit, : liv too small. (Discovered by lowesd)
## Warning in lo.wam(x, z, wz, fit$smooth, which, fit$smooth.frame,
## bf.maxit, : lv too small. (Discovered by lowesd)
plot(gam.lo.i)


–õ–æ–≥–∏—Å—Ç–∏—á–µ—Å–∫–∞—è GAM
–ü–æ—Å—Ç—Ä–æ–∏–º –ª–æ–≥–∏—Å—Ç–∏—á–µ—Å–∫—É—é GAM –¥–ª—è –≤—Å–µ—Ä–æ—è—Ç–Ω–æ—Å—Ç–∏ —Ç–æ–≥–æ, —á—Ç–æ wage –ø—Ä–µ–≤—ã—à–∞–µ—Ç 250.

gam.lr <- gam(I(wage > 250) ~ year + s(age, df = 5) + education, 
              family = 'binomial', data = Wage)
par(mfrow = c(1, 3))
plot(gam.lr, se = T, col = 'green')


# —É—Ä–æ–≤–Ω–∏ –æ–±—Ä–∞–∑–æ–≤–∞–Ω–∏—è –ø–æ –≥—Ä—É–ø–ø–∞–º —Ä–∞–∑–Ω–æ–≥–æ –¥–æ—Å—Ç–∞—Ç–∫–∞
table(education, I(wage > 250))
##                     
## education            FALSE TRUE
##   1. < HS Grad         268    0
##   2. HS Grad           966    5
##   3. Some College      643    7
##   4. College Grad      663   22
##   5. Advanced Degree   381   45
–í –∫–∞—Ç–µ–≥–æ—Ä–∏–∏ —Å —Å–∞–º—ã–º –Ω–∏–∑–∫–∏–º —É—Ä–æ–≤–Ω–µ–º –æ–±—Ä–∞–∑–æ–≤–∞–Ω–∏—è –Ω–µ—Ç wage > 250, –ø–æ—ç—Ç–æ–º—É —É–±–∏—Ä–∞–µ–º –µ—ë.

gam.lr.s <- gam(I(wage > 250) ~ year + s(age, df = 5) + education,
                family = 'binomial', data = Wage, 
                subset = (education != "1. < HS Grad"))
# –≥—Ä–∞—Ñ–∏–∫
par(mfrow = c(1, 3))
plot(gam.lr.s, se = T, col = 'green')


detach(Wage)
```


```{r, eval=FALSE}
# Nonlinear modeling
–ú–∞—Ç–µ–º–∞—Ç–∏—á–µ—Å–∫–æ–µ –º–æ–¥–µ–ª–∏—Ä–æ–≤–∞–Ω–∏–µ
–ü—Ä–∞–∫—Ç–∏–∫–∞ 8
–ù–µ–ª–∏–Ω–µ–π–Ω—ã–µ –º–æ–¥–µ–ª–∏
–í –ø—Ä–∞–∫—Ç–∏—á–µ—Å–∫–∏—Ö –ø—Ä–∏–º–µ—Ä–∞—Ö –Ω–∏–∂–µ –ø–æ–∫–∞–∑–∞–Ω–æ –∫–∞–∫:
    
    —Å—Ç—Ä–æ–∏—Ç—å —Ä–µ–≥—Ä–µ—Å—Å–∏–æ–Ω–Ω—ã–µ –¥–µ—Ä–µ–≤—å—è;
—Å—Ç—Ä–æ–∏—Ç—å –¥–µ—Ä–µ–≤—å—è –∫–ª–∞—Å—Å–∏—Ñ–∏–∫–∞—Ü–∏–∏;
–¥–µ–ª–∞—Ç—å –æ–±—Ä–µ–∑–∫—É –¥–µ—Ä–µ–≤–∞;
–∏—Å–ø–æ–ª—å–∑–æ–≤–∞—Ç—å –±—ç–≥–≥–∏–Ω–≥, –±—É—Å—Ç–∏–Ω–≥, —Å–ª—É—á–∞–π–Ω—ã–π –ª–µ—Å –¥–ª—è —É–ª—É—á—à–µ–Ω–∏—è –∫–∞—á–µ—Å—Ç–≤–∞ –ø—Ä–æ–≥–Ω–æ–∑–∏—Ä–æ–≤–∞–Ω–∏—è.
–ú–æ–¥–µ–ª–∏: –¥–µ—Ä–µ–≤—å—è —Ä–µ—à–µ–Ω–∏–π.
–î–∞–Ω–Ω—ã–µ: Sales {ISLR}, Boston {ISLR}

–ü–æ–¥—Ä–æ–±–Ω—ã–µ –∫–æ–º–º–µ–Ω—Ç–∞—Ä–∏–∏ –∫ –∫–æ–¥—É –ª–∞–±–æ—Ä–∞—Ç–æ—Ä–Ω—ã—Ö —Å–º. –≤ [1], –≥–ª–∞–≤–∞ 8.

library('tree')              # –¥–µ—Ä–µ–≤—å—è
## Warning: package 'tree' was built under R version 3.4.4
library('ISLR')              # –Ω–∞–±–æ—Ä—ã –¥–∞–Ω–Ω—ã—Ö
library('MASS')
library('randomForest')      # —Å–ª—É—á–∞–π–Ω—ã–π –ª–µ—Å
## Warning: package 'randomForest' was built under R version 3.4.4
## randomForest 4.6-14
## Type rfNews() to see new features/changes/bug fixes.
library('gbm')
## Warning: package 'gbm' was built under R version 3.4.4
## Loading required package: survival
## Loading required package: lattice
## Loading required package: splines
## Loading required package: parallel
## Loaded gbm 2.1.3
–î–µ—Ä–µ–≤—å—è —Ä–µ—à–µ–Ω–∏–π
–ó–∞–≥—Ä—É–∑–∏–º —Ç–∞–±–ª–∏—Ü—É —Å –¥–∞–Ω–Ω—ã–º–∏ –ø–æ –ø—Ä–æ–¥–∞–∂–∞–º –¥–µ—Ç—Å–∫–∏—Ö –∫—Ä–µ—Å–µ–ª –∏ –¥–æ–±–∞–≤–∏–º –∫ –Ω–µ–π –ø–µ—Ä–µ–º–µ–Ω–Ω—É—é High ‚Äì ‚Äú–≤—ã—Å–æ–∫–∏–µ –ø—Ä–æ–¥–∞–∂–∏‚Äù —Å–æ –∑–Ω–∞—á–µ–Ω–∏—è–º–∏:
    
    Yes –µ—Å–ª–∏ –ø—Ä–æ–¥–∞–∂–∏ –±–æ–ª—å—à–µ 8 (—Ç—ã—Å. —à—Ç.);
No –≤ –ø—Ä–æ—Ç–∏–≤–Ω–æ–º —Å–ª—É—á–∞–µ.
?Carseats
## starting httpd help server ... done
attach(Carseats)

# –Ω–æ–≤–∞—è –ø–µ—Ä–µ–º–µ–Ω–Ω–∞—è
High <- ifelse(Sales <= 8, "No", "Yes")

# –ø—Ä–∏—Å–æ–µ–¥–∏–Ω—è–µ–º –∫ —Ç–∞–±–ª–∏—Ü–µ –¥–∞–Ω–Ω—ã—Ö
Carseats <- data.frame(Carseats, High)
–°—Ç—Ä–æ–∏–º –¥–µ—Ä–µ–≤–æ –¥–ª—è –∫–∞—Ç–µ–≥–æ—Ä–∏–∞–ª—å–Ω–æ–≥–æ –æ—Ç–∫–ª–∏–∫–∞ High, –æ—Ç–±—Ä–æ—Å–∏–≤ –Ω–µ–ø—Ä–µ—Ä—ã–≤–Ω—ã–π –æ—Ç–∫–ª–∏–∫ Sales.

# –º–æ–¥–µ–ª—å –±–∏–Ω–∞—Ä–Ω–æ–≥–æ  –¥–µ—Ä–µ–≤–∞
tree.carseats <- tree(High ~ . -Sales, Carseats)
summary(tree.carseats)
## 
## Classification tree:
## tree(formula = High ~ . - Sales, data = Carseats)
## Variables actually used in tree construction:
## [1] "ShelveLoc"   "Price"       "Income"      "CompPrice"   "Population" 
## [6] "Advertising" "Age"         "US"         
## Number of terminal nodes:  27 
## Residual mean deviance:  0.4575 = 170.7 / 373 
## Misclassification error rate: 0.09 = 36 / 400
# –≥—Ä–∞—Ñ–∏–∫ —Ä–µ–∑—É–ª—å—Ç–∞—Ç–∞
plot(tree.carseats)            # –≤–µ—Ç–≤–∏
text(tree.carseats, pretty=0)  # –ø–æ–¥–ø–∏—Å–∏


tree.carseats                  # –ø–æ—Å–º–æ—Ç—Ä–µ—Ç—å –≤—Å—ë –¥–µ—Ä–µ–≤–æ –≤ –∫–æ–Ω—Å–æ–ª–∏
## node), split, n, deviance, yval, (yprob)
##       * denotes terminal node
## 
##   1) root 400 541.500 No ( 0.59000 0.41000 )  
##     2) ShelveLoc: Bad,Medium 315 390.600 No ( 0.68889 0.31111 )  
##       4) Price < 92.5 46  56.530 Yes ( 0.30435 0.69565 )  
##         8) Income < 57 10  12.220 No ( 0.70000 0.30000 )  
##          16) CompPrice < 110.5 5   0.000 No ( 1.00000 0.00000 ) *
##          17) CompPrice > 110.5 5   6.730 Yes ( 0.40000 0.60000 ) *
##         9) Income > 57 36  35.470 Yes ( 0.19444 0.80556 )  
##          18) Population < 207.5 16  21.170 Yes ( 0.37500 0.62500 ) *
##          19) Population > 207.5 20   7.941 Yes ( 0.05000 0.95000 ) *
##       5) Price > 92.5 269 299.800 No ( 0.75465 0.24535 )  
##        10) Advertising < 13.5 224 213.200 No ( 0.81696 0.18304 )  
##          20) CompPrice < 124.5 96  44.890 No ( 0.93750 0.06250 )  
##            40) Price < 106.5 38  33.150 No ( 0.84211 0.15789 )  
##              80) Population < 177 12  16.300 No ( 0.58333 0.41667 )  
##               160) Income < 60.5 6   0.000 No ( 1.00000 0.00000 ) *
##               161) Income > 60.5 6   5.407 Yes ( 0.16667 0.83333 ) *
##              81) Population > 177 26   8.477 No ( 0.96154 0.03846 ) *
##            41) Price > 106.5 58   0.000 No ( 1.00000 0.00000 ) *
##          21) CompPrice > 124.5 128 150.200 No ( 0.72656 0.27344 )  
##            42) Price < 122.5 51  70.680 Yes ( 0.49020 0.50980 )  
##              84) ShelveLoc: Bad 11   6.702 No ( 0.90909 0.09091 ) *
##              85) ShelveLoc: Medium 40  52.930 Yes ( 0.37500 0.62500 )  
##               170) Price < 109.5 16   7.481 Yes ( 0.06250 0.93750 ) *
##               171) Price > 109.5 24  32.600 No ( 0.58333 0.41667 )  
##                 342) Age < 49.5 13  16.050 Yes ( 0.30769 0.69231 ) *
##                 343) Age > 49.5 11   6.702 No ( 0.90909 0.09091 ) *
##            43) Price > 122.5 77  55.540 No ( 0.88312 0.11688 )  
##              86) CompPrice < 147.5 58  17.400 No ( 0.96552 0.03448 ) *
##              87) CompPrice > 147.5 19  25.010 No ( 0.63158 0.36842 )  
##               174) Price < 147 12  16.300 Yes ( 0.41667 0.58333 )  
##                 348) CompPrice < 152.5 7   5.742 Yes ( 0.14286 0.85714 ) *
##                 349) CompPrice > 152.5 5   5.004 No ( 0.80000 0.20000 ) *
##               175) Price > 147 7   0.000 No ( 1.00000 0.00000 ) *
##        11) Advertising > 13.5 45  61.830 Yes ( 0.44444 0.55556 )  
##          22) Age < 54.5 25  25.020 Yes ( 0.20000 0.80000 )  
##            44) CompPrice < 130.5 14  18.250 Yes ( 0.35714 0.64286 )  
##              88) Income < 100 9  12.370 No ( 0.55556 0.44444 ) *
##              89) Income > 100 5   0.000 Yes ( 0.00000 1.00000 ) *
##            45) CompPrice > 130.5 11   0.000 Yes ( 0.00000 1.00000 ) *
##          23) Age > 54.5 20  22.490 No ( 0.75000 0.25000 )  
##            46) CompPrice < 122.5 10   0.000 No ( 1.00000 0.00000 ) *
##            47) CompPrice > 122.5 10  13.860 No ( 0.50000 0.50000 )  
##              94) Price < 125 5   0.000 Yes ( 0.00000 1.00000 ) *
##              95) Price > 125 5   0.000 No ( 1.00000 0.00000 ) *
##     3) ShelveLoc: Good 85  90.330 Yes ( 0.22353 0.77647 )  
##       6) Price < 135 68  49.260 Yes ( 0.11765 0.88235 )  
##        12) US: No 17  22.070 Yes ( 0.35294 0.64706 )  
##          24) Price < 109 8   0.000 Yes ( 0.00000 1.00000 ) *
##          25) Price > 109 9  11.460 No ( 0.66667 0.33333 ) *
##        13) US: Yes 51  16.880 Yes ( 0.03922 0.96078 ) *
##       7) Price > 135 17  22.070 No ( 0.64706 0.35294 )  
##        14) Income < 46 6   0.000 No ( 1.00000 0.00000 ) *
##        15) Income > 46 11  15.160 Yes ( 0.45455 0.54545 ) *
–¢–µ–ø–µ—Ä—å –ø–æ—Å—Ç—Ä–æ–∏–º –¥–µ—Ä–µ–≤–æ –Ω–∞ –æ–±—É—á–∞—é—â–µ–π –≤—ã–±–æ—Ä–∫–µ –∏ –æ—Ü–µ–Ω–∏–º –æ—à–∏–±–∫—É –Ω–∞ —Ç–µ—Å—Ç–æ–≤–æ–π.

# —è–¥—Ä–æ –≥–µ–Ω–µ—Ä–∞—Ç–æ—Ä–∞ —Å–ª—É—á–∞–π–Ω—ã—Ö —á–∏—Å–µ–ª
set.seed(2)

# –æ–±—É—á–∞—é—â–∞—è –≤—ã–±–æ—Ä–∫–∞
train <- sample(1:nrow(Carseats), 200)

# —Ç–µ—Å—Ç–æ–≤–∞—è –≤—ã–±–æ—Ä–∫–∞
Carseats.test <- Carseats[-train,]
High.test <- High[-train]

# —Å—Ç—Ä–æ–∏–º –¥–µ—Ä–µ–≤–æ –Ω–∞ –æ–±—É—á–∞—é—â–µ–π –≤—ã–±–æ—Ä–∫–µ
tree.carseats <- tree(High ~ . -Sales, Carseats, subset = train)

# –¥–µ–ª–∞–µ–º –ø—Ä–æ–≥–Ω–æ–∑
tree.pred <- predict(tree.carseats, Carseats.test, type = "class")

# –º–∞—Ç—Ä–∏—Ü–∞ –Ω–µ—Ç–æ—á–Ω–æ—Å—Ç–µ–π
tbl <- table(tree.pred, High.test)
tbl
##          High.test
## tree.pred No Yes
##       No  86  27
##       Yes 30  57
# –æ—Ü–µ–Ω–∫–∞ —Ç–æ—á–Ω–æ—Å—Ç–∏
acc.test <- sum(diag(tbl))/sum(tbl)
acc.test
## [1] 0.715
–û–±–æ–±—â—ë–Ω–Ω–∞—è —Ö–∞—Ä–∞–∫—Ç–µ—Ä–∏—Å—Ç–∏–∫–∞ —Ç–æ—á–Ω–æ—Å—Ç–∏: –¥–æ–ª—è –≤–µ—Ä–Ω—ã—Ö –ø—Ä–æ–≥–Ω–æ–∑–æ–≤: 0.72.

–¢–µ–ø–µ—Ä—å –æ–±—Ä–µ–∑–∞–µ–º –¥–µ—Ä–µ–≤–æ, –∏—Å–ø–æ–ª—å–∑—É—è –≤ –∫–∞—á–µ—Å—Ç–≤–µ –∫—Ä–∏—Ç–µ—Ä–∏—è —á–∞—Å—Ç–æ—Ç—É –æ—à–∏–±–æ–∫ –∫–ª–∞—Å—Å–∏—Ñ–∏–∫–∞—Ü–∏–∏. –§—É–Ω–∫—Ü–∏—è cv.tree() –ø—Ä–æ–≤–æ–¥–∏—Ç –∫—Ä–æ—Å—Å-–≤–∞–ª–∏–¥–∞—Ü–∏—é –¥–ª—è –≤—ã–±–æ—Ä–∞ –ª—É—á—à–µ–≥–æ –¥–µ—Ä–µ–≤–∞, –∞—Ä–≥—É–º–µ–Ω—Ç prune.misclass –æ–∑–Ω–∞—á–∞–µ—Ç, —á—Ç–æ –º—ã –º–∏–Ω–∏–º–∏–∑–∏—Ä—É–µ–º –æ—à–∏–±–∫—É –∫–ª–∞—Å—Å–∏—Ñ–∏–∫–∞—Ü–∏–∏.

set.seed(3)
cv.carseats <- cv.tree(tree.carseats, FUN = prune.misclass)
# –∏–º–µ–Ω–∞ —ç–ª–µ–º–µ–Ω—Ç–æ–≤ –ø–æ–ª—É—á–µ–Ω–Ω–æ–≥–æ –æ–±—ä–µ–∫—Ç–∞
names(cv.carseats)
## [1] "size"   "dev"    "k"      "method"
# —Å–∞–º –æ–±—ä–µ–∫—Ç
cv.carseats
## $size
## [1] 19 17 14 13  9  7  3  2  1
## 
## $dev
## [1] 55 55 53 52 50 56 69 65 80
## 
## $k
## [1]       -Inf  0.0000000  0.6666667  1.0000000  1.7500000  2.0000000
## [7]  4.2500000  5.0000000 23.0000000
## 
## $method
## [1] "misclass"
## 
## attr(,"class")
## [1] "prune"         "tree.sequence"
# –≥—Ä–∞—Ñ–∏–∫–∏ –∏–∑–º–µ–Ω–µ–Ω–∏—è –ø–∞—Ä–∞–º–µ—Ç—Ä–æ–≤ –º–µ—Ç–æ–¥–∞ –ø–æ —Ö–æ–¥—É –æ–±—Ä–µ–∑–∫–∏ –¥–µ—Ä–µ–≤–∞ ###################

# 1. –æ—à–∏–±–∫–∞ —Å –∫—Ä–æ—Å—Å-–≤–∞–ª–∏–¥–∞—Ü–∏–µ–π –≤ –∑–∞–≤–∏—Å–∏–º–æ—Å—Ç–∏ –æ—Ç —á–∏—Å–ª–∞ —É–∑–ª–æ–≤
par(mfrow = c(1, 2))
plot(cv.carseats$size, cv.carseats$dev, type = "b",
     ylab = '–ß–∞—Å—Ç–æ—Ç–∞ –æ—à–∏–±–æ–∫ —Å –∫—Ä–æ—Å—Å-–≤–∞–ª. (dev)',
     xlab = '–ß–∏—Å–ª–æ —É–∑–ª–æ–≤ (size)')
# —Ä–∞–∑–º–µ—Ä –¥–µ—Ä–µ–≤–∞ —Å –º–∏–Ω–∏–º–∞–ª—å–Ω–æ–π –æ—à–∏–±–∫–æ–π
opt.size <- cv.carseats$size[cv.carseats$dev == min(cv.carseats$dev)]
abline(v = opt.size, col = 'red', 'lwd' = 2)     # —Å–æ–æ—Ç–≤. –≤–µ—Ä—Ç–∏–∫–∞–ª—å–Ω–∞—è –ø—Ä—è–º–∞—è
mtext(opt.size, at = opt.size, side = 1, col = 'red', line = 1)

# 2. –æ—à–∏–±–∫–∞ —Å –∫—Ä–æ—Å—Å-–≤–∞–ª–∏–¥–∞—Ü–∏–µ–π –≤ –∑–∞–≤–∏—Å–∏–º–æ—Å—Ç–∏ –æ—Ç —à—Ç—Ä–∞—Ñ–∞ –Ω–∞ —Å–ª–æ–∂–Ω–æ—Å—Ç—å
plot(cv.carseats$k, cv.carseats$dev, type = "b",
     ylab = '–ß–∞—Å—Ç–æ—Ç–∞ –æ—à–∏–±–æ–∫ —Å –∫—Ä–æ—Å—Å-–≤–∞–ª. (dev)',
     xlab = '–®—Ç—Ä–∞—Ñ –∑–∞ —Å–ª–æ–∂–Ω–æ—Å—Ç—å (k)')


–ö–∞–∫ –≤–∏–¥–Ω–æ –Ω–∞ –≥—Ä–∞—Ñ–∏–∫–µ —Å–ª–µ–≤–∞, –º–∏–Ω–∏–º—É–º —á–∞—Å—Ç–æ—Ç—ã –æ—à–∏–±–æ–∫ –¥–æ—Å—Ç–∏–≥–∞–µ—Ç—Å—è –ø—Ä–∏ —á–∏—Å–ª–µ —É–∑–ª–æ–≤ 9. –û—Ü–µ–Ω–∏–º —Ç–æ—á–Ω–æ—Å—Ç—å –¥–µ—Ä–µ–≤–∞ —Å 9 —É–∑–ª–∞–º–∏.

# –¥–µ—Ä–µ–≤–æ —Å 9 —É–∑–ª–∞–º–∏
prune.carseats <- prune.misclass(tree.carseats, best = 9)

# –≤–∏–∑—É–∞–ª–∏–∑–∞—Ü–∏—è
plot(prune.carseats)
text(prune.carseats, pretty = 0)


# –ø—Ä–æ–≥–Ω–æ–∑ –Ω–∞ —Ç–µ—Å—Ç–æ–≤—É—é –≤—ã–±–æ—Ä–∫—É
tree.pred <- predict(prune.carseats, Carseats.test, type = "class")

# –º–∞—Ç—Ä–∏—Ü–∞ –Ω–µ—Ç–æ—á–Ω–æ—Å—Ç–µ–π
tbl <- table(tree.pred, High.test)
tbl
##          High.test
## tree.pred No Yes
##       No  94  24
##       Yes 22  60
# –æ—Ü–µ–Ω–∫–∞ —Ç–æ—á–Ω–æ—Å—Ç–∏
acc.test <- sum(diag(tbl))/sum(tbl)
acc.test
## [1] 0.77
–¢–æ—á–Ω–æ—Å—Ç—å —ç—Ç–æ–π –º–æ–¥–µ–ª–∏ —á—É—Ç—å –≤—ã—à–µ —Ç–æ—á–Ω–æ—Å—Ç–∏ –∏—Å—Ö–æ–¥–Ω–æ–≥–æ –¥–µ—Ä–µ–≤–∞ –∏ —Å–æ—Å—Ç–∞–≤–ª—è–µ—Ç 0.77. –£–≤–µ–ª–∏—á–∏–≤ –∫–æ–ª–∏—á–µ—Å—Ç–≤–æ —É–∑–ª–æ–≤, –ø–æ–ª—É—á–∏–º –±–æ–ª–µ–µ –≥–ª—É–±–æ–∫–æ–µ –¥–µ—Ä–µ–≤–æ, –Ω–æ –º–µ–Ω–µ–µ —Ç–æ—á–Ω–æ–µ.

# –¥–µ—Ä–µ–≤–æ —Å 13 —É–∑–ª–∞–º–∏
prune.carseats <- prune.misclass(tree.carseats, best = 15)

# –≤–∏–∑—É–∞–ª–∏–∑–∞—Ü–∏—è
plot(prune.carseats)
text(prune.carseats, pretty = 0)


# –ø—Ä–æ–≥–Ω–æ–∑ –Ω–∞ —Ç–µ—Å—Ç–æ–≤—É—é –≤—ã–±–æ—Ä–∫—É
tree.pred <- predict(prune.carseats, Carseats.test, type = "class")

# –º–∞—Ç—Ä–∏—Ü–∞ –Ω–µ—Ç–æ—á–Ω–æ—Å—Ç–µ–π
tbl <- table(tree.pred, High.test)
tbl
##          High.test
## tree.pred No Yes
##       No  86  22
##       Yes 30  62
# –æ—Ü–µ–Ω–∫–∞ —Ç–æ—á–Ω–æ—Å—Ç–∏
acc.test <- sum(diag(tbl))/sum(tbl)
acc.test
## [1] 0.74
# —Å–±—Ä–∞—Å—ã–≤–∞–µ–º –≥—Ä–∞—Ñ–∏—á–µ—Å–∫–∏–µ –ø–∞—Ä–∞–º–µ—Ç—Ä—ã
par(mfrow = c(1, 1))
–†–µ–≥—Ä–µ—Å—Å–∏–æ–Ω–Ω—ã–µ –¥–µ—Ä–µ–≤—å—è
–í–æ—Å–ø–æ–ª—å–∑—É–µ–º—Å—è –Ω–∞–±–æ—Ä–æ–º –¥–∞–Ω–Ω—ã—Ö Boston.

?Boston

# –æ–±—É—á–∞—é—â–∞—è –≤—ã–±–æ—Ä–∫–∞
set.seed(1)
train <- sample(1:nrow(Boston), nrow(Boston)/2) # –æ–±—É—á–∞—é—â–∞—è –≤—ã–±–æ—Ä–∫–∞ -- 50%
–ü–æ—Å—Ç—Ä–æ–∏–º –¥–µ—Ä–µ–≤–æ —Ä–µ–≥—Ä–µ—Å—Å–∏–∏ –¥–ª—è –∑–∞–≤–∏—Å–∏–º–æ–π –ø–µ—Ä–µ–º–µ–Ω–Ω–æ–π medv: –º–µ–¥–∏–∞–Ω–Ω–∞—è —Å—Ç–æ–∏–º–æ—Å—Ç–∏ –¥–æ–º–æ–≤, –≤ –∫–æ—Ç–æ—Ä—ã—Ö –∂–∏–≤—É—Ç —Å–æ–±—Å—Ç–≤–µ–Ω–Ω–∏–∫–∏ (—Ç—ã—Å. –¥–æ–ª–ª.).

# –æ–±—É—á–∞–µ–º –º–æ–¥–µ–ª—å
tree.boston <- tree(medv ~ ., Boston, subset = train)
summary(tree.boston)
## 
## Regression tree:
## tree(formula = medv ~ ., data = Boston, subset = train)
## Variables actually used in tree construction:
## [1] "lstat" "rm"    "dis"  
## Number of terminal nodes:  8 
## Residual mean deviance:  12.65 = 3099 / 245 
## Distribution of residuals:
##      Min.   1st Qu.    Median      Mean   3rd Qu.      Max. 
## -14.10000  -2.04200  -0.05357   0.00000   1.96000  12.60000
# –≤–∏–∑—É–∞–ª–∏–∑–∞—Ü–∏—è
plot(tree.boston)
text(tree.boston, pretty = 0)


–°–Ω–æ–≤–∞ —Å–¥–µ–ª–∞–µ–º –æ–±—Ä–µ–∑–∫—É –¥–µ—Ä–µ–≤–∞ –≤ —Ü–µ–ª—è—Ö —É–ª—É—á—à–µ–Ω–∏—è –∫–∞—á–µ—Å—Ç–≤–∞ –ø—Ä–æ–≥–Ω–æ–∑–∞.

cv.boston <- cv.tree(tree.boston)

# —Ä–∞–∑–º–µ—Ä –¥–µ—Ä–µ–≤–∞ —Å –º–∏–Ω–∏–º–∞–ª—å–Ω–æ–π –æ—à–∏–±–∫–æ–π
plot(cv.boston$size, cv.boston$dev, type = 'b')
opt.size <- cv.boston$size[cv.boston$dev == min(cv.boston$dev)]
abline(v = opt.size, col = 'red', 'lwd' = 2)     # —Å–æ–æ—Ç–≤. –≤–µ—Ä—Ç–∏–∫–∞–ª—å–Ω–∞—è –ø—Ä—è–º–∞—è
mtext(opt.size, at = opt.size, side = 1, col = 'red', line = 1)


–í –¥–∞–Ω–Ω–æ–º —Å–ª—É—á–∞–µ–º –º–∏–Ω–∏–º—É–º –æ—à–∏–±–∫–∏ —Å–æ–æ—Ç–≤–µ—Ç—Å—Ç–≤—É–µ—Ç —Å–∞–º–æ–º—É —Å–ª–æ–∂–Ω–æ–º—É –¥–µ—Ä–µ–≤—É, —Å 8 —É–∑–ª–∞–º–∏. –ü–æ–∫–∞–∂–µ–º, –∫–∞–∫ –ø—Ä–∏ –∂–µ–ª–∞–Ω–∏–∏ –º–æ–∂–Ω–æ –æ–±—Ä–µ–∑–∞—Ç—å –¥–µ—Ä–µ–≤–æ –¥–æ 7 —É–∑–ª–æ–≤ (–æ—à–∏–±–∫–∞ –Ω–µ–Ω–∞–º–Ω–æ–≥–æ –≤—ã—à–µ, —á–µ–º –º–∏–Ω–∏–º–∞–ª—å–Ω–∞—è).

# –¥–µ—Ä–µ–≤–æ —Å 7 —É–∑–ª–∞–º–∏
prune.boston = prune.tree(tree.boston, best = 7)

# –≤–∏–∑—É–∞–ª–∏–∑–∞—Ü–∏—è
plot(prune.boston)
text(prune.boston, pretty = 0)


–ü—Ä–æ–≥–Ω–æ–∑ —Å–¥–µ–ª–∞–µ–º –ø–æ –Ω–µ–æ–±—Ä–µ–∑–∞–Ω–Ω–æ–º—É –¥–µ—Ä–µ–≤—É, —Ç.–∫. —Ç–∞–º –æ—à–∏–±–∫–∞, –æ—Ü–µ–Ω–µ–Ω–Ω–∞—è –ø–æ –º–µ—Ç–æ–¥—É –ø–µ—Ä–µ–∫—Ä—ë—Å—Ç–Ω–æ–π –ø—Ä–æ–≤–µ—Ä–∫–∏, –º–∏–Ω–∏–º–∞–ª—å–Ω–∞.

# –ø—Ä–æ–≥–Ω–æ–∑ –ø–æ –ª—É—á—à–µ–π –º–æ–¥–µ–ª–∏ (8 —É–∑–ª–æ–≤)
yhat <- predict(tree.boston, newdata = Boston[-train, ])
boston.test <- Boston[-train, "medv"]

# –≥—Ä–∞—Ñ–∏–∫ "–ø—Ä–æ–≥–Ω–æ–∑ -- —Ä–µ–∞–ª–∏–∑–∞—Ü–∏—è"
plot(yhat, boston.test)
# –ª–∏–Ω–∏—è –∏–¥–µ–∞–ª—å–Ω–æ–≥–æ –ø—Ä–æ–≥–Ω–æ–∑–∞
abline(0, 1)


# MSE –Ω–∞ —Ç–µ—Å—Ç–æ–≤–æ–π –≤—ã–±–æ—Ä–∫–µ
mse.test <- mean((yhat - boston.test)^2)
MSE –Ω–∞ —Ç–µ—Å—Ç–æ–≤–æ–π –≤—ã–±–æ—Ä–∫–µ —Ä–∞–≤–Ω–∞ 25.05 (—Ç—ã—Å.–¥–æ–ª–ª.).

–ë—ç–≥–≥–∏–Ω–≥ –∏ –º–µ—Ç–æ–¥ —Å–ª—É—á–∞–π–Ω–æ–≥–æ –ª–µ—Å–∞
–†–∞—Å—Å–º–æ—Ç—Ä–∏–º –±–æ–ª–µ–µ —Å–ª–æ–∂–Ω—ã–µ –º–µ—Ç–æ–¥—ã —É–ª—É—á—à–µ–Ω–∏—è –∫–∞—á–µ—Å—Ç–≤–∞ –¥–µ—Ä–µ–≤–∞. –ë—ç–≥–≥–∏–Ω–≥ ‚Äì —á–∞—Å—Ç–Ω—ã–π —Å–ª—É—á–∞–π —Å–ª—É—á–∞–π–Ω–æ–≥–æ –ª–µ—Å–∞ —Å m=p, –ø–æ—ç—Ç–æ–º—É –∏ —Ç–æ, –∏ –¥—Ä—É–≥–æ–µ –º–æ–∂–Ω–æ –ø–æ—Å—Ç—Ä–æ–∏—Ç—å —Ñ—É–Ω–∫—Ü–∏–µ–π randomForest().

–î–ª—è –Ω–∞—á–∞–ª–∞ –∏—Å–ø–æ–ª—å–∑—É–µ–º –±—ç–≥–≥–∏–Ω–≥, –ø—Ä–∏—á—ë–º –≤–æ–∑—å–º—ë–º –≤—Å–µ 13 –ø—Ä–µ–¥–∏–∫—Ç–æ—Ä–æ–≤ –Ω–∞ –∫–∞–∂–¥–æ–º —à–∞–≥–µ (–∞—Ä–≥—É–º–µ–Ω—Ç mtry).

# –±—ç–≥–≥–∏–Ω–≥ —Å 13 –ø—Ä–µ–¥–∏–∫—Ç–æ—Ä–∞–º–∏
set.seed(1)
bag.boston <- randomForest(medv ~ ., data = Boston, subset = train, 
                           mtry = 13, importance = TRUE)
bag.boston
## 
## Call:
##  randomForest(formula = medv ~ ., data = Boston, mtry = 13, importance = TRUE,      subset = train) 
##                Type of random forest: regression
##                      Number of trees: 500
## No. of variables tried at each split: 13
## 
##           Mean of squared residuals: 11.15723
##                     % Var explained: 86.49
# –ø—Ä–æ–≥–Ω–æ–∑
yhat.bag = predict(bag.boston, newdata = Boston[-train, ])

# –≥—Ä–∞—Ñ–∏–∫ "–ø—Ä–æ–≥–Ω–æ–∑ -- —Ä–µ–∞–ª–∏–∑–∞—Ü–∏—è"
plot(yhat.bag, boston.test)
# –ª–∏–Ω–∏—è –∏–¥–µ–∞–ª—å–Ω–æ–≥–æ –ø—Ä–æ–≥–Ω–æ–∑–∞
abline(0, 1)


# MSE –Ω–∞ —Ç–µ—Å—Ç–æ–≤–æ–π
mse.test <- mean((yhat.bag - boston.test)^2)
mse.test
## [1] 13.50808
–û—à–∏–±–∫–∞ –Ω–∞ —Ç–µ—Å—Ç–æ–≤–æ–π –≤—ã–±–æ—Ä–∫–µ —Ä–∞–≤–Ω–∞ 13.51.
–ú–æ–∂–Ω–æ –∏–∑–º–µ–Ω–∏—Ç—å —á–∏—Å–ª–æ –¥–µ—Ä–µ–≤—å–µ–≤ —Å –ø–æ–º–æ—â—å—é –∞—Ä–≥—É–º–µ–Ω—Ç–∞ ntree.

bag.boston <- randomForest(medv ~ ., data = Boston, subset = train,
                           mtry = 13, ntree = 25)

# –ø—Ä–æ–≥–Ω–æ–∑
yhat.bag <- predict(bag.boston, newdata = Boston[-train, ])

# MSE –Ω–∞ —Ç–µ—Å—Ç–æ–≤–æ–π
mse.test <- mean((yhat.bag - boston.test)^2)
mse.test
## [1] 13.94835
–ù–æ, –∫–∞–∫ –≤–∏–¥–Ω–æ, —ç—Ç–æ —Ç–æ–ª—å–∫–æ —É—Ö—É–¥—à–∞–µ—Ç –ø—Ä–æ–≥–Ω–æ–∑.
–¢–µ–ø–µ—Ä—å –ø–æ–ø—Ä–æ–±—É–µ–º –≤—ã—Ä–∞—Å—Ç–∏—Ç—å —Å–ª—É—á–∞–π–Ω—ã–π –ª–µ—Å. –ë–µ—Ä—ë–º 6 –ø—Ä–µ–¥–∏–∫—Ç–æ—Ä–æ–≤ –Ω–∞ –∫–∞–∂–¥–æ–º —à–∞–≥–µ.

# –æ–±—É—á–∞–µ–º –º–æ–¥–µ–ª—å
set.seed(1)
rf.boston <- randomForest(medv ~ ., data = Boston, subset = train,
                          mtry = 6, importance = TRUE)

# –ø—Ä–æ–≥–Ω–æ–∑
yhat.rf <- predict(rf.boston, newdata = Boston[-train, ])

# MSE –Ω–∞ —Ç–µ—Å—Ç–æ–≤–æ–π –≤—ã–±–æ—Ä–∫–µ
mse.test <- mean((yhat.rf - boston.test)^2)

# –≤–∞–∂–Ω–æ—Å—Ç—å –ø—Ä–µ–¥–∏–∫—Ç–æ—Ä–æ–≤
importance(rf.boston)  # –æ—Ü–µ–Ω–∫–∏ 
##           %IncMSE IncNodePurity
## crim    12.132320     986.50338
## zn       1.955579      57.96945
## indus    9.069302     882.78261
## chas     2.210835      45.22941
## nox     11.104823    1044.33776
## rm      31.784033    6359.31971
## age     10.962684     516.82969
## dis     15.015236    1224.11605
## rad      4.118011      95.94586
## tax      8.587932     502.96719
## ptratio 12.503896     830.77523
## black    6.702609     341.30361
## lstat   30.695224    7505.73936
varImpPlot(rf.boston)  # –≥—Ä–∞—Ñ–∏–∫–∏


–û—à–∏–±–∫–∞ –ø–æ –º–æ–¥–µ–ª–∏ —Å–ª—É—á–∞–π–Ω–æ–≥–æ –ª–µ—Å–∞ —Ä–∞–≤–Ω–∞ 11.66, —á—Ç–æ –Ω–∏–∂–µ, —á–µ–º –¥–ª—è –±—ç–≥–≥–∏–Ω–≥–∞.

–ë—É—Å—Ç–∏–Ω–≥
–ü–æ—Å—Ç—Ä–æ–∏–º 5000 —Ä–µ–≥—Ä–µ—Å—Å–∏–æ–Ω–Ω—ã—Ö –¥–µ—Ä–µ–≤—å–µ–≤ —Å –≥–ª—É–±–∏–Ω–æ–π 4.

set.seed(1)
boost.boston <- gbm(medv ~ ., data = Boston[train, ], distribution = "gaussian",
                    n.trees = 5000, interaction.depth = 4)
# –≥—Ä–∞—Ñ–∏–∫ –∏ —Ç–∞–±–ª–∏—Ü–∞ –æ—Ç–Ω–æ—Å–∏—Ç–µ–ª—å–Ω–æ–π –≤–∞–∂–Ω–æ—Å—Ç–∏ –ø–µ—Ä–µ–º–µ–Ω–Ω—ã—Ö
summary(boost.boston)

# –≥—Ä–∞—Ñ–∏–∫–∏ —á–∞—Å—Ç–Ω–æ–π –∑–∞–≤–∏—Å–∏–º–æ—Å—Ç–∏ –¥–ª—è –¥–≤—É—Ö –Ω–∞–∏–±–æ–ª–µ–µ –≤–∞–∂–Ω—ã—Ö –ø—Ä–µ–¥–∏–∫—Ç–æ—Ä–æ–≤
par(mfrow = c(1, 2))
plot(boost.boston, i = "rm")
plot(boost.boston, i = "lstat")


# –ø—Ä–æ–≥–Ω–æ–∑
yhat.boost <- predict(boost.boston, newdata = Boston[-train, ], n.trees = 5000)

# MSE –Ω–∞ —Ç–µ—Å—Ç–æ–≤–æ–π
mse.test <- mean((yhat.boost - boston.test)^2)
mse.test
## [1] 11.84434
–ù–∞—Å—Ç—Ä–æ–π–∫—É –±—É—Å—Ç–∏–Ω–≥–∞ –º–æ–∂–Ω–æ –¥–µ–ª–∞—Ç—å —Å –ø–æ–º–æ—â—å—é –≥–∏–ø–µ—Ä–ø–∞—Ä–∞–º–µ—Ç—Ä–∞ Œª (–∞—Ä–≥—É–º–µ–Ω—Ç shrinkage). –£—Å—Ç–∞–Ω–æ–≤–∏–º –µ–≥–æ —Ä–∞–≤–Ω—ã–º 0.2.

# –º–µ–Ω—è–µ–º –∑–Ω–∞—á–µ–Ω–∏–µ –≥–∏–ø–µ—Ä–ø–∞—Ä–∞–º–µ—Ç—Ä–∞ (lambda) –Ω–∞ 0.2 -- –∞—Ä–≥—É–º–µ–Ω—Ç shrinkage
boost.boston <- gbm(medv ~ ., data = Boston[train, ], distribution = "gaussian",
                    n.trees = 5000, interaction.depth = 4, 
                    shrinkage = 0.2, verbose = F)
# –ø—Ä–æ–≥–Ω–æ–∑
yhat.boost <- predict(boost.boston, newdata = Boston[-train, ], n.trees = 5000)

# MSE –∞ —Ç–µ—Å—Ç–æ–≤–æ–π
mse.test <- mean((yhat.boost - boston.test)^2)
mse.test
## [1] 11.51109
–¢–∞–∫–∏–º –æ–±—Ä–∞–∑–æ–º, –∏–∑–º–µ–Ω–∏–≤ –≥–∏–ø–µ—Ä–ø–∞—Ä–∞–º–µ—Ç—Ä, –º—ã –µ—â—ë –Ω–µ–º–Ω–æ–≥–æ —Å–Ω–∏–∑–∏–ª–∏ –æ—à–∏–±–∫—É –ø—Ä–æ–≥–Ω–æ–∑–∞.
```

<!--chapter:end:24_nonlinear_regression.Rmd-->

# Multiple linear regression

Source: –ê–Ω–∞–ª–∏–∑ –¥–∞–Ω–Ω—ã—Ö –≤ R. –ú–Ω–æ–∂–µ—Å—Ç–≤–µ–Ω–Ω–∞—è –ª–∏–Ω–µ–π–Ω–∞—è —Ä–µ–≥—Ä–µ—Å—Å–∏—è  

```{r, eval=FALSE}
# Dataset swiss
?swiss
swiss <- data.frame(swiss)
str(swiss)

# Histogram of fertility
hist(swiss$Fertility, col='red')

# Numeric predictors for Fertility prediction
fit <- lm(Fertility ~ Examination + Catholic, data = swiss)
summary(fit)
# the principal predictor is an 'examination' with negative correlation.

# Interaction of variables 'examination' and 'catholics' '*'
fit2 <- lm(Fertility ~ Examination*Catholic, data = swiss)
summary(fit2)

confint(fit2)

# Categorical predictors
# Histogram obviously have two parts -> we can split data for two factors
hist(swiss$Catholic, col = 'red')

# Lets split 'Catholics' for two groups: with many 'lots' and few 'few'
swiss$religious <- ifelse(swiss$Catholic > 60, 'Lots', 'Few')
swiss$religious <- as.factor(swiss$religious)

fit3 <- lm(Fertility ~ Examination + religious, data = swiss)
summary(fit3)

# Interaction of variables
fit4 <- lm(Fertility ~ religious*Examination, data = swiss)
summary(fit4)

# plots

ggplot(swiss, aes(x = Examination, y = Fertility)) + 
  geom_point() 

ggplot(swiss, aes(x = Examination, y = Fertility)) + 
  geom_point() + 
  geom_smooth()

ggplot(swiss, aes(x = Examination, y = Fertility)) + 
  geom_point() + 
  geom_smooth(method = 'lm')

ggplot(swiss, aes(x = Examination, y = Fertility, col = religious)) + 
  geom_point() 

ggplot(swiss, aes(x = Examination, y = Fertility, col = religious)) + 
  geom_point()  + 
  geom_smooth()

ggplot(swiss, aes(x = Examination, y = Fertility, col = religious)) + 
  geom_point()  + 
  geom_smooth(method = 'lm')


#

fit5 <- lm(Fertility ~ religious*Infant.Mortality*Examination, data = swiss)
summary(fit5)


# model comparison

rm(swiss)
swiss <- data.frame(swiss)

fit_full <- lm(Fertility ~ ., data = swiss)
summary(fit_full)

fit_reduced1 <- lm(Fertility ~ Infant.Mortality + Examination + Catholic + Education, data = swiss)
summary(fit_reduced1)

anova(fit_full, fit_reduced1)

fit_reduced2 <- lm(Fertility ~ Infant.Mortality + Education + Catholic + Agriculture, data = swiss)
summary(fit_reduced2)

anova(fit_full, fit_reduced2)

# model selection

optimal_fit <-  step(fit_full, direction = 'backward')
summary(optimal_fit)
```

<!--chapter:end:26_multiple_regression.Rmd-->

# Spline model

In this example we will generate data from a given function and then build a model using splines and estimate quality of the model.

## Generate dataset from a given function

```{R}
# parameters to generate a dataset
n.all <- 100             # number of observations
train.percent <- 0.85    # portion of the data for training
res.sd <- 1              # standard deviation of noise
x.min <- 5               # min limit of the data
x.max <- 105             # max limit of the data

# generate x
set.seed(1)       # to get reproducible results by randomizer
x <- runif(x.min, x.max, n = n.all)

# noise from normal destibution
set.seed(1)
res <- rnorm(mean = 0, sd = res.sd, n = n.all)

# generate y using a given function
y.func <- function(x) {4 - 2e-02*x + 5.5e-03*x^2 - 4.9e-05*x^3}

# add noise
y <- y.func(x) + res
```

## Split data for train and test

```{R}
# split dataset for training and test
set.seed(1)
# generate vector of chosen x for train data
inTrain <- sample(seq_along(x), size = train.percent*n.all)

# train data set
x.train <- x[inTrain]
y.train <- y[inTrain]

# test data set
x.test <- x[-inTrain]
y.test <- y[-inTrain]
```

## Diagram of the given function and generated datasets
```{R}
# lines of generated data for plot
x.line <- seq(x.min, x.max, length = n.all)
y.line <- y.func(x.line)

# PLOT
# generate plot by train data
par(mar = c(4, 4, 1, 1)) # reduce margins (optional)
plot(x.train, y.train,
     main = 'Generated data and original function',
     col = grey(0.2), bg = grey(0.2), pch = 21,
     xlab = 'X', ylab = 'Y', 
     xlim = c(x.min, x.max),
     ylim = c(min(y), max(y)), 
     cex = 1.2, cex.lab = 1.2, cex.axis = 1.2)

# add points of test data
points(x.test, y.test, col = 'red', bg = 'red', pch = 21)

# add the given function
lines(x.line, y.line, lwd = 2, lty = 2)

# add legend
legend('topleft', legend = c('train', 'test', 'f(X)'),
       pch = c(16, 16, NA), 
       col = c(grey(0.2), 'red', 'black'),  
       lty = c(0, 0, 2), lwd = c(1, 1, 2), cex = 1.2)
```

## Build a model using splines
We will compair sevaral models with degree of freedoms (df) from 2 to 40, where 2 correspond to a linear model.

```{r}
max.df <- 40                       # max degree of freedom (df)
# 
tbl <- data.frame(df = 2:max.df)   # data frame for writing errors
tbl$MSE.train <- 0                 # column 1: errors of train data
tbl$MSE.test <- 0                  # —Åcolumn 2: errors of test data

# generate models using for cycle
for (i in 2:max.df) {
    mod <- smooth.spline(x = x.train, y = y.train, df = i)
    
    # predicted values for train and test data using built model
    y.model.train <- predict(mod, data.frame(x = x.train))$y[, 1]
    y.model.test <- predict(mod, data.frame(x = x.test))$y[, 1]
    
    # MSE errors for train and test data
    MSE <- c(sum((y.train - y.model.train)^2) / length(x.train),
             sum((y.test - y.model.test)^2) / length(x.test))
    
    # write errors to the previously created data frame
    tbl[tbl$df == i, c('MSE.train', 'MSE.test')] <- MSE
}

# view first rows of the table
head(tbl, 4)
```

## Diagram of MSE for train and test data
```{r}
# plot MSE from our table
plot(x = tbl$df, y = tbl$MSE.test,
     main = "Changes of MSE from degrees of freedom",
     type = 'l', col = 'red', lwd = 2,
     xlab = 'spline degree of freedom', ylab = 'MSE',
     ylim = c(min(tbl$MSE.train, tbl$MSE.test), 
              max(tbl$MSE.train, tbl$MSE.test)),
     cex = 1.2, cex.lab = 1.2, cex.axis = 1.2)

# add 
points(x = tbl$df, y = tbl$MSE.test,
       pch = 21, col = 'red', bg = 'red')
lines(x = tbl$df, y = tbl$MSE.train, col = grey(0.3), lwd = 2)
# minimal MSE
abline(h = res.sd, lty = 2, col = grey(0.4), lwd = 2)

# add legend
legend('topright', legend = c('train', 'test'),
       pch = c(NA, 16), 
       col = c(grey(0.2), 'red'),  
       lty = c(1, 1), lwd = c(2, 2), cex = 1.2)

# df of minimal MSE for test data
min.MSE.test <- min(tbl$MSE.test)
df.min.MSE.test <- tbl[tbl$MSE.test == min.MSE.test, 'df']

# optimal df for precise model and maximal simplicity
df.my.MSE.test <- 6
my.MSE.test <- tbl[tbl$df == df.my.MSE.test, 'MSE.test']

# show the optimal solution
abline(v = df.my.MSE.test, 
       lty = 2, lwd = 2)
points(x = df.my.MSE.test, y = my.MSE.test, 
       pch = 15, col = 'blue')
mtext(df.my.MSE.test, 
      side = 1, line = -1, at = df.my.MSE.test, col = 'blue', cex = 1.2)
```

## Build optimal model and plot for the model
```{r}
mod.MSE.test <- smooth.spline(x = x.train, y = y.train, df = df.my.MSE.test)

# predict data for 250 x's to get smoothed curve
x.model.plot <- seq(x.min, x.max, length = 250)
y.model.plot <- predict(mod.MSE.test, data.frame(x = x.model.plot))$y[, 1]

# plot train data
par(mar = c(4, 4, 1, 1))
plot(x.train, y.train,
     main = "Initial data and the best fit model",
     col = grey(0.2), bg = grey(0.2), pch = 21,
     xlab = 'X', ylab = 'Y', 
     xlim = c(x.min, x.max),
     ylim = c(min(y), max(y)), 
     cex = 1.2, cex.lab = 1.2, cex.axis = 1.2)

# add test data
points(x.test, y.test, col = 'red', bg = 'red', pch = 21)

# function
lines(x.line, y.line,lwd = 2, lty = 2)

# add model
lines(x.model.plot, y.model.plot, lwd = 2, col = 'blue')

# legend
legend('topleft', legend = c('train', 'test', 'f(X)', 'model'),
       pch = c(16, 16, NA, NA), 
       col = c(grey(0.2), 'red', 'black', 'blue'),  
       lty = c(0, 0, 2, 1), lwd = c(1, 1, 2, 2), cex = 1.2)
```

## Bibliograpy
[An Introduction to Statistical Learning by Gareth James](http://faculty.marshall.usc.edu/gareth-james/)

<!--chapter:end:27_spline_model.Rmd-->

# Logistic Regression

The **logistic model** (or **logit model**) is a statistical model with input (independent variable) 
a continuous variable and output (dependent variable) a binary variable (discret choice, e.g. yes/no or 1/0).

## Confusion matrix  
A confusion matrix is a table that is often used to describe the performance of a classification model (or "classifier") on a set of test data for which the true values are known.  

| n=165       | Predicted: NO | Predicted: Yes-|
|-------------|---------------|----------------|
| Actual: No  | TN = 50       | FP = 10        |
| Actual: Yes | FN = 5        | TP = 100       |

TN - true negatives  
TP - true positives  
FN - false negatives  
FP - false posistives  

**Accuracy** - Overall, how often is the classifier correct?  
(TP+TN)/total = (100+50)/165 = 0.91

**Misclassification Rate** - Overall, how often is it wrong?  
(FP+FN)/total = (10+5)/165 = 0.09
equivalent to 1 - Accuracy
also known as "Error Rate"

**True Positive Rate**: When it's actually yes, how often does it predict yes?
TP/actual yes = 100/105 = 0.95
also known as "Sensitivity" or "Recall"

**False Positive Rate**: When it's actually no, how often does it predict yes?
FP/actual no = 10/60 = 0.17

**True Negative Rate**: When it's actually no, how often does it predict no?
TN/actual no = 50/60 = 0.83
equivalent to 1 minus False Positive Rate
also known as "Specificity"

**Precision**: When it predicts yes, how often is it correct?
TP/predicted yes = 100/110 = 0.91

**Prevalence**: How often does the yes condition actually occur in our sample?
actual yes/total = 105/165 = 0.64

**Null Error Rate**: This is how often you would be wrong if you always predicted the majority class. (In our example, the null error rate would be 60/165=0.36 because if you always predicted yes, you would only be wrong for the 60 "no" cases.) This can be a useful baseline metric to compare your classifier against. However, the best classifier for a particular application will sometimes have a higher error rate than the null error rate, as demonstrated by the Accuracy Paradox.  
**Cohen's Kappa**: This is essentially a measure of how well the classifier performed as compared to how well it would have performed simply by chance. In other words, a model will have a high Kappa score if there is a big difference between the accuracy and the null error rate. (More details about Cohen's Kappa.)  
**F Score**: This is a weighted average of the true positive rate (recall) and precision. (More details about the F Score.)  
**ROC Curve**: This is a commonly used graph that summarizes the performance of a classifier over all possible thresholds. It is generated by plotting the True Positive Rate (y-axis) against the False Positive Rate (x-axis) as you vary the threshold for assigning observations to a given class. (More details about ROC Curves.) 


**Example**:  
A group of 20 students spend between 0 and 6 hours studying for an exam. How does the number of hours spent studying affect the probability that the student will pass the exam?  

```{r}
hours <- c(0.5, 0.75, 1, 1.25, 1.5, 1.75, 1.75, 2, 2.25, 2.5, 2.75, 3, 3.25, 3.5, 4, 4.25, 4.5, 4.75, 5, 5.5)
pass <- c(0,0,0,0,0,0,1,0,1,0,1,0,1,0,1,1,1,1,1,1)
model = glm(pass ~ hours, family = binomial)
newdat <- data.frame(hours=seq(min(hours), max(hours),len=100))
newdat$pass = predict(model, newdata=newdat, type="response")

# plot
plot(pass ~ hours)
lines(pass ~ hours, newdat, col="red")

# data
data <- data.frame(hours=c(0.5, 0.75, 1, 1.25, 1.5, 1.75, 1.75, 2, 2.25, 2.5,
                         2.75, 3, 3.25, 3.5, 4, 4.25, 4.5, 4.75, 5, 5.5),
                 pass=c(0,0,0,0,0,0,1,0,1,0,1,0,1,0,1,1,1,1,1,1),
                 pass.predic = rep(NA, 20), # slot for predicted pass
                 pass.logit = rep(NA, 20))  # slot logit prediction
                 

# model
model <- glm(data$pass ~ data$hours, family = binomial)

# predict values of logit function
data$pass.logit <- predict(model, newdata=data, type='response')
# predict pass for threshold = 0.5
data$pass.predic <- ifelse(data$pass.logit > 0.5, 1, 0)

# Confusion matrix
library(caret)
caret::confusionMatrix(data =      factor(data$pass.predic),
                       reference = factor(data$pass))

```


```{r}
p <- seq(0,1, by=0.05)
data <- data.frame(probability=p, odds=p/(1-p))
head(data)
plot(data$odds~data$probability, type='o', pch=19, xlab='Probability', ylab='Odds')

plot(log(data$odds)~data$odds, type='o', pch=19, xlab='Odds', ylab='log(odds)')

plot(data$probability~log(data$odds), type='o', pch=19, xlab='log(odds)', ylab='Probability')

```

## Next part

```{r, eval=FALSE}
library(data.table)
df <- fread('https://raw.githubusercontent.com/suvarzz/data/master/data_classification.csv', header=T, sep=",")
head(df)

plot(df[pass==1][,!3], col='red')
points(df[pass==0][,!3], col='blue')

model.logit <- glm(pass ~ studied + slept, data = df, family = 'binomial')
summary(model.logit)
p.lda <- predict(model.logit, df, type = 'response')
df$predicted <- ifelse(p.lda > 0.5, 1, 0)
head(df)

a=-coef(model.logit)[1]/coef(model.logit)[2],
b=-coef(model.logit)[1]/coef(model.logit)[3])

b0 = coef(model.logit)[1]
b1 = mymodel$coefficients[[2]]
b2 = mymodel$coefficients[[3]]
z = b0 + (b1 * 1) + (b2 * 4)
p = 1 / (1 + exp(-z))

if p=0.5 => z = 0 => b0 + b1*x + b2*y => 
  
segments(0,10.87,9.26,0)
slept=(3.77-0.474*studied)/0.338
(0, 3.77/0.474) = (9.2, 0)
(3.77/0.474,0) = (0, 10.87)

segments(9.2,0, 0,10.87, lwd=2)
```

## NExt part
```{r, eval=FALSE}
# Example of logistic regression
# Source: 17 - –ê–Ω–∞–ª–∏–∑ –¥–∞–Ω–Ω—ã—Ö –≤ R. –õ–æ–≥–∏—Å—Ç–∏—á–µ—Å–∫–∞—è —Ä–µ–≥—Ä–µ—Å—Å–∏—è by Anatoliy Karpov

# Read data set train.csv:
# Statistics of students in a school
# gender - male/femail
# read, write, math - points for subjects
# hon - if honorary degree Y/N

# FIX combine train and test into one csv file. Split train and test inside this script
setwd("~/RData")
df <- read.csv("train.csv", sep=";")

# Visual inspection of the dataset
head(df)
str(df)
View(df)

# N-not the best mark, Y-the best mark
library(ggplot2)
ggplot(df, aes(read,math,col=gender))+geom_point()+facet_grid(.~hon)+
    theme(axis.text=element_text(size=25), axis.title=element_text(size=25, face='bold'))

# Apply logistic regression
# How hon depends on different variables: read, math, gender
fit <- glm(hon ~ read + math + gender, df, family = "binomial")
summary(fit)

# Meanings of coefficients:
# read-estimate: 0.06677 - if female, math is fixed, if read change to 1, then ln(odds) will be changed to 0.06677

# Get data from fit
exp(fit$coefficients)

# Predict model - ln(odds)
head(predict(object=fit))

# Predict model - return probability to get the best mark for every person
head(predict(object = fit, type = "response"))

# Add probabilities to get the best mark for every person in df
df$prob <- predict(object = fit, type = "response")

df

# Part 2
# ROC-curve of predicted model
library(ROCR)
# Predicted values and real values
pred_fit <- prediction(df$prob, df$hon)

# Calculate tpr - true positive rate and fpr - false positive rate
perf_fit <- performance(pred_fit, "tpr", "fpr")

# plot ROC-curve
plot(perf_fit, colorize=T, print.cutoffs.at = seq(0,1, by=0.1))

# Area under the curve: 0.87
auc <- performance(pred_fit, measure = "auc")
str(auc)

# How to detect the border and make a decision if student will get honorary degree
# Specificity - how good we can predict negative results
perf3 <- performance(pred_fit, x.measure = "cutoff", measure = "spec")
# Sencitivity - how good we can predict positive results
perf4 <- performance(pred_fit, x.measure = "cutoff", measure = "sens")
# –û–±—â–∞—è –∏–Ω—Ç–µ–Ω—Å–∏–≤–Ω–æ—Å—Ç—å –∫–ª–∞—Å—Å–∏—Ñ–∏–∫–∞—Ç–æ—Ä–∞
perf5 <- performance(pred_fit, x.measure = "cutoff", measure = "acc")

plot(perf3, col = 'red', lwd = 2)
plot(add=T, perf4, col = "green", lwd = 2)
plot(add=T, perf5, lwd = 2)

legend(x = 0.6, y = 0.3, c("spec", "sens", "accur"), lty=1, col=c("red", "green", "black"),
       bty='n', cex=1, lwd=2)
abline(v=0.255, lwd=2)
# The border is the intersection of all three curves

# Add column with prediced values Y/N
df$pred_resp <- factor(ifelse(df$prob > 0.255, 1, 0), labels=c("N", "Y"))

# 1 if prediction is correct, 0 if not correct
df$correct <- ifelse(df$pred_resp == df$hon, 1, 0)
df

# blue - correct classified, red - incorrect classified
# it is more difficult to predict positive result
ggplot(df, aes(prob, fill = factor(correct)))+
    geom_dotplot()+
    theme(axis.text=element_text(size=25),
          axis.title=element_text(size=25, face="bold"))

# Percent of positive predictions
mean(df$correct)

# Part 3 - Prediction using test data
test_df <- read.csv("test.csv", sep=";")
test_df

# Predict honorary members
test_df$prob_predict <- predict(fit, newdata=test_df, type="response")
test_df$pred_resp <- factor(ifelse(test_df$prob_predict > 0.255, 1, 0), labels=c("N", "Y"))
test_df
```

<!--chapter:end:28_logistic_regression.Rmd-->

# Clustering

5 classes of clustering methods:  
1. **Partitioning methods** - split into k-groups (k-means, k-dedoids (PAM), CLARA)  
2. **Hierarchical clustering**  
3. **Fuzzy clustering**  
4. **Density-based clustering**  
5. **Model-based clustering**  

```{r, eval=FALSE}

bv <- read.table("./DATA/beverage.csv", header=T, sep=";")  
head(bv)  
# no needs to normalize because all data is binary (0,1)

# Hierarchical clustering
# dist - calculate distances
# hclust - hierarchical clustering

clust.bv <- hclust(dist(bv[,2:9]), "ward.D")
clust.bv

# Plot clusters
plot(clust.bv)
plot(clust.bv, hang = -1)
rect.hclust(clust.bv, k=3, border="red") 

# Group data by clusters
groups <- cutree(clust.bv, k=3)
groups

# Percentage in broups by drinking different beverages
colMeans(bv[groups==1, 2:9])*100
colMeans(bv[groups==2, 2:9])*100
colMeans(bv[groups==3, 2:9])*100

# Interpretation
# 1. People who does not have specific preference
# 2. People who prefers cola and pepsi
# 3. Not clear (others)

# atributes of cluster analysis
names(clust.bv)

# chronic of combining
clust.bv$merge
clust.bv[1]

clust.bv$height
clust.bv$order
clust.bv$labels
clust.bv$method
clust.bv$call
clust.bv$dist.method

# Detect the best choice for number of cluster by elbow-plot
plot(1:33, clust.bv$height, type="l")

### Task. Analyse data and find groups of people
# Scores (0,10) of 10 tests for candidates to get a job.
# 1. Memorizing numbers
# 2. Math task
# 3. Solving tasks in dialoge
# 4. Algorithms
# 5. Self confidence
# 6. Work in group
# 7. Find solution
# 8. Collaboration
# 9. Acceptance by others

setwd("~/DataAnalysis")
job <- read.table("DATA/assess.dat", header=T, sep="\t")
job

# Clustering
clust.job <- hclust(dist(job[,3:ncol(job)]), "ward.D")
# no needs to normalize, because all numbers have the same min, max
plot(clust.job)  # visual number of clusters is 4

# Group data by clusters
groups <- cutree(clust.job, k=4)
groups
colMeans(job[groups==1, 3:12])*100

### Find clusters using k-means method

setwd("~/DataAnalysis")
bv <- read.table("DATA/beverage.csv", header=T, sep=";")
bv
dim(bv)
names(bv)

# k-means clustering, with initial 3 clusters
# nstart = x - run x times with different initial clusters
summ.1 = kmeans(bv[,2:9], 3, iter.max = 100)
names(summ.1)

# Objects by clusters
summ.1$cluster

# Centers of clusters
summ.1$centers
# 2 digits after point
options(digits=2)

t(summ.1$centers)
options(digits=7)

# Square summs
summ.1$withinss

# Summ of elements of vector
summ.1$tot.withinss

# sum(33*(apply(bv[,2:9], 2, sd))^2)
summ.1$totss
summ.1$tot.betweenss

# Size of clusters
summ.1$size

# Elbow plot to detect optimal number of clusters
wss <- (nrow(bv[,2:9])-1)*sum(apply(bv[,2:9],2,var))
for (i in 2:15) { wss[i] <- kmeans(bv[,2:9],
                centers=i)$tot.withinss }
plot(1:15, wss, type="b", xlab="Number of Clusters",
     ylab="Within groups sum of squares")

# We can see that diagram is rough. This is because clusters are not allways optimal
# To improve situation, we have to run many initiall start coordinates and choose the best
# option (add nstart=500):
wss <- (nrow(bv[,2:9])-1)*sum(apply(bv[,2:9],2,var))
for (i in 2:15) { wss[i] <- kmeans(bv[,2:9],
                                   centers=i, nstart=500)$tot.withinss }
plot(1:15, wss, type="b", xlab="Number of Clusters",
     ylab="Within groups sum of squares")
# Warnings means that iterations were not finished for some cases.

# Let's compair results for 3 and 4 clusters
summ.1 = kmeans(bv[,2:9], 3, iter.max=100)
summ.2 = kmeans(bv[,2:9], 4, iter.max=100)

# Compair clusters. How many elements in each cluster
# We can see how elements move if we take more clusters
table(summ.1$cluster, summ.2$cluster)

# Multidimentional scaling
# Project multidimentional data to 2d
bv.dist <- dist(bv[,2:9])
bv.mds <- cmdscale(bv.dist)
plot(bv.mds, col = summ.1$cluster, xlab="Index", ylab="")

# Detect optimal number of clusters
install.packages("NbClust")
library("NbClust")
Best <- NbClust(bv[,2:9],              # data 
                distance="euclidean",  # distance method
                min.nc=2,              # min number of clusters
                max.nc=8,             # max number of clusters
                method="ward.D",       # ward methodes 
                index = "alllong" )    # choose indices
```
## Next part

```{r, eval=FALSE}
library(cluster)
library(factoextra)


Distances:
stats::dist()
factoextra::get_dist()   # compute a distance matrix between the rows of a data matrix
factoextra::fviz_dist()  # visualize distance matrix
cluster::daisy()         # handle both numeric and not numeric (nominal, ordinal,...) data types

d <- factoextra::get_dist(USArrests, stand = TRUE, method = 'pearson')
factoextra::fviz_dist(d, gradient = list(low='blue', mid='white', high='red'))

#####
library(tidyverse)
library(cluster)
library(factoextra)

data <- USArrests %>% na.omit() %>% scale()
data
factoextra::fviz_nbclust(data, kmeans, method = 'gap_stat')


km.res <- kmeans(data, 3, nstart = 25)
factoextra::fviz_cluster(km.res, data = data,
                         ellipse.type = 'convex',
                         palette = 'jco',
                         repel = TRUE,
                         ggtheme = theme_minimal())

# PAM clustering
pam.res <- cluster::pam(data, 4)
factoextra::fviz_cluster(pam.res)

# CLARA clustering
clara.res <- clara(df, 2, samples = 50, pamLike = TRUE)
clara.res

dd <- cbind(df, cluster = clara.res$cluster)

# Medoids
clara.res$medoids

# Clustering
head(clara.res$clustering,10)
```

## Example
```{r setup, include=TRUE}
library(datasets)
head(iris)

# Plot Petal.Length ~ Petal.Width data
plot(iris$Petal.Length ~ iris$Petal.Width)
set.seed(20)

# Find number of clusters using wss
wss <- (nrow(iris[, 3:4])-1)*sum(apply(iris[, 3:4],2,var))
for (i in 2:15) wss[i] <- sum(kmeans(iris[, 3:4], i)$withinss)
plot(1:15, wss, type="b", xlab="Number of Clusters", ylab="Within groups sum of squares")

#More than 3 clusters give no obvious advantages

# Make k-means with 3 clasters
ncl <- 3
irisCluster <- kmeans(iris[, 3:4], ncl, nstart = 20)
irisCluster

# Compair result of clustering with real data (3 species of iris are in analysis)
table(irisCluster$cluster, iris$Species)

# Plot data
clusters <- split.data.frame(iris, irisCluster$cluster)
xlim <- c(min(iris$Petal.Width), max(iris$Petal.Width))
ylim <- c(min(iris$Petal.Length), max(iris$Petal.Length))
col <- c('red', 'green', 'blue')
plot(0, xlab='Petal width', ylab='Petal length', xlim=xlim, ylim=ylim)
for ( i in 1:ncl ) {
points(clusters[[i]]$Petal.Length ~ clusters[[i]]$Petal.Width, col=col[i], xlim=xlim, ylim=ylim)
}
```

## NEXT PART

```{r, eval=FALSE}
# K-Nearest Neighbors or KNN is a clustering algorithm
# k is known number of clusters (usually sqrt(N), between 3-10, but may be different)
# samples must be normalized x = (x - min(x))/(max(x)-min(x))

head(iris)
summary(iris)   # detailed view of the data set
str(iris)   # view data types, sample values, categorical values, etc
plot(iris)

#normalization function

min_max_normalizer <- function(x)
{
    num <- x - min(x) 
    denom <- max(x) - min(x)
    return (num/denom)
}

#normalizing iris data set
normalized_iris <- as.data.frame(lapply(iris[1:4], min_max_normalizer))

#viewing normalized data
summary(normalized_iris)

#checking the data constituency
table(iris$Species)

#set seed for randomization
set.seed(1234)

# setting the training-test split to 67% and 33% respectively
random_samples <- sample(2, nrow(iris), replace=TRUE, prob=c(0.67, 0.33))

# training data set
iris.training <- iris[
    random_samples ==1, 1:4] 

#training labels
iris.trainLabels <- iris[
    random_samples ==1, 5]


# test data set
iris.test <- iris[
    random_samples ==2, 1:4]

#testing labels
iris.testLabels <- iris[
    random_samples ==2, 5]

#setting library
library(class)

#executing knn for k=3
iris_model <- knn(train = iris.training, test = iris.test, cl = iris.trainLabels, k=3)

#summary of the model learnt
iris_model
```

<!--chapter:end:30_clustering.Rmd-->

# Learning Vector Quantization

Learning Vector Quantiztion (LVQ) is a supervised classification algorithm for binary and multiclass problems. LVQ is a special case of a neural network.  
LVQ model creates codebook vectors by learning training dataset. Codebook vectors represent class regions. They contain elements that placed around the respective class according to their matching level. If the element matches, it comes closer to the target class, if it does not match, it moves farther from it. With this codebooks, the model classifies new data.
[Here](http://jsalatas.ictpro.gr/implementation-of-competitive-learning-networks-for-weka/) is a nice explanation how it works.

There are several versions of **LVQ** function:  
`lvq1()`, `olvq1()`, `lvq2()`, `lvq3()`, `dlvq()`.    

```{r}
library(class) # olvq1()
library(caret) # to split data

# generate dataset
df <- iris

id = caret::createDataPartition(df$Species, p = .8, list = F)

train = df[id, ]
test = df[-id, ]

# initialize an LVQ codebook
cb = class::lvqinit(train[1:4], train$Species)

# training set in a codebook.
build.cb = class::olvq1(train[1:4], train$Species, cb)

# classify test set from LVQ Codebook for test data
predict = class::lvqtest(build.cb, test[1:4])

# confusion matrix.
caret::confusionMatrix(test$Species, predict)
```

<!--chapter:end:41_learning_vector_quantization.Rmd-->

# Bayesian inference

**Bayesian data analysis** the use of Bayesian inference to learn from data.  

**Bayesian inference** is a method for figuring out unobservable quantities 
given known facts that uses probability to describe the uncertainty over what 
the values of the unknown quantities could be.  

**Bayesian inference** is  conditioning on data, in order to learn about parameter values.  

**Bayesian inference** is a method of statistical inference in which Bayes' 
theorem is used to update the probability for a hypothesis as more evidence or 
information becomes available.  

**Bayes' theorem**  
$$P( \theta | D ) = frac{P( D | \theta) \pdot P( \theta)}{\sum{P( D | \theta) \pdot P( \theta ) }}$$

$\theta$ - Parameter values  
$D$ - New data  
$P(D | \theta)$ - Likelihood (the relative) probability of the data given different parameter value.  
$P( \theta )$ - Prior  
$\sum{P( D | \theta) \pdot P( \theta ) }$ - The total sum of the likelihod weighted by the prior.  

- A **prior** is a probability distribution that represents what the model knows before seeing the data.  
- A **posterior** is a probability distribution that represents what the model knows after having seen the data.  

**Bayesian inference components**:  

- Data  
- Bayesian Model:  
+ Generative model (`dbinom`, `dnorm`, `poisson` etc.)  
+ Priors  
- Computational methods:  
+ Rejection sampling  
+ Grid approximation  
+ Markov Chain Monte Carlo (MCMC)  

```{r, message = FALSE}
# Visualization function of Bayesian inference for binary data 
library(ggjoy)
library(ggExtra)
library(ggjoy)
library(tidyverse)

# prop_model function to visualize bayesian posterior distributions
prop_model <- function(data = c(), prior_prop = c(1, 1), n_draws = 10000, show_plot = TRUE) 
{
    data <- as.logical(data)
    proportion_success <- c(0, seq(0, 1, length.out = 100), 1)
    data_indices <- round(seq(0, length(data), length.out = min(length(data) + 
                                                                    1, 20)))
    post_curves <- map_dfr(data_indices, function(i) {
        value <- ifelse(i == 0, "Prior", ifelse(data[i], "Success", "Failure"))
        label <- paste0("n=", i)
        probability <- dbeta(proportion_success, prior_prop[1] + 
                                 sum(data[seq_len(i)]), prior_prop[2] + sum(!data[seq_len(i)]))
        probability <- probability/max(probability)
        tibble(value, label, proportion_success, probability)
    })
    post_curves$label <- fct_rev(factor(post_curves$label, levels = paste0("n=", 
                                                                           data_indices)))
    post_curves$value <- factor(post_curves$value, levels = c("Prior", 
                                                              "Success", "Failure"))
    p <- ggplot(post_curves, aes(x = proportion_success, y = label, 
                                 height = probability, fill = value)) + geom_joy(stat = "identity", 
                                                                                 color = "white", alpha = 1, panel_scaling = TRUE, size = 1) + 
        scale_y_discrete("", expand = c(0.01, 0)) + scale_x_continuous("Underlying proportion of success") + 
        scale_fill_manual(values = hcl(120 * 2:0 + 15, 100, 65), 
                          name = "", drop = FALSE, labels = c("Prior   ", "Success   ", 
                                                              "Failure   ")) + theme_light(base_size = 18) + 
        theme(legend.position = "top")
    if (show_plot) {
        print(p)
    }
    invisible(rbeta(n_draws, prior_prop[1] + sum(data), prior_prop[2] + 
                        sum(!data)))
}
```

## Simple model with one binary parameter

```{r}
# Generate 50 random binary data with P(1)=0.75
data <- sample(c(1, 0), prob = c(0.75, 0.25), size = 10, replace = TRUE)

# Visualize posteriors
posterior <- prop_model(data)

# View posterior
head(posterior)

# Center of the posterior
median(posterior)

# The credible interval (CI)
quantile(posterior, c(0.05, 0.95))

# Probability of successes > 0.3
sum(posterior > 0.3) / length(posterior)
```

**Generate binomial data**  
```{r}
p <- 0.42     # probability of success
n <- 100      # number of observations
# Create logical vector, TRUE if numeric value from uniform distribution < p
data <- c()
for (i in 1:n) {
    data[i] <- runif(1, min = 0, max = 1) < p
}
# convert logical to numeric 0/1
data <- as.numeric(data)
data

# the same vector using rbinom distribution
n = 100      # number of observations
size = 1     # 0 - fail, 1 - success, for 1 trial
p = 0.42     # probability of success
rbinom(n, size, p)
```

## Grid approximation

To get more visitors to your website you are considering paying for an ad to be 
**shown 100 times** on a popular social media site.  
According to the social media site, their ads get **clicked on 10% of the time**. 
How many visitors could your site get?  

We can fix % of clicks as 0.1 but we do not know exactly how many clicks people 
do (uncertain). We suggest the % of clicks is uniformly distributed between 0 and 20.  

Define model:  
$n_{ads}  = 100$
$p_{clicks} ~ Uniform(0.0, 0.2)$
$p_{visitors} ~ Binomial(n_{ads}, p_{clicks})$

```{r}
n <- 100000          # number of simulations
n_ads_shown <- 100   # number of shown ads

# probability that add will be clicked (assume uniform [0,0.2])
p_clicks <- runif(n, min = 0.0, max = 0.2)  

# MODEL: We assume that the binomial distribution is a reasonable generative model
n_visitors <- rbinom(n = n, 
                     size = n_ads_shown, 
                     prob = p_clicks)

# Visualize the prior model simulation
library(ggplot2)
df <- data.frame(x = n_visitors, y = p_clicks)
p <- ggplot(df, aes(n_visitors, p_clicks)) + geom_point() + theme_classic()
ggExtra::ggMarginal(p, type = "histogram",  bins = 18)
```

**Generate posterior distribution**  

After the first day of showing ads we registered that **13 people** clicked and 
**visited** site when the ad was shown 100 times.  

```{r}
# Prior distribution binomial model with uniform clicks
prior <- data.frame(p_clicks, n_visitors)

# To get posterior we subset from prior 13 visitors
posterior <- prior[prior$n_visitors == 13, ]
head(posterior)

# visualize posterior
p <- ggplot(posterior, aes(n_visitors, p_clicks)) + geom_point() + theme_classic()
ggExtra::ggMarginal(p, type = "histogram",  bins = 18)
```

For the next iteration we make **prior from posterior** and include new data for 
subsetting from prior to get posterior.  
```{r}
# Assign posterior to prior for the next iteration
prior <- posterior

# next iteration posterior ~ f(prior)
n <-  nrow(prior)
n_ads_shown <- 100
prior$n_visitors <- rbinom(n, 
                           size = n_ads_shown, 
                           prob = prior$p_clicks)

# plot of prior distributions
p <- ggplot(prior, aes(n_visitors, p_clicks)) + geom_point() + theme_classic()
ggExtra::ggMarginal(p, type = "histogram",  bins = 18)
```

Use result of the model for prediction:  

```{r}
# Calculate the probability that you will get 5 or more visitors.
sum(prior$n_visitors >= 5) / length(prior$n_visitors)

# Median number of visitors
median(prior$n_visitors)
```

**Updated model**
We will change prior distribution with knowledge we learn from people that the 
number of clicks are 5% and sometimes 2 or 8%.
We will use **beta distribution** to setup number of clicks to update prior in 
our model.  

```{r}
n <- 100000
n_ads_shown <- 100

# Change the prior on proportion_clicks
p_clicks <- rbeta(n, shape1 = 5, shape2 = 95)

# Updated model
n_visitors <- rbinom(n = n,
                     size = n_ads_shown, 
                     prob = p_clicks)

prior <- data.frame(p_clicks, n_visitors)

posterior <- prior[prior$n_visitors == 13, ]

# Plots the prior and the posterior
par(mfcol = c(1, 2))
hist(prior$p_clicks, xlim = c(0, 0.25))
hist(posterior$p_clicks, xlim = c(0, 0.25))
```

We tested for video banner and now we would like to compair distributions for:
- video ads (13 out of 100 clicked)  
- text ads (6 out of a 100 clicked).  

```{r}
n <- 100000
n_ads_shown <- 100
p_clicks <- runif(n, min = 0.0, max = 0.2)

n_visitors <- rbinom(n, 
                     size = n_ads_shown, 
                     prob = p_clicks)
prior <- data.frame(p_clicks, n_visitors)

# Create the posteriors for video and text ads
posterior_video <- prior[prior$n_visitors == 13, ]
posterior_text <- prior[prior$n_visitors == 6, ]

par(mfrow=c(1,2))
# Visualize the posteriors
hist(posterior_video$p_clicks, xlim = c(0, 0.25))
hist(posterior_text$p_clicks, xlim = c(0, 0.25))

# combine data into dataframe
# Make sizes of distrib. the same (4000) to fit to dataframe
posterior <- data.frame(video_prop = posterior_video$p_clicks[1:4000],
                        text_prop = posterior_text$p_clicks[1:4000])

# Calculate the posterior difference: video_prop - text_prop
posterior$prop_diff <- posterior$video_prop - posterior$text_prop

# Calculate the median of prop_diff
median(posterior$prop_diff)

# Calculate the proportion
mean(posterior$prop_diff > 0.0)

# Visualize prop_diff
hist(posterior$prop_diff)
```

**Decision analysis**  

Our analysis indicated that *the video ads are clicked more often*.  
We would like to estimate probable profit *video ads* vs *text ads* if we know:  

- Each visitor spends $2.53 on average on our website.  
- Video ad costs us $0.25 per click.   
- Text ad costs us $0.05 per click.  

```{r}
visitor_spend <- 2.53
video_cost <- 0.25
text_cost <- 0.05

# Add the column posterior$video_profit
posterior$video_profit <- posterior$video_prop * visitor_spend - video_cost

# Add the column posterior$text_profit
posterior$text_profit <- posterior$text_prop * visitor_spend - text_cost

head(posterior)

# Visualize the video_profit and text_profit columns
par(mfrow=c(1,2))
hist(posterior$video_profit)
hist(posterior$text_profit)

# Difference between video and text ad profits
posterior$profit_diff <- posterior$video_profit - posterior$text_profit

# Visualize posterior$profit_diff
hist(posterior$profit_diff)

# Calculate a "best guess" for the difference in profits
median(posterior$profit_diff)

# Calculate the probability that text ads are better than video ads
mean(posterior$profit_diff < 0)
```
Our analysis showed that text ads will bring probably more profit but the result 
is too uncertain and we need more data for making the correct decision.  

**Another model for another case**  

We would like to put up an ad on the site and pay for it per day.  
The site admin promise that we will get 19 clicks per day.  
How many daily clicks should we expect on average?  
We are going to use **Poisson distribution** for our new model which correspond 
to *number of successes in a period of time*.  

```{r}
n <- 100000
mean_clicks <- runif(n, min = 0, max = 80) # uniform dist between 0 and 80 clicks
# model
n_visitors <- rpois(n = n, mean_clicks)

prior <- data.frame(mean_clicks, n_visitors)

# first day trial showed 13 clicks
posterior <- prior[prior$n_visitors == 13, ]

# visualize prior and posterior
par(mfrow=c(1,2))
hist(prior$mean_clicks)
hist(posterior$mean_clicks)
```

## Grid approximation  

The following example shows how we can build the previous model more effective.  
We calculating the **distribution** using `dbinom` instead of **simulating** 
using `rbinom`.  
We can directly include condition in our model using `dbinom` instead of `rbinom`.  

Instead of using resul of our trial *13 visitors after showing ads 100 times* we 
can condition for all posible number of visitors `seq(0,100)`.  

```{r}
n_ads_shown <- 100
p_clicks <- 0.1
n_visitors <- seq(0, 100) # instead of 13 we get for all possible n visitors
# model
prob <- dbinom(n_visitors,
               size = n_ads_shown, 
               prob = p_clicks)
head(prob)

# Plot the distribution
plot(n_visitors, prob, type = "h")
```

**Calculating a joint distribution**  

```{r}
n_ads_shown <- 100
p_clicks <- seq(0, 1, by = 0.01)
n_visitors <- seq(0, 100, by = 1)

# Define a grid over all the parameter combinations you need to evaluate
pars <- expand.grid(proportion_clicks = p_clicks,
                    n_visitors = n_visitors)

pars$prior <- dunif(pars$proportion_clicks, min = 0, max = 0.2)

pars$likelihood <- dbinom(pars$n_visitors, 
                          size = n_ads_shown, 
                          prob = pars$proportion_clicks)

### According to Bayes' theorem:
# Combined probability by the rule of probabilities multiplication
pars$probability <- pars$likelihood * pars$prior
# normalize to the total number to get sum of all probabilities eq 1
pars$probability <- pars$probability / sum(pars$probability)
###

# Conditioning on the data for n_visitors == 6  
pars <- pars[pars$n_visitors == 6, ]

# Normalize again to get sum of all probabilities eq 1
pars$probability <- pars$probability / sum(pars$probability)

# Plot the posterior pars$probability
plot(pars$proportion_clicks, pars$probability, type = "h")
```

We can directly condition and change `n_visitors <- 6`. In this case we do not 
need to subset by `pars[pars$n_visitors == 6, ]`. Result will be the same.  

## Model of birth weights using normal distribution

Let's assume that the Normal distribution is a decent model of birth weight data.  

```{r}
# Assign mu and sigma
m <- 3500   # central value (mean weight)
s <- 600    # deviation

weight_dist <- rnorm(n = 100000, mean = m, sd = s)
hist(weight_dist, 60, xlim = c(0, 6000))
```

We calculating the **distribution** using `dnorm` instead of **simulating** using `rnorm`.  

```{r}
# Create weight
weight <- seq(0, 6000, by = 100)  # 100 g increment

# Calculate likelihood
likelihood <- dnorm(weight, m, s)

# Plot the distribution of weight
plot(weight, likelihood, type = "h")
```

Here is a small data set with the birth weights of six newborn babies in grams.  
`c(3164, 3362, 4435, 3542, 3578, 4529)`  

Mark: What to do with the data? Should we condition it to get posterior?  

## A Bayesian model of Zombie IQ
Check video 
```{r, bi-432}
temp <- c(19, 23, 20, 17, 23)
mu <- seq(8, 30, by = 0.5)
sigma <- seq(0.1, 10, by = 0.3)
pars <- expand.grid(mu = mu, sigma = sigma)

pars$mu_prior    <- dnorm(pars$mu,    mean = 18, sd = 5)
pars$sigma_prior <- dunif(pars$sigma, min = 0, max = 10)

pars$prior <- pars$mu_prior * pars$sigma_prior

for (i in 1:nrow(pars)) {
    likelihoods <- dnorm(temp, pars$mu[i], pars$sigma[i])
    pars$likelihood[i] <- prod(likelihoods)
}

pars$probability <- pars$likelihood * pars$prior
pars$probability <- pars$probability / sum(pars$probability)

sample_indices <- sample(1:nrow(pars), size = 10000, 
                         replace = TRUE, prob = pars$probability)
head(sample_indices)
pars_sample <- pars[sample_indices, c("mu", "sigma")]
head(pars_sample)

hist(pars_sample$mu, 30)
quantile(pars_sample$mu, c(0.05, 0.95))

pred_temp <- rnorm(10000, mean = pars_sample$mu, sd = pars_sample$sigma)
hist(pred_temp, 30)

# Probability of 18C
sum(pred_temp >= 18) / length(pred_temp)
```
for the model of temperature but use zombies IQs instead. 
Define model:  
$\mu = Normal(mean:18, sd:5)$
$\sigma = Uniform(min:0, max:10)$
$temp = 19,23,...$
How much we can learn about the mean zombie IQ from this data. 
We need to calculate the probability of each parameter combination in pars.  
Use Bayes Theorem to calculate these probabilities and assign them to 
`pars$probability` to complete the model.  

```{r, bi-476}
# The IQ of a bunch of zombies
iq <- c(55, 44, 34, 18, 51, 40, 40, 49, 48, 46)

# Defining the parameter grid
pars <- expand.grid(mu = seq(0, 150, length.out = 100), 
                    sigma = seq(0.1, 50, length.out = 100))

# Defining and calculating the prior density for each parameter combination
pars$mu_prior <- dnorm(pars$mu, mean = 100, sd = 100)
pars$sigma_prior <- dunif(pars$sigma, min = 0.1, max = 50)
pars$prior <- pars$mu_prior * pars$sigma_prior

# Calculating the likelihood for each parameter combination
for(i in 1:nrow(pars)) {
  likelihoods <- dnorm(iq, pars$mu[i], pars$sigma[i])
  pars$likelihood[i] <- prod(likelihoods)
}

# Calculate the probability of each parameter combination
pars$probability <- pars$likelihood * pars$prior / sum(pars$likelihood * pars$prior)
```

Calculate new parameters.  

```{r}
head(pars)
sample_indices <- sample(nrow(pars), size = 10000,
                         replace = TRUE, prob = pars$probability)
head(sample_indices)

# Sample from pars to calculate some new measures
pars_sample <- pars[sample_indices, c("mu", "sigma")]

# Calculate quantiles
quantile(pars_sample$mu, c(0.025, 0.5, 0.975))

head(pars_sample)
pred_iq <- rnorm(10000, 
                 mean = pars_sample$mu, 
                 sd = pars_sample$sigma)

# Calculate the probability that zombi has IQ > 60
sum(pred_iq >= 60) / length(pred_iq)

par(mfrow=c(1,2))
# Visualize the mean IQ
hist(pars_sample$mu, 100)
# Visualize pred_iq
hist(pred_iq)
```

## The BEST models 
The t-test is a classical statistical procedure used to compare the 
means of two data sets.  

In 2013 John Kruschke developed a souped-up Bayesian version of the 
t-test he named **BEST** (standing for **B**ayesian **E**stimation **S**upersedes 
the **t**-test).  

We would like to compair IQ of two groups of 10 patiens on different diets (a and b).  

```{r}
# The IQ of patients.
iq_a <- c(55, 44, 34, 18, 51, 40, 40, 49, 48, 46)
iq_b <- c(44, 52, 42, 66, 53, 42, 55, 57, 56, 51)

# Calculate the mean difference in IQ between the two groups
mean(iq_b) - mean(iq_a)

```{r, message=FALSE}
require(BEST)

# Fit the BEST model to the data from both groups
require(BEST)
best_posterior <- BEST::BESTmcmc(iq_b, iq_a)
```
```{r}
# Plot the model result
plot(best_posterior)
```

The Bayesian model behind BEST assumes that the generative model for 
the data is a **t-distribution**; a more flexible distribution than the normal 
distribution as it assumes that data points might be outliers to some degree. 
This makes BEST's estimate of the mean difference robust to outliers in the data.  

**Sources**  

[Fundamentals of Bayesian Data Analysis in R on Datacamp](https://learn.datacamp.com/courses/fundamentals-of-bayesian-data-analysis-in-r)

<!--chapter:end:42_bayesian_inference.Rmd-->

# Naive Bayes classifiers

**Naive Bayes classifiers** are a family of simple "probabilistic classifiers" 
based on applying Bayes' theorem with strong (na√Øve) independence assumptions 
between the features.  

They are among the simplest Bayesian network models, but coupled with kernel 
density estimation, they can achieve higher accuracy levels.  

Na√Øve Bayes classifiers are highly scalable, requiring a number of parameters 
linear in the number of variables (features/predictors) in a learning problem. 
Maximum-likelihood training can be done by evaluating a closed-form expression, 
which takes linear time, rather than by expensive iterative approximation as 
used for many other types of classifiers.  

$P(c|x) = \frac{P(x|c)(P(c))}{P(x)}$, where  
$P(c|x)$ - posteriour probability  
$P(x|c)$ - Likelihood  
$P(c)$ - Class Prior Probbility  
$P(x)$ - Predictor Prior Probability  

<!--chapter:end:43_naive_bayes.Rmd-->

# Markov Chain Monte Carlo (MCMC)

We have a seq of conditional probabilities:  
We know conditional probabilities for the weather tomorrow (t+1) depending of the weather today (t). R for raily day and S for sunny day.  
$P(S_{t+1} | R_t) = 0.5$  
$P(R_{t+1} | R_t) = 0.5$  
$P(R_{t+1} | S_t) = 0.1$  
$P(S_{t+1} | S_t) = 0.9$  


What is the probability of sunny (S) or rainy (R) day?  
Simulate data using conditional probabilities:  
Markov Chain 1:  
S-S-R-R-S-S-S-S-R-R-R-S-S-S

Markov Chain 2:  
R-S-S-S-S-S-R-S-S-S-S-S-R-R  

After simulation we calculate probabilities of sunny (S) and rainy (R) days:  
P(S) = 0.833
P(R) = 0.167

<!--chapter:end:44_monte_carlo.Rmd-->

# Simple Markov process


Here, we will consider a simple example of Markov process with implementation in R.  
The following example is taken from [Bodo Winter website](http://www.bodowinter.com).  

A **Markov process** is characterized by (1) **a finite set of states** and (2) **fixed transition probabilities between the states**.  

Let's consider an example. Assume you have a classroom, with students who could be either in the state **alert** or in the state **bored**. And	 then, at any given time point, there's a certain probability of an alert student becoming bored (say 0.2), and there's a probability of a  bored student becoming alert (say 0.25).  

Let's say there are 20 alert and 80 bored students in a particular class. This is your initial condition at time point	$t$. Given the	transition probabilities above,	what's the number of alert and bored students at the next point in time, $t+1$?  	
Multiply 20 by 0.2 (=4) and these will be the alert students that	turn bored.	 
And	then multiply 80 by 0.25 (=20) and these will be the bored students that turn alert.  
So, at $t+1$, there's going to be 20-4+20 alert students. And there's going to be 80+4-20 bored students. Before, 80% of the students were bored and now, only 64% of the students are bored. Conversely, 36% are alert.  

A handy way of representing this Markov process is by defining a transition probability matrix:  

|           | A   | B    |
|-----------|-----|------|
| A$_{t+1}$ | 0.8 | 0.25 |
| B$_{t+1}$ | 0.2 | 0.75 |

What this matrix says is: A proportion of 0.8 of the people who are in state A (alert) will also be at state A at time point $t+1$. And, a proportion of 0.25 of the people who are in state B (bored) will switch to alert at t+1. This is what the first row says. The next row is simply one minus the probabilities of the first row, because probabilities (or proportions) have to add up to 1. Now think about multiplying this matrix with the initial proportions of alert and bored students that we had above. 0.8 are bored and 0.2 are alert. In linear algebra this would look the following way:  

<div>
$$
\begin{bmatrix}
 0.8 & 0.25 \\
 0.2 & 0.75
\end{bmatrix}\times\begin{bmatrix}
 0.2 \\
 0.8
\end{bmatrix} = \begin{bmatrix}
 0.8\times0.2 + 0.25\times0.8 \\
 0.2\times0.2 + 0.75\times0.8
\end{bmatrix} = \begin{bmatrix}
0.36 \\
0.64
\end{bmatrix}
$$
</div>

The results of these calculations are exactly the proportions that we saw above: 36% alert student and 64% bored students.  

Now, you might ask yourself: What happens if this process continues? What happens at $t+2$, $t+3$ etc.? Will it be the case that at one point there are no bored students any more? Let's simulate this in R and find out! Let's call this **tpm** for **transition probability matrix**:  

```{R}
tpm = matrix(c(0.8,0.25, 0.2,0.75), nrow=2, byrow=TRUE)
colnames(tpm) = c('A','B')
rownames(tpm) = c('At+1', 'Bt+1')
tpm
```

Again this matrix shows that 0.8 students who were in state A at time point t will still be in state A at $t+1$. And 0.25 students who were in state B at time point t will be in state A at $t+1$. The second row has a similar interpretation for alert and bored students becoming bored at $t+1$. Remember that Markov processes assume fixed transition probabilities. This means that in the simulation that we'll be doing, we leave the transition probability matrix unchanged. However, we will define a vector of the actual proportions ‚Äì and these are allowed to change. In time, we expect more and more students to become alert, because the transition probability from B to A (which, to remind you, was 0.25) is higher than from A to B (which was 0.2).  

Let's start our simulation by setting the initial condition as 0.1 students are alert and 0.9 students are bored and define a matrix called **sm** (short for **student matrix**):  
```{R}
sm = as.matrix(c(0.1, 0.9))
rownames(sm)= c('A', 'B')
sm
```
Now let's repeat by looping:

```{R}
for(i in 1:10){
    sm = tpm %*% sm
    }
```
Here, we're looping 10 times and on each iteration, we multiply the matrix **tpm** with the student matrix **sm**. We take this result and store it in **sm**. This means that at the next iteration, our fixed **transition probability matrix** will be multiplied by a different student matrix, allowing for the proportions to slowly change over time.  
R operator '%*%' is used for matrix multiplication

Outcome of our ten loop iterations:
```{R}
sm
```

So, after 10 iterations of the Markov process, we now have about 55% alert students and 45% bored ones. What is interesting to me is that even though 80% of the people who are alert at one time point remain alert at the next time point, the process only converged on 55% alert and 45% bored after 10 iterations.

Let's reset our initial condition to (0.1 alert and 0.9 bored students) and run a thousand iterations. 
```{R}
for(i in 1:1000){
    sm = tpm %*% sm
    }
sm
```

A 1000 iterations, and we seem to be zoning in onto ~55% and ~44%. This phenomenon is called **Markov convergence**. You could run even more iterations, and your outcome would get closer and closer to 0.5555 (to infinity). So, the model converges on an equilibrium. However, this is not a fixed equilibrium. It's not the case that the Markov process comes to a hold or that nobody changes states between alertness and boredness any more. The equilibrium that we're dealing with here is a statistical equilibrium, where the proportions of alert and bored students remain the same. but there still is constant change (at each time step, 0.2 alert students become bored and 0.25 bored students become alert). Markov models always converge to a statistical equilibrium if the conditions (1) and (2) above are met, and if you can get from any state within your Markov model to any other state (in the case of just two states, that clearly is the case). What's so cool about this is that it is, at first sight, fairly counterintuitive.

At least when I thought about the transition probabilities for the first time, I somehow expected all students to become alert but as we saw, that's not the case. Moreover, this process is not sensitive to initial conditions. That means that when you start with any proportion of alert or bored students (even extreme ones such as 0.0001 alert students), the process will reach the statistical equilibrium ‚Äì albeit sometimes a little faster or slower. You can play around with different values for the **sm** object to explore this property of Markov convergence. Another interesting thing is that the process is impervious to intervention: Say, you introduced something that made more students alert ‚Äì the Markov model would quickly get back to equilibrium. So Markov processes are essentially ahistorical processes: history doesn't matter. Even with extreme initial conditions or extreme interventions, the process quickly converges to the equilibrium defined by the transition probabilities. The only way to persistently change the system is to change the transition probabilities. Finally, what I find so cool about Markov processes is their computational simplicity.  

### Sources
[Bodo Winter website](http://www.bodowinter.com)

<!--chapter:end:45_simple_markov_process.Rmd-->

# Tree-based models

Classification and regression trees (CART) are a non-parametric decision tree learning technique that produces either classification or regression trees, depending on whether the dependent variable is categorical or numeric, respectively.  

CART is both a generic term to describe tree algorithms and also a specific name for Breiman‚Äôs original algorithm for constructing classification and regression trees.  

* **Decision Tree**: A tree-shaped graph or model of decisions used to determine a course of action or show a statistical probability.  
* **Classification Tree**: A decision tree that performs classification (predicts a categorical response).  
* **Regression Tree**: A decision tree that performs regression (predicts a numeric response).  
* **Split Point**: A split point occurs at each node of the tree where a decision is made (e.g. x > 7 vs. x ‚â§ 7).  
* **Terminal Node**: A terminal node is a node which has no descendants (child nodes). Also called a ‚Äúleaf node.‚Äù  

**Properties of Trees**  
* Can handle huge datasets.  
* Can handle mixed predictors implicitly ‚Äì numeric and categorical.  
* Easily ignore redundant variables.  
* Handle missing data elegantly through surrogate splits.  
* Small trees are easy to interpret.  
* Large trees are hard to interpret.  
* Prediction performance is often poor (high variance).  

**Tree Algorithms**  
There are a handful of different tree algorithms in addition to Breiman‚Äôs original CART algorithm. Namely, ID3, C4.5 and C5.0, all created by Ross Quinlan. C5.0 is an improvement over C4.5, however, the C4.5 algorithm is still quite popular since the multi-threaded version of C5.0 is proprietary (although the single threaded is released as GPL).  

**CART vs C4.5**  
Here are some of the differences between CART and C4.5:

* Tests in CART are always binary, but C4.5 allows two or more outcomes.  
* CART uses the Gini diversity index to rank tests, whereas C4.5 uses information-based criteria.  
* CART prunes trees using a cost-complexity model whose parameters are estimated by cross-validation; C4.5 uses a single-pass algorithm derived from binomial confidence limits.  
* With respect to missing data, CART looks for surrogate tests that approximate the outcomes when the tested attribute has an unknown value, but C4.5 apportions the case probabilistically among the outcomes.  

Decision trees are formed by a collection of rules based on variables in the modeling data set:  

1. Rules based on variables‚Äô values are selected to get the best split to differentiate observations based on the dependent variable.  
2. Once a rule is selected and splits a node into two, the same process is applied to each ‚Äúchild‚Äù node (i.e. it is a recursive procedure).  
3. Splitting stops when CART detects no further gain can be made, or some pre-set stopping rules are met. (Alternatively, the data are split as much as possible and then the tree is later pruned.)  

Each branch of the tree ends in a terminal node. Each observation falls into one and exactly one terminal node, and each terminal node is uniquely defined by a set of rules.  

## Classification Tree example
Let's use the data frame kyphosis to predict a type of deformation (kyphosis) after surgery, from age in months (Age), number of vertebrae involved (Number), and the highest vertebrae operated on (Start).

```{r}
# Classification Tree with rpart
library(rpart)

# grow tree
fit <- rpart(Kyphosis ~ Age + Number + Start,
   method="class", data=kyphosis)

printcp(fit) # display the results
plotcp(fit) # visualize cross-validation results
summary(fit) # detailed summary of splits

# plot tree
plot(fit, uniform=TRUE,
   main="Classification Tree for Kyphosis")
text(fit, use.n=TRUE, all=TRUE, cex=.8)

# create attractive postscript plot of tree
post(fit, title = "Classification Tree for Kyphosis")

# prune the tree
pfit<- prune(fit, cp=fit$cptable[which.min(fit$cptable[,"xerror"]),"CP"])

# plot the pruned tree
#FIXME: pfit is not a tree just a root error
#plot(pfit, uniform=TRUE,
#   main="Pruned Classification Tree for Kyphosis")
#text(pfit, use.n=TRUE, all=TRUE, cex=.8)
#post(pfit, file = "c:/ptree.ps",
#   title = "Pruned Classification Tree for Kyphosis")
```

## Regression Tree example
```{r}
# Regression Tree Example
library(rpart)

# grow tree
fit <- rpart(Mileage~Price + Country + Reliability + Type,
   method="anova", data=cu.summary)

printcp(fit) # display the results
plotcp(fit) # visualize cross-validation results
summary(fit) # detailed summary of splits

# create additional plots
par(mfrow=c(1,2)) # two plots on one page
rsq.rpart(fit) # visualize cross-validation results  

# plot tree
plot(fit, uniform=TRUE,
   main="Regression Tree for Mileage ")
text(fit, use.n=TRUE, all=TRUE, cex=.8)

# create attractive postcript plot of tree
post(fit, file = "tree2.ps",
   title = "Regression Tree for Mileage ")

# prune the tree
pfit<- prune(fit, cp=0.01160389) # from cptable   

# plot the pruned tree
plot(pfit, uniform=TRUE,
   main="Pruned Regression Tree for Mileage")
text(pfit, use.n=TRUE, all=TRUE, cex=.8)
post(pfit, file = "ptree2.ps",
   title = "Pruned Regression Tree for Mileage")
```

**Sources**  
[Tree-Based Models at Quick-R by datacamp](https://www.statmethods.net/advstats/cart.html)  
[UseR! Machine Learnign Turorial](https://koalaverse.github.io/machine-learning-in-R/decision-trees.html)

<!--chapter:end:46_tree-based_models.Rmd-->

